<!DOCTYPE html>
<html>
  <!-- meta/link... -->
  



<head>
  <meta http-equiv="content-type" content="text/html; charset=utf-8">
  <meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport">
  <!-- Global site tag (gtag.js) - Google Analytics -->

<script async src="https://www.googletagmanager.com/gtag/js?id=UA-139249935-1"></script>
<script>
    window.dataLayer = window.dataLayer || [];
    function gtag() {
        dataLayer.push(arguments);
    }

    gtag('js', new Date());
    gtag('config', 'UA-139249935-1');
</script>


  <title>Segmentation modelså®æˆ˜ | äº†å°˜å…°è‹¥çš„å°å‘</title>

  <link rel="icon" type="image/x-icon, image/vnd.microsoft.icon" href="/favicon.ico">
  <link rel="stylesheet" href="https://at.alicdn.com/t/font_1911880_c1nvbyezg17.css">
  <link href="https://unpkg.com/@fortawesome/fontawesome-free/css/all.min.css" rel="stylesheet">
  <link href="/js/swiper/swiper@5.4.1.min.css" rel="stylesheet">
  
  
    <script>
        var themeModelId = '';
        if (themeModelId) {
            localStorage.setItem('modelId', themeModelId);
        }
    </script>
    
    <script async src="https://cdn.jsdelivr.net/gh/yuang01/live2d-widget@latest/autoload.js"></script>
    <script>
        var live2dOpen = eval('true') || false;
        if (!live2dOpen) {
            localStorage.setItem('waifu-display', 1609323474481);
        }
    </script>
  
  
  
<link rel="stylesheet" href="/css/animate.min.css">

  
<link rel="stylesheet" href="/css/style.css">

  
  
    <link href="https://unpkg.com/@fancyapps/ui@5.0/dist/fancybox/fancybox.css" rel="stylesheet">
  
  
    
<link rel="stylesheet" href="/js/shareJs/share.min.css">

  
  <style>
        @media (max-width: 992px) {
            #waifu {
                display: none;
            }
        }
    </style>
    <script defer src="//cdn.bootcss.com/pace/1.0.2/pace.min.js"></script>
    <link href="//cdn.bootcss.com/pace/1.0.2/themes/pink/pace-theme-flash.css" rel="stylesheet">

    
    
    <!-- ä¾èµ–äºjqueryå’Œvue -->
    <script src="https://unpkg.com/jquery@3.5.1/dist/jquery.min.js"></script>
    <script src="https://unpkg.com/vue@2.6.11/dist/vue.min.js"></script>

    <!-- import link -->
    
        
            
        
            
        
    
    <!-- import script -->
    
        
            
        
            
        
    

<meta name="generator" content="Hexo 6.3.0"><link rel="alternate" href="/atom.xml" title="äº†å°˜å…°è‹¥çš„å°å‘" type="application/atom+xml">
</head>

  
  <!-- é¢„åŠ è½½åŠ¨ç”» -->
  
  
  <div class="preloader_2" id="loader">
  <div class="loader"></div>
</div>

  <script>
    var endLoading = function () {
      document.body.style.overflow = 'auto';
      document.getElementById('loader').classList.add("loaded");
    }
    window.addEventListener('DOMContentLoaded', endLoading);
  </script>


  <body>
    <!-- åˆ¤æ–­æ˜¯å¦ä¸ºæš—é»‘é£æ ¼ -->
    <!-- åˆ¤æ–­æ˜¯å¦ä¸ºé»‘å¤œæ¨¡å¼ -->
<script defer>
  let isDark = JSON.parse(localStorage.getItem('dark')) || JSON.parse('false');

  if (isDark) {
    $(document.body).addClass('darkModel');
  }
</script>

    <!-- éœ€è¦åœ¨ä¸Šé¢åŠ è½½çš„js -->
    <script>
  function loadScript(src, cb) {
    return new Promise(resolve => {
      setTimeout(function () {
        var HEAD = document.getElementsByTagName("head")[0] || document.documentElement;
        var script = document.createElement("script");
        script.setAttribute("type", "text/javascript");
        if (cb) {
          if (JSON.stringify(cb)) {
            for (let p in cb) {
              if (p == "onload") {
                script[p] = () => {
                  cb[p]()
                  resolve()
                }
              } else {
                script[p] = cb[p]
                script.onload = resolve
              }
            }
          } else {
            script.onload = () => {
              cb()
              resolve()
            };
          }
        } else {
          script.onload = resolve
        }
        script.setAttribute("src", src);
        HEAD.appendChild(script);
      });
    });
  }

  //https://github.com/filamentgroup/loadCSS
  var loadCSS = function (src) {
    return new Promise(resolve => {
      setTimeout(function () {
        var link = document.createElement('link');
        link.rel = "stylesheet";
        link.href = src;
        link.onload = resolve;
        document.getElementsByTagName("head")[0].appendChild(link);
      });
    });
  };

</script> 

<!-- è½®æ’­å›¾æ‰€éœ€è¦çš„js -->
<script src="/js/swiper/swiper.min.js"></script>
<script src="/js/swiper/vue-awesome-swiper.js"></script>
<script src="/js/swiper/swiper.animate1.0.3.min.js"></script>

<script type="text/javascript">
  Vue.use(window.VueAwesomeSwiper)
</script>


  <script src="/js/vue-typed-js/index.js"></script>


<!-- é¦–é¡µçš„å…¬å‘Šæ»šåŠ¨æ’ä»¶çš„jséœ€è¦é‡æ–°åŠ è½½ -->
<script src="/js/vue-seamless-scroll/index.js"></script>

<!-- æ‰“å­—æœºæ•ˆæœjs -->
<script src="https://unpkg.com/typed.js@2.0.11"></script>


    <div id="safearea">
      <main class="main" id="pjax-container">
        <!-- å¤´éƒ¨å¯¼èˆª -->
        
<header class="header  " 
  id="navHeader"
  style="position: fixed;
  left: 0; top: 0; z-index: 10;width: 100%;"
>
  <div class="header-content">
    <div class="bars">
      <div id="appDrawer" class="sidebar-image">
  <div class="drawer-box-icon">
    <i class="fas fa-bars" aria-hidden="true" @click="showDialogDrawer"></i>
  </div>
  
  <transition name="fade">
    <div class="drawer-box_mask" v-cloak style="display: none;" v-show="visible" @click.self="cancelDialogDrawer">
    </div>
  </transition>
  <div class="drawer-box" :class="{'active': visible}">
    <div class="drawer-box-head bg-color">
      <img class="drawer-box-head_logo lazyload placeholder" src="/medias/head.jpg" class="lazyload placeholder" data-srcset="/medias/head.jpg" srcset="https://pic1.zhimg.com/v2-cd38920285d125be80b3eb504052c550_b.webp" alt="logo">
      <h3 class="drawer-box-head_title">äº†å°˜å…°è‹¥çš„å°å‘</h3>
      <h5 class="drawer-box-head_desc">åˆ†äº«å¾®ç”Ÿç‰©ç”Ÿç‰©ä¿¡æ¯å­¦åˆ†ææ–¹æ³•ï¼Œæ¬¢è¿åŠ å…¥QQç¾¤äº¤æµ945751012ï¼Œä¸æ¥å—ç¾¤å†…å¹¿å‘Šï¼</h5>
    </div>
    
    <div class="drawer-box-content">
      <ul class="drawer-box-content_menu">
        
          
            <li class="drawer-box-content_item" style="position: relative;">
              
                <a href="/" class="drawer-menu-item-link">
                  
                    <i class="fas fa-home" aria-hidden="true"></i>
                  
                  <span class="name">é¦–é¡µ</span>
                </a>
              
            </li>
          
            <li class="drawer-box-content_item" style="position: relative;">
              
                <a href="javascript:;" class="drawer-menu-item-link has-children" @click="openOrCloseMenu(1)">
                  <span>
                    
                      <i class="fas fa-archive"></i>
                    
                    <span class="name">å½’æ¡£</span>
                  </span>
                  <i class="fas fa-chevron-left arrow " :class="{'icon-rotate': isOpen(1)}" aria-hidden="true"></i>
                </a>
                <ul class="drawer-sub-menu" v-if="isOpen(1)">
                  
                  <li>
                    <a href="/archives">
                      
                      <i class="fas fa-archive" style="margin-top: -20px;"></i>
                      
                      <span>Archives</span>
                    </a>
                  </li>
                  
                  <li>
                    <a href="/categories">
                      
                      <i class="fas fa-bookmark" style="margin-top: -20px;"></i>
                      
                      <span>Categories</span>
                    </a>
                  </li>
                  
                  <li>
                    <a href="/tags">
                      
                      <i class="fas fa-tags" style="margin-top: -20px;"></i>
                      
                      <span>Tags</span>
                    </a>
                  </li>
                  
                </ul>
              
            </li>
          
            <li class="drawer-box-content_item" style="position: relative;">
              
                <a href="javascript:;" class="drawer-menu-item-link has-children" @click="openOrCloseMenu(2)">
                  <span>
                    
                      <i class="fas fa-user"></i>
                    
                    <span class="name">å…³äº</span>
                  </span>
                  <i class="fas fa-chevron-left arrow " :class="{'icon-rotate': isOpen(2)}" aria-hidden="true"></i>
                </a>
                <ul class="drawer-sub-menu" v-if="isOpen(2)">
                  
                  <li>
                    <a href="/about">
                      
                      <i class="fas fa-user-secret" style="margin-top: -20px;"></i>
                      
                      <span>About</span>
                    </a>
                  </li>
                  
                  <li>
                    <a href="/me">
                      
                      <i class="fas fa-user-md" style="margin-top: -20px;"></i>
                      
                      <span>æˆ‘</span>
                    </a>
                  </li>
                  
                  <li>
                    <a href="/skip/box">
                      
                      <i class="fas fa-box-open" style="margin-top: -20px;"></i>
                      
                      <span>é­”ç›’</span>
                    </a>
                  </li>
                  
                </ul>
              
            </li>
          
            <li class="drawer-box-content_item" style="position: relative;">
              
                <a href="javascript:;" class="drawer-menu-item-link has-children" @click="openOrCloseMenu(3)">
                  <span>
                    
                      <i class="fas fa-link"></i>
                    
                    <span class="name">æ›´å¤š</span>
                  </span>
                  <i class="fas fa-chevron-left arrow " :class="{'icon-rotate': isOpen(3)}" aria-hidden="true"></i>
                </a>
                <ul class="drawer-sub-menu" v-if="isOpen(3)">
                  
                  <li>
                    <a href="/gallery">
                      
                      <i class="fas fa-camera-retro" style="margin-top: -20px;"></i>
                      
                      <span>å›¾åº“</span>
                    </a>
                  </li>
                  
                  <li>
                    <a target="_blank" rel="noopener" href="https://liaochenlanruo.fun/reference/">
                      
                      <i class="fab fa-unity" style="margin-top: -20px;"></i>
                      
                      <span>é€ŸæŸ¥è¡¨</span>
                    </a>
                  </li>
                  
                  <li>
                    <a href="/friends">
                      
                      <i class="fas fa-link" style="margin-top: -20px;"></i>
                      
                      <span>å‹æƒ…è¿æ¥</span>
                    </a>
                  </li>
                  
                </ul>
              
            </li>
          
            <li class="drawer-box-content_item" style="position: relative;">
              
                <a target="_blank" rel="noopener" href="https://liaochenlanruo.gitee.io" class="drawer-menu-item-link">
                  
                    <i class="fas fa-external-link" aria-hidden="true"></i>
                  
                  <span class="name">English</span>
                </a>
              
            </li>
          
            <li class="drawer-box-content_item" style="position: relative;">
              
                <a target="_blank" rel="noopener" href="https://www.travellings.cn/go-by-clouds.html" class="drawer-menu-item-link">
                  
                    <i class="fas fa-subway" aria-hidden="true"></i>
                  
                  <span class="name">å¼€å¾€</span>
                </a>
              
            </li>
          
        
        
          <li class="drawer-box-content_item">
            <a target="_blank" rel="noopener" href="http://github.com/liaochenlanruo">
              <i class="fas fa-github" aria-hidden="true"></i>
              <span>Github</span>
            </a>
          </li>
        
      </ul>
    </div>
  </div>
</div>

<script>
  var body = document.body || document.documentElement || window;
  var vm = new Vue({
    el: '#appDrawer',
    data: {
      visible: false,
      top: 0,
      openArr: [],
    },
    computed: {
    },
    mounted() {
    },
    methods: {
      isOpen(index) {
        if (this.openArr.includes(index)) {
          return true;
        } else {
          return false;
        }
      },
      openOrCloseMenu(curIndex) {
        const index = this.openArr.indexOf(curIndex);
        if (index !== -1) {
          this.openArr.splice(index, 1);
        } else {
          this.openArr.push(curIndex);
        }
      },
      showDialogDrawer() {
        this.visible = true;
        // é˜²æ­¢é¡µé¢æ»šåŠ¨ï¼Œåªèƒ½è®©å¼¹æ¡†æ»šåŠ¨
        this.top = $(document).scrollTop()
        body.style.cssText = 'width: 100%; height: 100%;overflow: hidden;';
      },
      cancelDialogDrawer() {
        this.visible = false;
        body.removeAttribute('style');
        $(document).scrollTop(this.top)
      }
    },
    created() {}
  })
</script>

    </div>
    <div class="blog-title" id="author-avatar">
      
        <div class="avatar">
          <img src="/medias/head.jpg" class="lazyload placeholder" data-srcset="/medias/head.jpg" srcset="https://pic1.zhimg.com/v2-cd38920285d125be80b3eb504052c550_b.webp" alt="logo">
        </div>
      
      <a href="/" class="logo">äº†å°˜å…°è‹¥çš„å°å‘</a>
    </div>
    <nav class="navbar">
      <ul class="menu">
        
          
            <li class="menu-item" style="position: relative;">
              
                <a href="/" class="menu-item-link" title="é¦–é¡µ">
                  
                    <i class="fas fa-home" aria-hidden="true"></i>
                  
                  <span class="name">é¦–é¡µ</span>
                </a>
              
            </li>
          
            <li class="menu-item" style="position: relative;">
              
                <a href="javascript:;" class="menu-item-link" title="å½’æ¡£">
                  
                    <i class="fas fa-archive"></i>
                  
                  <span class="name">å½’æ¡£</span>
                  <i class="fas fa-chevron-down arrow" aria-hidden="true"></i>
                </a>
                <ul class="sub-menu">
                  
                  <li>
                    <a href="/archives">
                      
                      <i class="fas fa-archive" style="margin-top: -20px;"></i>
                      
                      <span>Archives</span>
                    </a>
                  </li>
                  
                  <li>
                    <a href="/categories">
                      
                      <i class="fas fa-bookmark" style="margin-top: -20px;"></i>
                      
                      <span>Categories</span>
                    </a>
                  </li>
                  
                  <li>
                    <a href="/tags">
                      
                      <i class="fas fa-tags" style="margin-top: -20px;"></i>
                      
                      <span>Tags</span>
                    </a>
                  </li>
                  
                </ul>
              
            </li>
          
            <li class="menu-item" style="position: relative;">
              
                <a href="javascript:;" class="menu-item-link" title="å…³äº">
                  
                    <i class="fas fa-user"></i>
                  
                  <span class="name">å…³äº</span>
                  <i class="fas fa-chevron-down arrow" aria-hidden="true"></i>
                </a>
                <ul class="sub-menu">
                  
                  <li>
                    <a href="/about">
                      
                      <i class="fas fa-user-secret" style="margin-top: -20px;"></i>
                      
                      <span>About</span>
                    </a>
                  </li>
                  
                  <li>
                    <a href="/me">
                      
                      <i class="fas fa-user-md" style="margin-top: -20px;"></i>
                      
                      <span>æˆ‘</span>
                    </a>
                  </li>
                  
                  <li>
                    <a href="/skip/box">
                      
                      <i class="fas fa-box-open" style="margin-top: -20px;"></i>
                      
                      <span>é­”ç›’</span>
                    </a>
                  </li>
                  
                </ul>
              
            </li>
          
            <li class="menu-item" style="position: relative;">
              
                <a href="javascript:;" class="menu-item-link" title="æ›´å¤š">
                  
                    <i class="fas fa-link"></i>
                  
                  <span class="name">æ›´å¤š</span>
                  <i class="fas fa-chevron-down arrow" aria-hidden="true"></i>
                </a>
                <ul class="sub-menu">
                  
                  <li>
                    <a href="/gallery">
                      
                      <i class="fas fa-camera-retro" style="margin-top: -20px;"></i>
                      
                      <span>å›¾åº“</span>
                    </a>
                  </li>
                  
                  <li>
                    <a target="_blank" rel="noopener" href="https://liaochenlanruo.fun/reference/">
                      
                      <i class="fab fa-unity" style="margin-top: -20px;"></i>
                      
                      <span>é€ŸæŸ¥è¡¨</span>
                    </a>
                  </li>
                  
                  <li>
                    <a href="/friends">
                      
                      <i class="fas fa-link" style="margin-top: -20px;"></i>
                      
                      <span>å‹æƒ…è¿æ¥</span>
                    </a>
                  </li>
                  
                </ul>
              
            </li>
          
            <li class="menu-item" style="position: relative;">
              
                <a target="_blank" rel="noopener" href="https://liaochenlanruo.gitee.io" class="menu-item-link" title="English">
                  
                    <i class="fas fa-external-link" aria-hidden="true"></i>
                  
                  <span class="name">English</span>
                </a>
              
            </li>
          
            <li class="menu-item" style="position: relative;">
              
                <a target="_blank" rel="noopener" href="https://www.travellings.cn/go-by-clouds.html" class="menu-item-link" title="å¼€å¾€">
                  
                    <i class="fas fa-subway" aria-hidden="true"></i>
                  
                  <span class="name">å¼€å¾€</span>
                </a>
              
            </li>
          
        
      </ul>
      
      
        <div id="appSearch">
  <div class="search"  @click="showDialog()"><i class="fas fa-search" aria-hidden="true"></i></div>
  <transition name="fade">
    <div class="message-box_wrapper" style="display: none;" v-cloak v-show="dialogVisible" @click.self="cancelDialogVisible()">
      <div class="message-box animated bounceInDown">
        <h2>
          <span>
            <i class="fas fa-search" aria-hidden="true"></i>
            <span class="title">æœ¬åœ°æœç´¢</span>
          </span>
          <i class="fas fa-times close" pointer style="float:right;" aria-hidden="true" @click.self="cancelDialogVisible()"></i>
        </h2>
        <form class="site-search-form">
          <input type="text"
            placeholder="è¯·è¾“å…¥å…³é”®å­—"
            id="local-search-input" 
            @click="getSearchFile()"
            class="st-search-input"
            v-model="searchInput"
          />
        </form>
        <div class="result-wrapper">
          <div id="local-search-result" class="local-search-result-cls"></div>
        </div>
      </div>
    </div>
  </transition>
</div>
<script src="/js/local_search.js"></script>
<script>
  var body = document.body || document.documentElement || window;
  var vm = new Vue({
    el: '#appSearch',
    data: {
      dialogVisible: false,
      searchInput: '',
      top: 0,
    },
    computed: {
    },
    mounted() {
      window.addEventListener('pjax:complete', () => {
        this.cancelDialogVisible();
      })
    },
    methods: {
      showDialog() {
        this.dialogVisible = true;
        // é˜²æ­¢é¡µé¢æ»šåŠ¨ï¼Œåªèƒ½è®©å¼¹æ¡†æ»šåŠ¨
        this.top = $(document).scrollTop()
        body.style.cssText = 'overflow: hidden;';
      },
      getSearchFile() {
        if (!this.searchInput) {
          getSearchFile("/search.xml");
        }
      },
      cancelDialogVisible() {
        this.dialogVisible = false;
        body.removeAttribute('style');
        $(document).scrollTop(this.top)
      },
    },
    created() {}
  })
</script>
<!-- è§£å†³åˆ·æ–°é¡µé¢é—ªçƒé—®é¢˜ï¼Œå¯ä»¥åœ¨å…ƒç´ ä¸Šæ·»åŠ display: none, æˆ–è€…ç”¨vue.extendæ–¹æ³•ï¼Œè¯¦æƒ…ï¼šhttps://blog.csdn.net/qq_31393401/article/details/81017912 -->
<!-- ä¸‹é¢æ˜¯æœç´¢åŸºæœ¬å†™æ³• -->
<!-- <script type="text/javascript" id="local.search.active">
  var inputArea = document.querySelector("#local-search-input");
  inputArea.onclick   = function(){ getSearchFile(); this.onclick = null }
  inputArea.onkeydown = function(){ if(event.keyCode == 13) return false }
</script> -->

      

    </nav>
  </div>
  
    <a target="_blank" rel="noopener" href="http://github.com/liaochenlanruo" class="github-corner color-primary" aria-label="View source on GitHub"><svg width="60" height="60" viewBox="0 0 250 250" style="fill:#fff; position: absolute; top: 0; border: 0; right: 0;" aria-hidden="true"><path d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"></path><path d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5,78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5,87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2" fill="currentColor" style="transform-origin: 130px 106px;" class="octo-arm"></path><path d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3,112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z" fill="currentColor" class="octo-body"></path></svg></a><style>.github-corner:hover .octo-arm{animation:octocat-wave 560ms ease-in-out}@keyframes octocat-wave{0%,100%{transform:rotate(0)}20%,60%{transform:rotate(-25deg)}40%,80%{transform:rotate(10deg)}}@media (max-width:500px){.github-corner:hover .octo-arm{animation:none}.github-corner .octo-arm{animation:octocat-wave 560ms ease-in-out}}</style>
  
  
    <div id="he-plugin-simple"></div>
    <script defer>
      WIDGET = {
        CONFIG: {
          "modules": "012",
          "background": 5,
          "tmpColor": "4A4A4A",
          "tmpSize": 16,
          "cityColor": "4A4A4A",
          "citySize": 16,
          "aqiSize": 16,
          "weatherIconSize": 24,
          "alertIconSize": 18,
          "padding": "10px 10px 10px 10px",
          "shadow": "1",
          "language": "auto",
          "borderRadius": 5,
          "fixed": "false",
          "vertical": "middle",
          "horizontal": "center",
          "key": "2784dd3fcb1e4f0f9a9b579bf69641f2"
        }
      }
    </script>
    <script defer src="https://widget.qweather.net/simple/static/js/he-simple-common.js?v=2.0"></script> 
    
</header>
        <!-- å†…å®¹åŒºåŸŸ -->
        
<!-- prismjs ä»£ç é«˜äº® -->




<div class="bg-dark-floor" style="position: fixed;left: 0;top: 0;width: 100%;height: 100%;z-index: -1;"></div>


  <!-- æ–‡ç« è¯¦æƒ…é¡µé¡¶éƒ¨å›¾ç‰‡å’Œæ ‡é¢˜ -->




<div class="post-detail-header" id="thumbnail_canvas" style="background-repeat: no-repeat; background-size: cover; 
  background-position: center center;position: relative;background-image:url('https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/custom/bgs/thumb_230.webp')">
  <div class="post-detail-header-mask"></div>
  <canvas id="header_canvas"style="position:absolute;bottom:0;pointer-events:none;"></canvas>
  
  <div class="post-detail-header_info-box">
    <div class="title-box">
      <span class="title">
        Segmentation modelså®æˆ˜
      </span>
    </div>
    
    
      
        <span class="post-detail-header_date">
          <i class="fas fa-calendar"></i> å‘è¡¨äºï¼š2025-11-16 |
        </span>
      

      
        <span class="post-detail-header_categories">
          <i class="iconfont iconbookmark1"></i> åˆ†ç±»ï¼š
          
            <a href="/categories/BioAI/" class="post-detail-header_category">
              BioAI
            </a>
          
        </span>
      

      
        <div class="post-detail-header_wordcount">
          <span class="totalcount">
            <i class="fas fa-file-text-o"></i> å­—æ•°ç»Ÿè®¡: 7.3k |
          </span>
  
          <span class="min2read">
            <i class="fas fa-clock"></i> é˜…è¯»æ—¶é•¿: 39åˆ†é’Ÿ |
          </span>
  
          
            <span class="reading">
              <i class="fas fa-eye"></i> é˜…è¯»é‡ï¼š<span id="busuanzi_value_page_pv"></span>
            </span>
          
        </div>
      
    
  </div>
  
  
    <script defer src="/js/bubble/bubble.js"></script>
  
</div>





<div class="post-detail-content post-row" 
  style="padding-top: 0px;">
  <div class="main-content">
    <article class="post post-detail">
      <div class="post-content">
        <h1 id="1-Segmentation-models-ç®€ä»‹"><a href="#1-Segmentation-models-ç®€ä»‹" class="headerlink" title="1. Segmentation models ç®€ä»‹"></a>1. Segmentation models ç®€ä»‹</h1><p>ä½¿ç”¨Transformeréª¨å¹²ç½‘ç»œçš„åˆ†å‰²æ¨¡å‹ï¼ˆå¦‚Nucleotide Transformerã€Enformerã€Borzoiï¼‰å¯ç”¨äºå•æ ¸è‹·é…¸åˆ†è¾¨ç‡ä¸‹çš„åŸºå› ç»„å…ƒä»¶é¢„æµ‹ã€‚ä¾‹å¦‚ï¼Œ<code>SegmentNT</code>èƒ½åœ¨é•¿è¾¾30kbçš„åºåˆ—ï¼ˆå¯æ‰©å±•è‡³50kbpï¼‰ä¸­<strong>é¢„æµ‹14ç§ä¸åŒç±»åˆ«çš„äººç±»åŸºå› ç»„å…ƒä»¶</strong>ï¼Œå¹¶è¡¨ç°å‡ºä¼˜å¼‚çš„æ€§èƒ½ã€‚</p>
<p>æ‰€æœ‰æ¨¡å‹å‡æ­é…ä¸€ç»´U-Netåˆ†å‰²å¤´ï¼Œä»¥å•æ ¸è‹·é…¸åˆ†è¾¨ç‡é¢„æµ‹åºåˆ—ä¸­å¤šç§åŸºå› ç»„å…ƒä»¶çš„ä½ç½®ã€‚è¿™äº›å…ƒä»¶åŒ…æ‹¬åŸºå› å…ƒä»¶ï¼ˆè›‹ç™½è´¨ç¼–ç åŸºå› ã€é•¿é“¾éç¼–ç RNAã€5â€™éç¿»è¯‘åŒºã€3â€™éç¿»è¯‘åŒºã€å¤–æ˜¾å­ã€å†…å«å­ã€å‰ªæ¥å—ä½“ä½ç‚¹å’Œä¾›ä½“ä½ç‚¹ï¼‰å’Œè°ƒæ§å…ƒä»¶ï¼ˆpolyA signalã€ç»„ç»‡éç‰¹å¼‚æ€§å’Œç»„ç»‡ç‰¹å¼‚æ€§å¯åŠ¨å­åŠå¢å¼ºå­ï¼Œä»¥åŠCTCFç»“åˆä½ç‚¹ï¼‰ã€‚</p>
<ul>
<li>ğŸ“œ <strong><a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41592-025-02881-2">Read the Paper (Nature Methods 2025)</a></strong></li>
<li>ğŸ¤— <strong><a target="_blank" rel="noopener" href="https://huggingface.co/collections/InstaDeepAI/segmentnt-65eb4941c57808b4a3fe1319">SegmentNT Hugging Face Collection</a></strong></li>
<li>ğŸš€ <strong><a target="_blank" rel="noopener" href="https://colab.research.google.com/#fileId=https%3A//huggingface.co/InstaDeepAI/segment_nt/blob/main/inference_segment_nt.ipynb">SegmentNT Inference Notebook (HF)</a></strong></li>
</ul>
<img src="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/images/post/BioAI/03/Fig01.webp" class="lazyload placeholder" data-srcset="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/images/post/BioAI/03/Fig01.webp" srcset="https://pic1.zhimg.com/v2-cd38920285d125be80b3eb504052c550_b.webp" alt= "Performance on downstream tasks" width="600">

<p><em>Fig. 1: SegmentNT localizes genomics elements at nucleotide resolution.</em></p>
<h1 id="2-å¦‚ä½•ä½¿ç”¨-ğŸš€"><a href="#2-å¦‚ä½•ä½¿ç”¨-ğŸš€" class="headerlink" title="2. å¦‚ä½•ä½¿ç”¨ ğŸš€"></a>2. å¦‚ä½•ä½¿ç”¨ ğŸš€</h1><h2 id="2-1-å®‰è£…å¹¶åŠ è½½æ¨¡å—"><a href="#2-1-å®‰è£…å¹¶åŠ è½½æ¨¡å—" class="headerlink" title="2.1 å®‰è£…å¹¶åŠ è½½æ¨¡å—"></a>2.1 å®‰è£…å¹¶åŠ è½½æ¨¡å—</h2><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">!pip install boto3</span><br><span class="line">!pip install matplotlib</span><br><span class="line">!pip install biopython</span><br><span class="line">!pip install dm-haiku</span><br></pre></td></tr></table></figure>

<pre><code>Looking in indexes: https://pypi.org/simple, https://packagecloud.io/github/git-lfs/pypi/simple
Requirement already satisfied: boto3 in /home/hugo/anaconda3/envs/trix/lib/python3.10/site-packages (1.28.2)
Requirement already satisfied: botocore&lt;1.32.0,&gt;=1.31.2 in /home/hugo/anaconda3/envs/trix/lib/python3.10/site-packages (from boto3) (1.31.2)
Requirement already satisfied: jmespath&lt;2.0.0,&gt;=0.7.1 in /home/hugo/anaconda3/envs/trix/lib/python3.10/site-packages (from boto3) (1.0.1)
Requirement already satisfied: s3transfer&lt;0.7.0,&gt;=0.6.0 in /home/hugo/anaconda3/envs/trix/lib/python3.10/site-packages (from boto3) (0.6.1)
Requirement already satisfied: python-dateutil&lt;3.0.0,&gt;=2.1 in /home/hugo/anaconda3/envs/trix/lib/python3.10/site-packages (from botocore&lt;1.32.0,&gt;=1.31.2-&gt;boto3) (2.8.2)
Requirement already satisfied: urllib3&lt;1.27,&gt;=1.25.4 in /home/hugo/anaconda3/envs/trix/lib/python3.10/site-packages (from botocore&lt;1.32.0,&gt;=1.31.2-&gt;boto3) (1.26.16)
Requirement already satisfied: six&gt;=1.5 in /home/hugo/anaconda3/envs/trix/lib/python3.10/site-packages (from python-dateutil&lt;3.0.0,&gt;=2.1-&gt;botocore&lt;1.32.0,&gt;=1.31.2-&gt;boto3) (1.16.0)

[1m[[0m[34;49mnotice[0m[1;39;49m][0m[39;49m A new release of pip is available: [0m[31;49m24.1.1[0m[39;49m -&gt; [0m[32;49m24.1.2[0m
[1m[[0m[34;49mnotice[0m[1;39;49m][0m[39;49m To update, run: [0m[32;49mpip install --upgrade pip[0m
Looking in indexes: https://pypi.org/simple, https://packagecloud.io/github/git-lfs/pypi/simple
Requirement already satisfied: matplotlib in /home/hugo/anaconda3/envs/trix/lib/python3.10/site-packages (3.7.2)
Requirement already satisfied: contourpy&gt;=1.0.1 in /home/hugo/anaconda3/envs/trix/lib/python3.10/site-packages (from matplotlib) (1.1.0)
Requirement already satisfied: cycler&gt;=0.10 in /home/hugo/anaconda3/envs/trix/lib/python3.10/site-packages (from matplotlib) (0.11.0)
Requirement already satisfied: fonttools&gt;=4.22.0 in /home/hugo/anaconda3/envs/trix/lib/python3.10/site-packages (from matplotlib) (4.41.0)
Requirement already satisfied: kiwisolver&gt;=1.0.1 in /home/hugo/anaconda3/envs/trix/lib/python3.10/site-packages (from matplotlib) (1.4.4)
Requirement already satisfied: numpy&gt;=1.20 in /home/hugo/anaconda3/envs/trix/lib/python3.10/site-packages (from matplotlib) (1.25.1)
Requirement already satisfied: packaging&gt;=20.0 in /home/hugo/anaconda3/envs/trix/lib/python3.10/site-packages (from matplotlib) (23.1)
Requirement already satisfied: pillow&gt;=6.2.0 in /home/hugo/anaconda3/envs/trix/lib/python3.10/site-packages (from matplotlib) (10.0.0)
Requirement already satisfied: pyparsing&lt;3.1,&gt;=2.3.1 in /home/hugo/anaconda3/envs/trix/lib/python3.10/site-packages (from matplotlib) (3.0.9)
Requirement already satisfied: python-dateutil&gt;=2.7 in /home/hugo/anaconda3/envs/trix/lib/python3.10/site-packages (from matplotlib) (2.8.2)
Requirement already satisfied: six&gt;=1.5 in /home/hugo/anaconda3/envs/trix/lib/python3.10/site-packages (from python-dateutil&gt;=2.7-&gt;matplotlib) (1.16.0)

[1m[[0m[34;49mnotice[0m[1;39;49m][0m[39;49m A new release of pip is available: [0m[31;49m24.1.1[0m[39;49m -&gt; [0m[32;49m24.1.2[0m
[1m[[0m[34;49mnotice[0m[1;39;49m][0m[39;49m To update, run: [0m[32;49mpip install --upgrade pip[0m
Looking in indexes: https://pypi.org/simple, https://packagecloud.io/github/git-lfs/pypi/simple
Requirement already satisfied: biopython in /home/hugo/anaconda3/envs/trix/lib/python3.10/site-packages (1.81)
Requirement already satisfied: numpy in /home/hugo/anaconda3/envs/trix/lib/python3.10/site-packages (from biopython) (1.25.1)

[1m[[0m[34;49mnotice[0m[1;39;49m][0m[39;49m A new release of pip is available: [0m[31;49m24.1.1[0m[39;49m -&gt; [0m[32;49m24.1.2[0m
[1m[[0m[34;49mnotice[0m[1;39;49m][0m[39;49m To update, run: [0m[32;49mpip install --upgrade pip[0m
Looking in indexes: https://pypi.org/simple, https://packagecloud.io/github/git-lfs/pypi/simple
Requirement already satisfied: dm-haiku in /home/hugo/anaconda3/envs/trix/lib/python3.10/site-packages (0.0.9)
Requirement already satisfied: absl-py&gt;=0.7.1 in /home/hugo/anaconda3/envs/trix/lib/python3.10/site-packages (from dm-haiku) (1.4.0)
Requirement already satisfied: jmp&gt;=0.0.2 in /home/hugo/anaconda3/envs/trix/lib/python3.10/site-packages (from dm-haiku) (0.0.4)
Requirement already satisfied: numpy&gt;=1.18.0 in /home/hugo/anaconda3/envs/trix/lib/python3.10/site-packages (from dm-haiku) (1.25.1)
Requirement already satisfied: tabulate&gt;=0.8.9 in /home/hugo/anaconda3/envs/trix/lib/python3.10/site-packages (from dm-haiku) (0.9.0)

[1m[[0m[34;49mnotice[0m[1;39;49m][0m[39;49m A new release of pip is available: [0m[31;49m24.1.1[0m[39;49m -&gt; [0m[32;49m24.1.2[0m
[1m[[0m[34;49mnotice[0m[1;39;49m][0m[39;49m To update, run: [0m[32;49mpip install --upgrade pip[0m
</code></pre>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> os</span><br><span class="line"></span><br><span class="line"><span class="keyword">try</span>:</span><br><span class="line">    <span class="keyword">import</span> nucleotide_transformer</span><br><span class="line"><span class="keyword">except</span>:</span><br><span class="line">    !pip install git+https://github.com/instadeepai/nucleotide-transformer@main |tail -n <span class="number">1</span></span><br><span class="line">    <span class="keyword">import</span> nucleotide_transformer</span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> <span class="string">&quot;COLAB_TPU_ADDR&quot;</span> <span class="keyword">in</span> os.environ:</span><br><span class="line">    <span class="keyword">from</span> jax.tools <span class="keyword">import</span> colab_tpu</span><br><span class="line"></span><br><span class="line">    colab_tpu.setup_tpu()</span><br></pre></td></tr></table></figure>

<h2 id="2-2-SegmentNT"><a href="#2-2-SegmentNT" class="headerlink" title="2.2 SegmentNT"></a>2.2 SegmentNT</h2><p>âš ï¸ SegmentNT æ¨¡å‹é‡‡ç”¨ <a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41592-024-02523-z">æ ¸è‹·é…¸è½¬æ¢å™¨ï¼ˆNTï¼‰</a> ä½œä¸ºéª¨å¹²ç½‘ç»œï¼Œè®­ç»ƒæ•°æ®ä¸º 30,000 ä¸ªæ ¸è‹·é…¸åºåˆ—ï¼ˆå¯¹åº” 5001 ä¸ªæ ‡è®°ï¼ŒåŒ…å« CLS æ ‡è®°ï¼‰ã€‚ä½†ç ”ç©¶è¡¨æ˜ï¼ŒSegmentNT å¯æ³›åŒ–è‡³ 50,000 bpï¼ˆç¢±åŸºå¯¹ï¼‰çš„åºåˆ—ã€‚ç”±äº 30,000 bp çš„è®­ç»ƒé•¿åº¦è¶…è¿‡äº†æ ¸è‹·é…¸è½¬æ¢å™¨æ‰€èƒ½å¤„ç†çš„æœ€å¤§é•¿åº¦ï¼ˆ2048 ä¸ª 6-mer æ ‡è®°ï¼‰ï¼Œå› æ­¤é‡‡ç”¨äº† Yarn ç¼©æ”¾æŠ€æœ¯ã€‚</p>
<p>é»˜è®¤æƒ…å†µä¸‹ï¼Œ<code>ç¼©æ”¾ç³»æ•°ï¼ˆrescaling factorï¼‰</code> å·²è®¾ç½®ä¸ºè®­ç»ƒæ—¶ä½¿ç”¨çš„å€¼ã€‚è‹¥éœ€å¯¹ 30kbp è‡³ 50kbp ä¹‹é—´çš„åºåˆ—è¿›è¡Œæ¨ç†ï¼Œéœ€åœ¨ <code>get_pretrained_segment_nt_model</code> å‡½æ•°ä¸­ä¼ å…¥ <code>rescaling_factor</code> å‚æ•°ï¼Œå…¶å€¼è®¡ç®—å…¬å¼ä¸ºï¼š</p>
<p><strong>rescaling_factor &#x3D; æœ€å¤§æ ¸è‹·é…¸æ•° &#x2F; æ ¸è‹·é…¸è½¬æ¢å™¨æœ€å¤§æ ‡è®°æ•°</strong></p>
<p>å…¶ä¸­ï¼Œ<code>æ¨ç†æ—¶ DNA æ ‡è®°æ•°ï¼ˆnum_dna_tokens_inferenceï¼‰</code> æŒ‡æ¨ç†è¿‡ç¨‹ä¸­çš„æ ‡è®°æ€»æ•°ï¼ˆä¾‹å¦‚ 40008 ä¸ªç¢±åŸºå¯¹çš„åºåˆ—å¯¹åº” 6669 ä¸ªæ ‡è®°ï¼‰ï¼Œ<code>æ ¸è‹·é…¸è½¬æ¢å™¨æœ€å¤§æ ‡è®°æ•°ï¼ˆmax_num_tokens_ntï¼‰</code> ä¸ºéª¨å¹²ç½‘ç»œæ ¸è‹·é…¸è½¬æ¢å™¨çš„è®­ç»ƒæœ€å¤§æ ‡è®°æ•°ï¼Œå³ <code>2048</code>ã€‚</p>
<p>ğŸš§ <font color="#FF0000">SegmentNT æ¨¡å‹ä¸æ”¯æŒè¾“å…¥åºåˆ—ä¸­åŒ…å«ä»»ä½• â€œ<strong>N</strong>â€ ç¢±åŸºã€‚åŸå› æ˜¯æ¯ä¸ªæ ¸è‹·é…¸éœ€è¢«æ ‡è®°åŒ–ä¸º <code>6-mer</code> å½¢å¼ï¼Œè€ŒåŒ…å«ä¸€ä¸ªæˆ–å¤šä¸ª â€œNâ€ ç¢±åŸºçš„åºåˆ—æ— æ³•æ»¡è¶³è¿™ä¸€æ ‡è®°åŒ–è¦æ±‚ã€‚</font></p>
<h3 id="2-2-1-ğŸ”-ç¤ºä¾‹ä»£ç "><a href="#2-2-1-ğŸ”-ç¤ºä¾‹ä»£ç " class="headerlink" title="2.2.1 ğŸ” ç¤ºä¾‹ä»£ç "></a>2.2.1 ğŸ” <strong>ç¤ºä¾‹ä»£ç </strong></h3><p>ä¸‹é¢è¿™æ®µä»£ç çš„æ ¸å¿ƒåŠŸèƒ½æ˜¯ä½¿ç”¨é¢„è®­ç»ƒçš„<code>segment_nt</code>æ¨¡å‹å¯¹ DNA åºåˆ—è¿›è¡ŒåŸºå› ç»„ç‰¹å¾é¢„æµ‹ï¼ˆå¦‚é¢„æµ‹åºåˆ—æ˜¯å¦ä¸ºå†…å«å­ intron ç­‰ï¼‰ï¼Œå¹¶å±•ç¤ºäº†ä»æ¨¡å‹åŠ è½½ã€æ•°æ®é¢„å¤„ç†åˆ°å¹¶è¡Œæ¨ç†çš„å®Œæ•´æµç¨‹ã€‚å…·ä½“åŒ…æ‹¬ï¼š</p>
<ul>
<li><strong>ç¯å¢ƒé…ç½®</strong>ï¼šæŒ‡å®š JAX ä½¿ç”¨ CPU è®¾å¤‡ï¼Œé¿å…å†…å­˜æ³„æ¼ï¼ŒåŒæ—¶å‡†å¤‡å¤šè®¾å¤‡å¹¶è¡Œè®¡ç®—ç¯å¢ƒã€‚</li>
<li><strong>æ¨¡å‹åŠ è½½</strong>ï¼šåŠ è½½é¢„è®­ç»ƒçš„segment_ntæ¨¡å‹ï¼ŒåŒ…æ‹¬æ¨¡å‹å‚æ•°ã€å‰å‘è®¡ç®—å‡½æ•°ã€åˆ†è¯å™¨å’Œé…ç½®ä¿¡æ¯ï¼Œå¹¶é€šè¿‡ Haiku å’Œ JAX çš„å¹¶è¡Œæ¥å£ï¼ˆpmapï¼‰é€‚é…å¤šè®¾å¤‡è®¡ç®—ã€‚</li>
<li><strong>æ•°æ®å¤„ç†</strong>ï¼šå°†è¾“å…¥çš„ DNA åºåˆ—é€šè¿‡åˆ†è¯å™¨è½¬æ¢ä¸ºæ¨¡å‹å¯è¯†åˆ«çš„ token idsï¼Œå¹¶è½¬æ¢ä¸º JAX æ•°ç»„æ ¼å¼ã€‚</li>
<li><strong>å¹¶è¡Œæ¨ç†</strong>ï¼šåœ¨å¤šä¸ª CPU è®¾å¤‡ä¸Šå¹¶è¡Œè¿è¡Œæ¨¡å‹ï¼Œè¾“å‡º DNA åºåˆ—å¯¹åº”å„ç±»åŸºå› ç»„ç‰¹å¾çš„é¢„æµ‹æ¦‚ç‡ï¼Œå¹¶é‡ç‚¹æå–äº† â€œintronâ€ï¼ˆå†…å«å­ï¼‰çš„é¢„æµ‹æ¦‚ç‡ã€‚</li>
</ul>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># å¯¼å…¥æ‰€éœ€åº“ï¼šhaikuç”¨äºæ„å»ºç¥ç»ç½‘ç»œï¼Œjaxç”¨äºé«˜æ€§èƒ½æ•°å€¼è®¡ç®—ï¼Œjax.numpyæ˜¯jaxçš„numpyæ¥å£</span></span><br><span class="line"><span class="comment"># nucleotide_transformer.pretrainedä¸­çš„get_pretrained_segment_nt_modelç”¨äºè·å–é¢„è®­ç»ƒçš„æ ¸è‹·é…¸åºåˆ—æ¨¡å‹</span></span><br><span class="line"><span class="keyword">import</span> haiku <span class="keyword">as</span> hk</span><br><span class="line"><span class="keyword">import</span> jax</span><br><span class="line"><span class="keyword">import</span> jax.numpy <span class="keyword">as</span> jnp</span><br><span class="line"><span class="keyword">from</span> nucleotide_transformer.pretrained <span class="keyword">import</span> get_pretrained_segment_nt_model</span><br><span class="line"></span><br><span class="line"><span class="comment"># é…ç½®JAXé»˜è®¤ä½¿ç”¨CPUä½œä¸ºè®¡ç®—è®¾å¤‡ï¼Œé¿å…åœ¨å…¶ä»–è®¾å¤‡ï¼ˆå¦‚GPUï¼‰ä¸Šå¯èƒ½å‡ºç°çš„å†…å­˜æ³„æ¼é—®é¢˜ï¼Œå¢å¼ºä»£ç ç¨³å®šæ€§</span></span><br><span class="line">jax.config.update(<span class="string">&quot;jax_platform_name&quot;</span>, <span class="string">&quot;cpu&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># æŒ‡å®šè®¡ç®—åç«¯ä¸ºCPUï¼Œè·å–æ‰€æœ‰å¯ç”¨çš„CPUè®¾å¤‡ï¼Œå¹¶è®¡ç®—è®¾å¤‡æ•°é‡</span></span><br><span class="line">backend = <span class="string">&quot;cpu&quot;</span></span><br><span class="line">devices = jax.devices(backend)</span><br><span class="line">num_devices = <span class="built_in">len</span>(devices)</span><br><span class="line"><span class="built_in">print</span>(<span class="string">f&quot;Devices found: <span class="subst">&#123;devices&#125;</span>&quot;</span>)  <span class="comment"># æ‰“å°æ‰¾åˆ°çš„è®¾å¤‡ä¿¡æ¯</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># DNAåºåˆ—tokenæ•°é‡ï¼ˆä¸åŒ…å«å‰ç¼€çš„CLS tokenï¼‰éœ€è¦èƒ½è¢«ä¸‹é‡‡æ ·å—æ•°é‡çš„2çš„å¹‚æ•´é™¤ï¼ˆæ­¤å¤„ä¸‹é‡‡æ ·å—å¯¹åº”çš„å€¼ä¸º4ï¼‰</span></span><br><span class="line">max_num_nucleotides = <span class="number">8</span>  <span class="comment"># å®šä¹‰æœ€å¤§æ ¸è‹·é…¸æ•°é‡</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># æ–­è¨€æ£€æŸ¥ï¼šç¡®ä¿max_num_nucleotidesèƒ½è¢«4æ•´é™¤ï¼Œå¦åˆ™æŠ›å‡ºé”™è¯¯ï¼ˆä¿è¯æ¨¡å‹è¾“å…¥å°ºå¯¸å…¼å®¹ï¼‰</span></span><br><span class="line"><span class="keyword">assert</span> max_num_nucleotides % <span class="number">4</span> == <span class="number">0</span>, (</span><br><span class="line">    <span class="string">&quot;The number of DNA tokens (excluding the CLS token prepended) needs to be dividible by&quot;</span></span><br><span class="line">     <span class="string">&quot;2 to the power of the number of downsampling block, i.e 4.&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># è·å–é¢„è®­ç»ƒçš„segment_ntæ¨¡å‹ï¼šåŒ…æ‹¬æ¨¡å‹å‚æ•°ã€å‰å‘è®¡ç®—å‡½æ•°ã€åˆ†è¯å™¨å’Œæ¨¡å‹é…ç½®</span></span><br><span class="line"><span class="comment"># å…¶ä¸­ï¼š</span></span><br><span class="line"><span class="comment"># - model_nameæŒ‡å®šæ¨¡å‹åä¸º&quot;segment_nt&quot;</span></span><br><span class="line"><span class="comment"># - embeddings_layers_to_saveæŒ‡å®šä¿å­˜ç¬¬29å±‚çš„åµŒå…¥ç‰¹å¾</span></span><br><span class="line"><span class="comment"># - attention_maps_to_saveæŒ‡å®šä¿å­˜(1,4)å’Œ(7,10)ä½ç½®çš„æ³¨æ„åŠ›å›¾</span></span><br><span class="line"><span class="comment"># - max_positionsæŒ‡å®šæœ€å¤§ä½ç½®æ•°ï¼ˆåŒ…å«CLS tokenï¼Œå› æ­¤ä¸ºmax_num_nucleotides + 1ï¼‰</span></span><br><span class="line">parameters, forward_fn, tokenizer, config = get_pretrained_segment_nt_model(</span><br><span class="line">    model_name=<span class="string">&quot;segment_nt&quot;</span>,</span><br><span class="line">    embeddings_layers_to_save=(<span class="number">29</span>,),</span><br><span class="line">    attention_maps_to_save=((<span class="number">1</span>, <span class="number">4</span>), (<span class="number">7</span>, <span class="number">10</span>)),</span><br><span class="line">    max_positions=max_num_nucleotides + <span class="number">1</span>,</span><br><span class="line">)</span><br><span class="line"><span class="comment"># ç”¨haiku.transformè½¬æ¢å‰å‘å‡½æ•°ï¼Œä½¿å…¶ç¬¦åˆhaikuçš„å‡½æ•°å¼ç¼–ç¨‹èŒƒå¼</span></span><br><span class="line">forward_fn = hk.transform(forward_fn)</span><br><span class="line"><span class="comment"># ä½¿ç”¨jax.pmapå¯¹å‰å‘è®¡ç®—å‡½æ•°è¿›è¡Œå¹¶è¡Œæ˜ å°„ï¼ŒæŒ‡å®šè®¡ç®—è®¾å¤‡ï¼Œdonate_argnums=(0,)è¡¨ç¤ºå…è®¸é‡Šæ”¾å‚æ•°çš„å†…å­˜</span></span><br><span class="line">apply_fn = jax.pmap(forward_fn.apply, devices=devices, donate_argnums=(<span class="number">0</span>,))</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment"># å‡†å¤‡è¾“å…¥æ•°æ®å¹¶è¿›è¡Œåˆ†è¯</span></span><br><span class="line">sequences = [<span class="string">&quot;ATTCCGATTCCGATTCCAACGGATTATTCCGATTAACCGATTCCAATT&quot;</span>, <span class="string">&quot;ATTTCTCTCTCTCTCTGAGATCGATGATTTCTCTCTCATCGAACTATG&quot;</span>]  <span class="comment"># ä¸¤ä¸ªDNAåºåˆ—ç¤ºä¾‹</span></span><br><span class="line"><span class="comment"># å¯¹åºåˆ—è¿›è¡Œæ‰¹é‡åˆ†è¯ï¼Œæå–token idsï¼ˆå¿½ç•¥å…ƒç»„ä¸­çš„ç¬¬ä¸€ä¸ªå…ƒç´ ï¼Œä¿ç•™token idsï¼‰</span></span><br><span class="line">tokens_ids = [b[<span class="number">1</span>] <span class="keyword">for</span> b <span class="keyword">in</span> tokenizer.batch_tokenize(sequences)]</span><br><span class="line"><span class="comment"># å°†token idsè½¬æ¢ä¸ºjaxæ•°ç»„ï¼ˆint32ç±»å‹ï¼‰ï¼Œä½œä¸ºæ¨¡å‹è¾“å…¥</span></span><br><span class="line">tokens = jnp.asarray(tokens_ids, dtype=jnp.int32)</span><br><span class="line"></span><br><span class="line"><span class="comment"># åˆå§‹åŒ–éšæœºç§å­ï¼Œå¹¶å°†éšæœºé”®å¤åˆ¶åˆ°æ‰€æœ‰è®¾å¤‡ï¼ˆå¤šè®¾å¤‡å¹¶è¡Œæ—¶ä¿æŒä¸€è‡´æ€§ï¼‰</span></span><br><span class="line">random_key = jax.random.PRNGKey(seed=<span class="number">0</span>)</span><br><span class="line">keys = jax.device_put_replicated(random_key, devices=devices)</span><br><span class="line"><span class="comment"># å°†æ¨¡å‹å‚æ•°å¤åˆ¶åˆ°æ‰€æœ‰è®¾å¤‡</span></span><br><span class="line">parameters = jax.device_put_replicated(parameters, devices=devices)</span><br><span class="line"><span class="comment"># å°†è¾“å…¥tokenå¤åˆ¶åˆ°æ‰€æœ‰è®¾å¤‡</span></span><br><span class="line">tokens = jax.device_put_replicated(tokens, devices=devices)</span><br><span class="line"></span><br><span class="line"><span class="comment"># å¯¹è¾“å…¥åºåˆ—è¿›è¡Œæ¨¡å‹æ¨ç†</span></span><br><span class="line">outs = apply_fn(parameters, keys, tokens)</span><br><span class="line"><span class="comment"># ä»æ¨ç†ç»“æœä¸­è·å–åŸºå› ç»„ç‰¹å¾çš„logitsï¼ˆæœªå½’ä¸€åŒ–çš„æ¦‚ç‡ï¼‰</span></span><br><span class="line">logits = outs[<span class="string">&quot;logits&quot;</span>]</span><br><span class="line"><span class="comment"># å°†logitsé€šè¿‡softmaxè½¬æ¢ä¸ºæ¦‚ç‡ï¼Œå¹¶å–æœ€åä¸€ä¸ªç»´åº¦çš„ç»“æœ</span></span><br><span class="line">probabilities = jnp.asarray(jax.nn.softmax(logits, axis=-<span class="number">1</span>))[...,-<span class="number">1</span>]</span><br><span class="line"><span class="built_in">print</span>(<span class="string">f&quot;Probabilities shape: <span class="subst">&#123;probabilities.shape&#125;</span>&quot;</span>)  <span class="comment"># æ‰“å°æ¦‚ç‡çš„å½¢çŠ¶</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># æ‰“å°æ¨¡å‹å¯é¢„æµ‹çš„åŸºå› ç»„ç‰¹å¾ï¼ˆå¦‚intronã€exonç­‰ï¼‰</span></span><br><span class="line"><span class="built_in">print</span>(<span class="string">f&quot;Features inferred: <span class="subst">&#123;config.features&#125;</span>&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># è·å–&quot;intron&quot;ï¼ˆå†…å«å­ï¼‰ç‰¹å¾åœ¨ç‰¹å¾åˆ—è¡¨ä¸­çš„ç´¢å¼•</span></span><br><span class="line">idx_intron = config.features.index(<span class="string">&quot;intron&quot;</span>)</span><br><span class="line"><span class="comment"># æå–&quot;intron&quot;å¯¹åº”çš„æ¦‚ç‡</span></span><br><span class="line">probabilities_intron = probabilities[..., idx_intron]</span><br><span class="line"><span class="built_in">print</span>(<span class="string">f&quot;Intron probabilities shape: <span class="subst">&#123;probabilities_intron.shape&#125;</span>&quot;</span>)  <span class="comment"># æ‰“å°å†…å«å­æ¦‚ç‡çš„å½¢çŠ¶</span></span><br></pre></td></tr></table></figure>

<p>æ”¯æŒçš„æ¨¡å‹å:</p>
<ul>
<li><strong>segment_nt</strong></li>
<li><strong>segment_nt_multi_species</strong></li>
</ul>
<h3 id="2-2-2-å®Œæ•´ä»£ç "><a href="#2-2-2-å®Œæ•´ä»£ç " class="headerlink" title="2.2.2 å®Œæ•´ä»£ç "></a>2.2.2 å®Œæ•´ä»£ç </h3><p>ä¸‹é¢çš„ä»£ç å±•ç¤ºäº†å¦‚ä½•å¯¹ 10kb å’Œ 50kb åºåˆ—è¿›è¡Œæ¨ç†ï¼Œä»¥åŠå¦‚ä½•ç»˜åˆ¶æ¦‚ç‡åˆ†å¸ƒå›¾ä»¥å¤ç°è®ºæ–‡ä¸­çš„å›¾ 1ï¼ˆFig.1dï¼‰ã€‚</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> Bio <span class="keyword">import</span> SeqIO</span><br><span class="line"><span class="keyword">import</span> gzip</span><br><span class="line"><span class="keyword">import</span> haiku <span class="keyword">as</span> hk</span><br><span class="line"><span class="keyword">import</span> jax</span><br><span class="line"><span class="keyword">import</span> jax.numpy <span class="keyword">as</span> jnp</span><br><span class="line"><span class="keyword">import</span> numpy <span class="keyword">as</span> np</span><br><span class="line"><span class="keyword">import</span> seaborn <span class="keyword">as</span> sns</span><br><span class="line"><span class="keyword">from</span> typing <span class="keyword">import</span> <span class="type">List</span></span><br><span class="line"><span class="keyword">import</span> matplotlib.pyplot <span class="keyword">as</span> plt</span><br><span class="line"><span class="keyword">from</span> nucleotide_transformer.pretrained <span class="keyword">import</span> get_pretrained_segment_nt_model</span><br><span class="line"></span><br><span class="line"><span class="comment"># Specify &quot;cpu&quot; as default (but you can decide to use GPU or TPU in the next cell)</span></span><br><span class="line">jax.config.update(<span class="string">&quot;jax_platform_name&quot;</span>, <span class="string">&quot;cpu&quot;</span>)</span><br></pre></td></tr></table></figure>

<pre><code>Devices found: [CpuDevice(id=0)]
</code></pre>
<p><strong>ï¼ˆ1ï¼‰æŒ‡å®šåç«¯è®¾å¤‡</strong></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Use either &quot;cpu&quot;, &quot;gpu&quot; or &quot;tpu&quot;</span></span><br><span class="line">backend = <span class="string">&quot;cpu&quot;</span></span><br><span class="line">devices = jax.devices(backend)</span><br><span class="line">num_devices = <span class="built_in">len</span>(devices)</span><br><span class="line"><span class="built_in">print</span>(<span class="string">f&quot;Devices found: <span class="subst">&#123;devices&#125;</span>&quot;</span>)</span><br></pre></td></tr></table></figure>

<p><strong>ï¼ˆ2ï¼‰å®šä¹‰ç”¨äºç»˜åˆ¶æ¦‚ç‡çš„å‡½æ•°</strong></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># seaborn settings</span></span><br><span class="line">sns.set_style(<span class="string">&quot;whitegrid&quot;</span>)</span><br><span class="line">sns.set_context(</span><br><span class="line">    <span class="string">&quot;notebook&quot;</span>,</span><br><span class="line">    font_scale=<span class="number">1</span>,</span><br><span class="line">    rc=&#123;</span><br><span class="line">        <span class="string">&quot;font.size&quot;</span>: <span class="number">14</span>,</span><br><span class="line">        <span class="string">&quot;axes.titlesize&quot;</span>: <span class="number">18</span>,</span><br><span class="line">        <span class="string">&quot;axes.labelsize&quot;</span>: <span class="number">18</span>,</span><br><span class="line">        <span class="string">&quot;xtick.labelsize&quot;</span>: <span class="number">16</span>,</span><br><span class="line">        <span class="string">&quot;ytick.labelsize&quot;</span>: <span class="number">16</span>,</span><br><span class="line">        <span class="string">&quot;legend.fontsize&quot;</span>: <span class="number">16</span>,</span><br><span class="line">        &#125;</span><br><span class="line">)</span><br><span class="line"></span><br><span class="line">plt.rcParams[<span class="string">&#x27;xtick.bottom&#x27;</span>] = <span class="literal">True</span></span><br><span class="line">plt.rcParams[<span class="string">&#x27;ytick.left&#x27;</span>] = <span class="literal">True</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># set colors</span></span><br><span class="line">colors = sns.color_palette(<span class="string">&quot;Set2&quot;</span>).as_hex()</span><br><span class="line">colors2 = sns.color_palette(<span class="string">&quot;husl&quot;</span>).as_hex()</span><br><span class="line"></span><br><span class="line"><span class="comment"># Rearrange order of the features to match Fig.3 from the paper.</span></span><br><span class="line">features_rearranged = [</span><br><span class="line"> <span class="string">&#x27;protein_coding_gene&#x27;</span>,</span><br><span class="line"> <span class="string">&#x27;lncRNA&#x27;</span>,</span><br><span class="line"> <span class="string">&#x27;5UTR&#x27;</span>,</span><br><span class="line"> <span class="string">&#x27;3UTR&#x27;</span>,</span><br><span class="line"> <span class="string">&#x27;exon&#x27;</span>,</span><br><span class="line"> <span class="string">&#x27;intron&#x27;</span>,</span><br><span class="line"> <span class="string">&#x27;splice_donor&#x27;</span>,</span><br><span class="line"> <span class="string">&#x27;splice_acceptor&#x27;</span>,</span><br><span class="line"> <span class="string">&#x27;promoter_Tissue_specific&#x27;</span>,</span><br><span class="line"> <span class="string">&#x27;promoter_Tissue_invariant&#x27;</span>,</span><br><span class="line"> <span class="string">&#x27;enhancer_Tissue_specific&#x27;</span>,</span><br><span class="line"> <span class="string">&#x27;enhancer_Tissue_invariant&#x27;</span>,</span><br><span class="line"> <span class="string">&#x27;CTCF-bound&#x27;</span>,</span><br><span class="line"> <span class="string">&#x27;polyA_signal&#x27;</span>,</span><br><span class="line">]</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">plot_features</span>(<span class="params"></span></span><br><span class="line"><span class="params">    predicted_probabilities_all,</span></span><br><span class="line"><span class="params">    seq_length: <span class="built_in">int</span>,</span></span><br><span class="line"><span class="params">    features: <span class="type">List</span>[<span class="built_in">str</span>],</span></span><br><span class="line"><span class="params">    order_to_plot: <span class="type">List</span>[<span class="built_in">str</span>],</span></span><br><span class="line"><span class="params">    fig_width=<span class="number">8</span>,</span></span><br><span class="line"><span class="params"></span>):</span><br><span class="line">    <span class="string">&quot;&quot;&quot;</span></span><br><span class="line"><span class="string">    Function to plot labels and predicted probabilities.</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">    Args:</span></span><br><span class="line"><span class="string">        predicted_probabilities_all: Probabilities per genomic feature for each</span></span><br><span class="line"><span class="string">            nucleotides in the DNA sequence.</span></span><br><span class="line"><span class="string">        seq_length: DNA sequence length.</span></span><br><span class="line"><span class="string">        feature: Genomic features to plot.</span></span><br><span class="line"><span class="string">        order_to_plot: Order in which to plot the genomic features. This needs to be</span></span><br><span class="line"><span class="string">            specified in order to match the order presented in the Fig.3 of the paper</span></span><br><span class="line"><span class="string">        fig_width: Width of the figure</span></span><br><span class="line"><span class="string">    &quot;&quot;&quot;</span></span><br><span class="line"></span><br><span class="line">    sc = <span class="number">1.8</span></span><br><span class="line">    n_panels = <span class="number">7</span></span><br><span class="line"></span><br><span class="line">    <span class="comment"># fig, axes = plt.subplots(n_panels, 1, figsize=(fig_width * sc, (n_panels + 2) * sc), height_ratios=[6] + [2] * (n_panels-1))</span></span><br><span class="line">    _, axes = plt.subplots(n_panels, <span class="number">1</span>, figsize=(fig_width * sc, (n_panels + <span class="number">4</span>) * sc))</span><br><span class="line"></span><br><span class="line">    <span class="keyword">for</span> n, feat <span class="keyword">in</span> <span class="built_in">enumerate</span>(order_to_plot):</span><br><span class="line">        feat_id = features.index(feat)</span><br><span class="line">        prob_dist = predicted_probabilities_all[:, feat_id]</span><br><span class="line"></span><br><span class="line">        <span class="comment"># Use the appropriate subplot</span></span><br><span class="line">        ax = axes[n // <span class="number">2</span>]</span><br><span class="line"></span><br><span class="line">        <span class="keyword">try</span>:</span><br><span class="line">            id_color = colors[feat_id]</span><br><span class="line">        <span class="keyword">except</span>:</span><br><span class="line">            id_color = colors2[feat_id - <span class="number">8</span>]</span><br><span class="line">        ax.plot(</span><br><span class="line">            prob_dist,</span><br><span class="line">            color=id_color,</span><br><span class="line">            label=feat,</span><br><span class="line">            linestyle=<span class="string">&quot;-&quot;</span>,</span><br><span class="line">            linewidth=<span class="number">1.5</span>,</span><br><span class="line">        )</span><br><span class="line">        ax.set_xlim(<span class="number">0</span>, seq_length)</span><br><span class="line">        ax.grid(<span class="literal">False</span>)</span><br><span class="line">        ax.spines[<span class="string">&#x27;bottom&#x27;</span>].set_color(<span class="string">&#x27;black&#x27;</span>)</span><br><span class="line">        ax.spines[<span class="string">&#x27;top&#x27;</span>].set_color(<span class="string">&#x27;black&#x27;</span>)</span><br><span class="line">        ax.spines[<span class="string">&#x27;right&#x27;</span>].set_color(<span class="string">&#x27;black&#x27;</span>)</span><br><span class="line">        ax.spines[<span class="string">&#x27;left&#x27;</span>].set_color(<span class="string">&#x27;black&#x27;</span>)</span><br><span class="line"></span><br><span class="line">    <span class="keyword">for</span> a <span class="keyword">in</span> <span class="built_in">range</span> (<span class="number">0</span>,n_panels):</span><br><span class="line">        axes[a].set_ylim(<span class="number">0</span>, <span class="number">1.05</span>)</span><br><span class="line">        axes[a].set_ylabel(<span class="string">&quot;Prob.&quot;</span>)</span><br><span class="line">        axes[a].legend(loc=<span class="string">&quot;upper left&quot;</span>, bbox_to_anchor=(<span class="number">1</span>, <span class="number">1</span>), borderaxespad=<span class="number">0</span>)</span><br><span class="line">        <span class="keyword">if</span> a != (n_panels-<span class="number">1</span>):</span><br><span class="line">            axes[a].tick_params(axis=<span class="string">&#x27;x&#x27;</span>, which=<span class="string">&#x27;both&#x27;</span>, bottom=<span class="literal">True</span>, top=<span class="literal">False</span>, labelbottom=<span class="literal">False</span>)</span><br><span class="line"></span><br><span class="line">    <span class="comment"># Set common x-axis label</span></span><br><span class="line">    axes[-<span class="number">1</span>].set_xlabel(<span class="string">&quot;Nucleotides&quot;</span>)</span><br><span class="line">    <span class="comment"># axes[0].axis(&#x27;off&#x27;)  # Turn off the axis</span></span><br><span class="line">    axes[n_panels-<span class="number">1</span>].grid(<span class="literal">False</span>)</span><br><span class="line">    axes[n_panels-<span class="number">1</span>].tick_params(axis=<span class="string">&#x27;y&#x27;</span>, which=<span class="string">&#x27;both&#x27;</span>, left=<span class="literal">True</span>, right=<span class="literal">False</span>, labelleft=<span class="literal">True</span>, labelright=<span class="literal">False</span>)</span><br><span class="line"></span><br><span class="line">    axes[<span class="number">0</span>].set_title(<span class="string">&quot;Probabilities predicted over all genomics features&quot;</span>, fontweight=<span class="string">&quot;bold&quot;</span>)</span><br><span class="line"></span><br><span class="line">    plt.show()</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p><strong>ï¼ˆ3ï¼‰è·å–äººçš„ 20å·æŸ“è‰²ä½“åºåˆ—</strong></p>
<p>ä¸ºäº†é‡ç°Segment-NTè®ºæ–‡ä¸­çš„å›¾è¡¨ï¼Œæˆ‘ä»¬åœ¨æ­¤æ£€ç´¢ä¸‹è½½äººç±»20å·æŸ“è‰²ä½“çš„æ–‡ä»¶</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">! wget https://ftp.ensembl.org/pub/release-<span class="number">111</span>/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.chromosome<span class="number">.20</span>.fa.gz</span><br></pre></td></tr></table></figure>

<pre><code>--2024-07-23 09:38:24--  https://ftp.ensembl.org/pub/release-111/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.chromosome.20.fa.gz
Resolving ftp.ensembl.org (ftp.ensembl.org)... 193.62.193.169
Connecting to ftp.ensembl.org (ftp.ensembl.org)|193.62.193.169|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18833053 (18M) [application/x-gzip]
Saving to: â€˜Homo_sapiens.GRCh38.dna.chromosome.20.fa.gzâ€™

Homo_sapiens.GRCh38 100%[===================&gt;]  17,96M  1,77MB/s    in 10s     

2024-07-23 09:38:35 (1,75 MB/s) - â€˜Homo_sapiens.GRCh38.dna.chromosome.20.fa.gzâ€™ saved [18833053/18833053]
</code></pre>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">fasta_path = <span class="string">&quot;Homo_sapiens.GRCh38.dna.chromosome.20.fa.gz&quot;</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">with</span> gzip.<span class="built_in">open</span>(fasta_path, <span class="string">&quot;rt&quot;</span>) <span class="keyword">as</span> handle:</span><br><span class="line">    record = <span class="built_in">next</span>(SeqIO.parse(handle, <span class="string">&quot;fasta&quot;</span>))</span><br><span class="line">    chr20 = <span class="built_in">str</span>(record.seq)</span><br></pre></td></tr></table></figure>

<p><strong>ï¼ˆ4ï¼‰å¯¹10kbåŸºå› ç»„åºåˆ—è¿›è¡Œæ¨æ–­</strong>ï¼ˆä¸éœ€è¦æ”¹å˜å‰å‘å‡½æ•°ä¸­çš„é‡æ–°ç¼©æ”¾å› å­ï¼‰</p>
<p><strong>â‘  å®ä¾‹åŒ–SegmentNTæ¨ç†å‡½æ•°</strong></p>
<p>ä»¥ä¸‹ä»£ç å…è®¸æ‚¨ä¸‹è½½å…¶ä¸­ä¸€ä¸ªSegment-NTæ¨¡å‹çš„æƒé‡ã€‚å®ƒä¼šè¿”å›æƒé‡å­—å…¸ã€haikuå‰å‘å‡½æ•°ã€åˆ†è¯å™¨å’Œé…ç½®å­—å…¸ã€‚</p>
<p>ä¸<code>get_pretrained_nucleotide_transformer</code>å‡½æ•°ç±»ä¼¼ï¼Œæ‚¨è¿˜å¯ä»¥æŒ‡å®šï¼š</p>
<ol>
<li>æ‚¨å¸Œæœ›æ”¶é›†åµŒå…¥çš„å±‚ï¼ˆä¾‹å¦‚ï¼Œ(5, 10, 20) è¡¨ç¤ºè·å–ç¬¬5ã€10å’Œ20å±‚çš„åµŒå…¥ï¼‰</li>
<li>æ‚¨å¸Œæœ›æ”¶é›†çš„æ³¨æ„åŠ›å›¾ï¼ˆä¾‹å¦‚ï¼Œ((1,4), (7,18)) è¡¨ç¤ºè·å–å¯¹åº”äºç¬¬1å±‚ç¬¬4å¤´å’Œç¬¬7å±‚ç¬¬18å¤´çš„æ³¨æ„åŠ›å›¾ï¼‰ã€‚è¯·å‚è€ƒé…ç½®ä»¥æŸ¥çœ‹æ¨¡å‹ä¸­çš„å±‚æ•°å’Œå¤´æ•°ã€‚</li>
<li>æ‚¨å°†è¿›è¡Œæ¨ç†è®¡ç®—çš„åºåˆ—ä¸­çš„æœ€å¤§æ ‡è®°æ•°ã€‚æ‚¨å¯ä»¥è¾“å…¥ä¸è¶…è¿‡æ¨¡å‹é…ç½®ä¸­æŒ‡å®šçš„å€¼ï¼ˆåŒ…å«å°†è‡ªåŠ¨æ·»åŠ åˆ°åºåˆ—å¼€å¤´çš„ç±»æ ‡è®°ï¼‰ï¼Œä½†ä¸ºäº†ä¼˜åŒ–å†…å­˜å’Œæ¨ç†æ—¶é—´ï¼Œæˆ‘ä»¬å»ºè®®å°†æ­¤æ•°å­—å°½å¯èƒ½è®¾å°ã€‚</li>
</ol>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># The number of DNA tokens (excluding the CLS token prepended) needs to be dividible by</span></span><br><span class="line"><span class="comment"># the square of the number of downsampling block, i.e 4.</span></span><br><span class="line">max_num_nucleotides = <span class="number">1668</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">assert</span> max_num_nucleotides % <span class="number">4</span> == <span class="number">0</span>, (</span><br><span class="line">    <span class="string">&quot;The number of DNA tokens (excluding the CLS token prepended) needs to be dividible by&quot;</span></span><br><span class="line">     <span class="string">&quot;2 to the power of the number of downsampling block, i.e 4.&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># If max_num_nucleotides is larger than what was used to train Segment-NT, the rescaling</span></span><br><span class="line"><span class="comment"># factor needs to be adapted.</span></span><br><span class="line"><span class="keyword">if</span> max_num_nucleotides + <span class="number">1</span> &gt; <span class="number">5001</span>:</span><br><span class="line">    inference_rescaling_factor = (max_num_nucleotides + <span class="number">1</span>) / <span class="number">2048</span></span><br><span class="line"><span class="keyword">else</span>:</span><br><span class="line">    inference_rescaling_factor=<span class="literal">None</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># If this download fails at one point, restarting it will not work.</span></span><br><span class="line"><span class="comment"># Before rerunning the cell, make sure to delete the cache by executing:</span></span><br><span class="line"><span class="comment"># ! rm -rf ~/.cache/nucleotide_transformer/</span></span><br><span class="line">parameters, forward_fn, tokenizer, config = get_pretrained_segment_nt_model(</span><br><span class="line">    model_name=<span class="string">&quot;segment_nt&quot;</span>,</span><br><span class="line">    rescaling_factor=inference_rescaling_factor,</span><br><span class="line">    embeddings_layers_to_save=(<span class="number">29</span>,),</span><br><span class="line">    attention_maps_to_save=((<span class="number">1</span>, <span class="number">4</span>), (<span class="number">7</span>, <span class="number">10</span>)),</span><br><span class="line">    max_positions=max_num_nucleotides + <span class="number">1</span>,</span><br><span class="line">)</span><br><span class="line">forward_fn = hk.transform(forward_fn)</span><br><span class="line">apply_fn = jax.pmap(forward_fn.apply, devices=devices, donate_argnums=(<span class="number">0</span>,))</span><br><span class="line"></span><br><span class="line"><span class="comment"># Put required quantities for the inference on the devices. This step is not</span></span><br><span class="line"><span class="comment"># reproduced in the second inference since the quantities will already be loaded</span></span><br><span class="line"><span class="comment"># on the devices !</span></span><br><span class="line">random_key = jax.random.PRNGKey(seed=<span class="number">0</span>)</span><br><span class="line">keys = jax.device_put_replicated(random_key, devices=devices)</span><br><span class="line">parameters = jax.device_put_replicated(parameters, devices=devices)</span><br></pre></td></tr></table></figure>

<pre><code>Downloading model&#39;s hyperparameters json file...
Downloaded model&#39;s hyperparameters.
Downloading model&#39;s weights...
Downloaded model&#39;s weights...
</code></pre>
<p><strong>â‘¡ å¯¹DNAåºåˆ—è¿›è¡Œåˆ†è¯</strong></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">idx_start = <span class="number">2650520</span></span><br><span class="line"></span><br><span class="line">idx_stop = idx_start + max_num_nucleotides*<span class="number">6</span></span><br><span class="line"></span><br><span class="line">sequences = [chr20[idx_start:idx_stop]]</span><br><span class="line">tokens_ids = [b[<span class="number">1</span>] <span class="keyword">for</span> b <span class="keyword">in</span> tokenizer.batch_tokenize(sequences)]</span><br><span class="line">tokens_str = [b[<span class="number">0</span>] <span class="keyword">for</span> b <span class="keyword">in</span> tokenizer.batch_tokenize(sequences)]</span><br><span class="line"></span><br><span class="line"><span class="comment"># This stacks the batch so that it is repeated across the devices. This is done</span></span><br><span class="line"><span class="comment"># in order to allow for replication even if one has more than one device.</span></span><br><span class="line"><span class="comment"># To take advantage of the multiple devices and infer different sequences on</span></span><br><span class="line"><span class="comment"># each of the devices, make sure to change this line into a reshape.</span></span><br><span class="line"><span class="comment"># a reshape</span></span><br><span class="line">tokens = jnp.stack([jnp.asarray(tokens_ids, dtype=jnp.int32)]*num_devices, axis=<span class="number">0</span>)</span><br><span class="line">tokens.shape</span><br><span class="line"></span><br></pre></td></tr></table></figure>




<pre><code>(1, 1, 1669)
</code></pre>
<p><strong>â‘¢ å¯¹ç”Ÿæˆçš„æ‰¹æ¬¡è¿›è¡Œæ¨ç†</strong></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Infer</span></span><br><span class="line">outs = apply_fn(parameters, keys, tokens)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Obtain the logits over the genomic features</span></span><br><span class="line">logits = outs[<span class="string">&quot;logits&quot;</span>]</span><br><span class="line"><span class="comment"># Transform them on probabilities</span></span><br><span class="line">probabilities = np.asarray(jax.nn.softmax(logits, axis=-<span class="number">1</span>))[...,-<span class="number">1</span>]</span><br></pre></td></tr></table></figure>

<pre><code>/home/hugo/anaconda3/envs/trix/lib/python3.10/site-packages/jax/interpreters/mlir.py:622: UserWarning: Some donated buffers were not usable: ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,8192]), ShapedArray(float32[4096,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[4107,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[4107]), ShapedArray(float32[1024,4107]), ShapedArray(float32[1024]), ShapedArray(float32[1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[1024]), ShapedArray(float32[168]), ShapedArray(float32[1024,168]), ShapedArray(float32[1024]), ShapedArray(float32[3,1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[3,1024,1024]), ShapedArray(float32[2048]), ShapedArray(float32[3,1024,2048]), ShapedArray(float32[2048]), ShapedArray(float32[3,2048,2048]), ShapedArray(float32[1024]), ShapedArray(float32[3,1024,1024]), ShapedArray(float32[1024]), ShapedArray(float32[3,1024,1024]), ShapedArray(float32[2048]), ShapedArray(float32[3,2048,2048]), ShapedArray(float32[2048]), ShapedArray(float32[3,2048,2048]), ShapedArray(float32[1024]), ShapedArray(float32[3,1024,2048]), ShapedArray(float32[1024]), ShapedArray(float32[3,1024,1024]).
See an explanation at https://jax.readthedocs.io/en/latest/faq.html#buffer-donation.
  warnings.warn(f&quot;Some donated buffers were not usable: &#123;&#39;, &#39;.join(unused_donations)&#125;.\n&#123;msg&#125;&quot;)
</code></pre>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">probabilities.shape</span><br></pre></td></tr></table></figure>




<pre><code>(1, 1, 10008, 14)
</code></pre>
<p><strong>â‘£ ç»˜åˆ¶è¿™æ¡DNAåºåˆ—ä¸Š14ä¸ªåŸºå› ç»„ç‰¹å¾çš„æ¦‚ç‡å›¾</strong></p>
<p>è¯·æ³¨æ„ï¼ŒSegmentNTè®ºæ–‡ä¸­çš„å›¾1æ˜¯ç”¨SegmentNT-10kbå®ç°çš„ï¼Œè€Œè¿™é‡Œä½¿ç”¨çš„æ˜¯SegmentNT-30kbï¼Œè¿™å°±è§£é‡Šäº†ä¸ºä»€ä¹ˆæ¦‚ç‡å¹¶éå®Œå…¨ç›¸åŒã€‚</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">plot_features(</span><br><span class="line">    probabilities[<span class="number">0</span>,<span class="number">0</span>],</span><br><span class="line">    probabilities.shape[-<span class="number">2</span>],</span><br><span class="line">    fig_width=<span class="number">20</span>,</span><br><span class="line">    features=config.features,</span><br><span class="line">    order_to_plot=features_rearranged</span><br><span class="line">)</span><br></pre></td></tr></table></figure>
<p><img src="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/images/post/BioAI/03/Segmentation_models%E5%AE%9E%E6%88%98_26_0.png" class="lazyload placeholder" data-srcset="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/images/post/BioAI/03/Segmentation_models%E5%AE%9E%E6%88%98_26_0.png" srcset="https://pic1.zhimg.com/v2-cd38920285d125be80b3eb504052c550_b.webp" alt="png"></p>
<p><strong>ï¼ˆ5ï¼‰å¯¹50kbçš„åŸºå› ç»„åºåˆ—è¿›è¡Œæ¨æ–­å¹¶ç»˜å›¾</strong>ï¼ˆéœ€è¦æ›´æ”¹å‰å‘å‡½æ•°ä¸­çš„é‡æ–°ç¼©æ”¾å› å­ï¼‰</p>
<p><strong>â‘  å®ä¾‹åŒ–SegmentNTæ¨ç†å‡½æ•°</strong></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># The number of DNA tokens (excluding the CLS token prepended) needs to be dividible by</span></span><br><span class="line"><span class="comment"># the square of the number of downsampling block, i.e 4.</span></span><br><span class="line">max_num_nucleotides = <span class="number">8332</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">assert</span> max_num_nucleotides % <span class="number">4</span> == <span class="number">0</span>, (</span><br><span class="line">    <span class="string">&quot;The number of DNA tokens (excluding the CLS token prepended) needs to be dividible by&quot;</span></span><br><span class="line">     <span class="string">&quot;2 to the power of the number of downsampling block, i.e 4.&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># If max_num_nucleotides is larger than what was used to train Segment-NT, the rescaling</span></span><br><span class="line"><span class="comment"># factor needs to be adapted.</span></span><br><span class="line"><span class="keyword">if</span> max_num_nucleotides + <span class="number">1</span> &gt; <span class="number">5001</span>:</span><br><span class="line">    inference_rescaling_factor = (max_num_nucleotides + <span class="number">1</span>) / <span class="number">2048</span></span><br><span class="line"><span class="keyword">else</span>:</span><br><span class="line">    inference_rescaling_factor=<span class="literal">None</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># The parameters have already been downloaded above</span></span><br><span class="line"><span class="comment"># so we do not instantiate them. However we instantiate a new forward function</span></span><br><span class="line"><span class="comment"># where the context length extension needed is in effect.</span></span><br><span class="line">_, forward_fn, tokenizer, config = get_pretrained_segment_nt_model(</span><br><span class="line">    model_name=<span class="string">&quot;segment_nt&quot;</span>,</span><br><span class="line">    rescaling_factor=inference_rescaling_factor,</span><br><span class="line">    embeddings_layers_to_save=(<span class="number">29</span>,),</span><br><span class="line">    attention_maps_to_save=((<span class="number">1</span>, <span class="number">4</span>), (<span class="number">7</span>, <span class="number">10</span>)),</span><br><span class="line">    max_positions=max_num_nucleotides + <span class="number">1</span>,</span><br><span class="line">)</span><br><span class="line">forward_fn = hk.transform(forward_fn)</span><br><span class="line">apply_fn = jax.pmap(forward_fn.apply, devices=devices, donate_argnums=(<span class="number">0</span>,))</span><br></pre></td></tr></table></figure>

<p><strong>â‘¡ å¯¹DNAåºåˆ—è¿›è¡Œåˆ†è¯</strong></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">idx_start = <span class="number">5099984</span></span><br><span class="line">idx_stop = idx_start + max_num_nucleotides*<span class="number">6</span></span><br><span class="line"></span><br><span class="line">sequences = [chr20[idx_start:idx_stop]]</span><br><span class="line">tokens_ids = [b[<span class="number">1</span>] <span class="keyword">for</span> b <span class="keyword">in</span> tokenizer.batch_tokenize(sequences)]</span><br><span class="line">tokens_str = [b[<span class="number">0</span>] <span class="keyword">for</span> b <span class="keyword">in</span> tokenizer.batch_tokenize(sequences)]</span><br><span class="line"></span><br><span class="line"><span class="comment"># This stacks the batch so that it is repeated across the devices. This is done</span></span><br><span class="line"><span class="comment"># in order to allow for replication even if one has more than one device.</span></span><br><span class="line"><span class="comment"># To take advantage of the multiple devices and infer different sequences on</span></span><br><span class="line"><span class="comment"># each of the devices, make sure to change this line into a reshape.</span></span><br><span class="line"><span class="comment"># a reshape</span></span><br><span class="line">tokens = jnp.asarray(tokens_ids, dtype=jnp.int32)[<span class="literal">None</span>, :]</span><br><span class="line">tokens.shape, idx_stop</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<pre><code>((1, 1, 8333), 5149976)
</code></pre>
<p><strong>â‘¢ å¯¹ç”Ÿæˆçš„æ‰¹æ¬¡è¿›è¡Œæ¨ç†</strong></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Infer</span></span><br><span class="line">outs = apply_fn(parameters, keys, tokens)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Obtain the logits over the genomic features</span></span><br><span class="line">logits = outs[<span class="string">&quot;logits&quot;</span>]</span><br><span class="line"><span class="comment"># Transform them on probabilities</span></span><br><span class="line">probabilities = np.asarray(jax.nn.softmax(logits, axis=-<span class="number">1</span>))[...,-<span class="number">1</span>]</span><br></pre></td></tr></table></figure>

<p><strong>â‘£ ç»˜åˆ¶è¿™æ¡DNAåºåˆ—ä¸Š14ä¸ªåŸºå› ç»„ç‰¹å¾çš„æ¦‚ç‡å›¾</strong></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">plot_features(</span><br><span class="line">    probabilities[<span class="number">0</span>,<span class="number">0</span>],</span><br><span class="line">    probabilities.shape[-<span class="number">2</span>],</span><br><span class="line">    fig_width=<span class="number">20</span>,</span><br><span class="line">    features=config.features,</span><br><span class="line">    order_to_plot=features_rearranged</span><br><span class="line">)</span><br></pre></td></tr></table></figure>
<p><img src="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/images/post/BioAI/03/Segmentation_models%E5%AE%9E%E6%88%98_35_0.png" class="lazyload placeholder" data-srcset="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/images/post/BioAI/03/Segmentation_models%E5%AE%9E%E6%88%98_35_0.png" srcset="https://pic1.zhimg.com/v2-cd38920285d125be80b3eb504052c550_b.webp" alt="png"></p>
<h2 id="2-3-SegmentEnformer"><a href="#2-3-SegmentEnformer" class="headerlink" title="2.3 SegmentEnformer"></a>2.3 SegmentEnformer</h2><p>SegmentEnformerå€ŸåŠ©äº†<a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41592-021-01252-x">Enformer</a>ï¼Œå®ƒç§»é™¤äº†é¢„æµ‹å¤´ï¼Œå¹¶ç”¨ä¸€ä¸ªä¸€ç»´U-Netåˆ†å‰²å¤´å–è€Œä»£ä¹‹ï¼Œä»¥å•æ ¸è‹·é…¸åˆ†è¾¨ç‡<strong>é¢„æµ‹åºåˆ—ä¸­å¤šç§åŸºå› ç»„å…ƒä»¶çš„ä½ç½®</strong>ã€‚</p>
<h3 id="2-3-1-ğŸ”-ç¤ºä¾‹"><a href="#2-3-1-ğŸ”-ç¤ºä¾‹" class="headerlink" title="2.3.1 ğŸ” ç¤ºä¾‹"></a>2.3.1 ğŸ” ç¤ºä¾‹</h3><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> haiku <span class="keyword">as</span> hk</span><br><span class="line"><span class="keyword">import</span> jax</span><br><span class="line"><span class="keyword">import</span> jax.numpy <span class="keyword">as</span> jnp</span><br><span class="line"><span class="keyword">import</span> numpy <span class="keyword">as</span> np</span><br><span class="line"></span><br><span class="line"><span class="keyword">from</span> nucleotide_transformer.enformer.pretrained <span class="keyword">import</span> get_pretrained_segment_enformer_model</span><br><span class="line"><span class="keyword">from</span> nucleotide_transformer.enformer.features <span class="keyword">import</span> FEATURES</span><br><span class="line"></span><br><span class="line"><span class="comment"># Initialize CPU as default JAX device. This makes the code robust to memory leakage on</span></span><br><span class="line"><span class="comment"># the devices.</span></span><br><span class="line">jax.config.update(<span class="string">&quot;jax_platform_name&quot;</span>, <span class="string">&quot;cpu&quot;</span>)</span><br><span class="line"></span><br><span class="line">backend = <span class="string">&quot;cpu&quot;</span></span><br><span class="line">devices = jax.devices(backend)</span><br><span class="line">num_devices = <span class="built_in">len</span>(devices)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Load model</span></span><br><span class="line">parameters, state, forward_fn, tokenizer, config = get_pretrained_segment_enformer_model()</span><br><span class="line">forward_fn = hk.transform_with_state(forward_fn)</span><br><span class="line"></span><br><span class="line">apply_fn = jax.pmap(forward_fn.apply, devices=devices, donate_argnums=(<span class="number">0</span>,))</span><br><span class="line">random_key = jax.random.PRNGKey(seed=<span class="number">0</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Replicate over devices</span></span><br><span class="line">keys = jax.device_put_replicated(random_key, devices=devices)</span><br><span class="line">parameters = jax.device_put_replicated(parameters, devices=devices)</span><br><span class="line">state = jax.device_put_replicated(state, devices=devices)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Get data and tokenize it</span></span><br><span class="line">sequences = [<span class="string">&quot;A&quot;</span> * <span class="number">196_608</span>]</span><br><span class="line">tokens_ids = [b[<span class="number">1</span>] <span class="keyword">for</span> b <span class="keyword">in</span> tokenizer.batch_tokenize(sequences)]</span><br><span class="line">tokens = jnp.stack([jnp.asarray(tokens_ids, dtype=jnp.int32)] * num_devices, axis=<span class="number">0</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Infer</span></span><br><span class="line">outs, state = apply_fn(parameters, state, keys, tokens)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Obtain the logits over the genomic features</span></span><br><span class="line">logits = outs[<span class="string">&quot;logits&quot;</span>]</span><br><span class="line"><span class="comment"># Transform them on probabilities</span></span><br><span class="line">probabilities = np.asarray(jax.nn.softmax(logits, axis=-<span class="number">1</span>))[..., -<span class="number">1</span>]</span><br><span class="line"></span><br><span class="line"><span class="comment"># Get probabilities associated with intron</span></span><br><span class="line">idx_intron = FEATURES.index(<span class="string">&quot;intron&quot;</span>)</span><br><span class="line">probabilities_intron = probabilities[..., idx_intron]</span><br><span class="line"><span class="built_in">print</span>(<span class="string">f&quot;Intron probabilities shape: <span class="subst">&#123;probabilities_intron.shape&#125;</span>&quot;</span>)</span><br></pre></td></tr></table></figure>

<h3 id="2-3-2-å®Œæ•´ä»£ç "><a href="#2-3-2-å®Œæ•´ä»£ç " class="headerlink" title="2.3.2 å®Œæ•´ä»£ç "></a>2.3.2 å®Œæ•´ä»£ç </h3><p>ä¸‹åˆ—ä»£ç å±•ç¤ºäº†å¦‚ä½•å¯¹ä¸€ä¸ª196,608bp çš„åºåˆ—è¿›è¡Œæ¨æ–­å¹¶ç»˜åˆ¶æ¦‚ç‡å›¾ã€‚æ¨¡å—å¯¼å…¥ã€åŸºå› ç»„ä¸‹è½½ã€ç»˜å›¾å‡½æ•°å¦‚å‰æ‰€è¿°ï¼Œä¸å†é‡å¤ã€‚</p>
<p><strong>ï¼ˆ1ï¼‰å®ä¾‹åŒ–SegmentEnformeræ¨ç†å‡½æ•°</strong></p>
<p>ä»¥ä¸‹ä»£ç å…è®¸æ‚¨ä¸‹è½½SegmentEnformerçš„æƒé‡ã€‚å®ƒä¼šè¿”å›æƒé‡å­—å…¸ã€haikuå‰å‘å‡½æ•°ã€åˆ†è¯å™¨å’Œé…ç½®å­—å…¸ã€‚</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">parameters, state, forward_fn, tokenizer, config = get_pretrained_segment_enformer_model()</span><br></pre></td></tr></table></figure>

<pre><code>Downloading model&#39;s weights...
</code></pre>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">forward_fn = hk.transform_with_state(forward_fn)</span><br><span class="line">apply_fn = jax.pmap(forward_fn.apply, devices=devices, donate_argnums=(<span class="number">0</span>,))</span><br><span class="line"></span><br><span class="line"><span class="comment"># Put required quantities for the inference on the devices. This step is not</span></span><br><span class="line"><span class="comment"># reproduced in the second inference since the quantities will already be loaded</span></span><br><span class="line"><span class="comment"># on the devices !</span></span><br><span class="line">random_key = jax.random.PRNGKey(seed=<span class="number">0</span>)</span><br><span class="line">keys = jax.device_put_replicated(random_key, devices=devices)</span><br><span class="line">parameters = jax.device_put_replicated(parameters, devices=devices)</span><br><span class="line">state = jax.device_put_replicated(state, devices=devices)</span><br></pre></td></tr></table></figure>

<p><strong>ï¼ˆ2ï¼‰DNAåºåˆ—åˆ†è¯</strong></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">num_nucleotides = <span class="number">196_608</span></span><br><span class="line">idx_start = <span class="number">2650520</span></span><br><span class="line"></span><br><span class="line">idx_stop = idx_start + num_nucleotides</span><br><span class="line"></span><br><span class="line">sequences = [chr20[idx_start:idx_stop]]</span><br><span class="line">tokens_ids = [b[<span class="number">1</span>] <span class="keyword">for</span> b <span class="keyword">in</span> tokenizer.batch_tokenize(sequences)]</span><br><span class="line">tokens_str = [b[<span class="number">0</span>] <span class="keyword">for</span> b <span class="keyword">in</span> tokenizer.batch_tokenize(sequences)]</span><br><span class="line"></span><br><span class="line"><span class="comment"># This stacks the batch so that it is repeated across the devices. This is done</span></span><br><span class="line"><span class="comment"># in order to allow for replication even if one has more than one device.</span></span><br><span class="line"><span class="comment"># To take advantage of the multiple devices and infer different sequences on</span></span><br><span class="line"><span class="comment"># each of the devices, make sure to change this line into a reshape.</span></span><br><span class="line"><span class="comment"># a reshape</span></span><br><span class="line">tokens = jnp.stack([jnp.asarray(tokens_ids, dtype=jnp.int32)]*num_devices, axis=<span class="number">0</span>)</span><br><span class="line">tokens.shape</span><br><span class="line"></span><br></pre></td></tr></table></figure>


<pre><code>(1, 1, 196608)
</code></pre>
<p><strong>ï¼ˆ3ï¼‰å¯¹ç”Ÿæˆçš„æ‰¹æ¬¡è¿›è¡Œæ¨æ–­</strong></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Infer</span></span><br><span class="line">outs, state = apply_fn(parameters, state, keys, tokens) </span><br><span class="line"></span><br><span class="line"><span class="comment"># Obtain the logits over the genomic features</span></span><br><span class="line">logits = outs[<span class="string">&quot;logits&quot;</span>]</span><br><span class="line"><span class="comment"># Transform them on probabilities</span></span><br><span class="line">probabilities = np.asarray(jax.nn.softmax(logits, axis=-<span class="number">1</span>))[...,-<span class="number">1</span>]</span><br></pre></td></tr></table></figure>


<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">probabilities.shape</span><br></pre></td></tr></table></figure>


<pre><code>(1, 1, 196608, 14)
</code></pre>
<p><strong>ï¼ˆ4ï¼‰ç»˜åˆ¶è¯¥DNAåºåˆ—ä¸­14ä¸ªåŸºå› ç»„ç‰¹å¾çš„æ¦‚ç‡å›¾</strong></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">plot_features(</span><br><span class="line">    probabilities[<span class="number">0</span>, <span class="number">0</span>],</span><br><span class="line">    probabilities.shape[-<span class="number">2</span>],</span><br><span class="line">    fig_width=<span class="number">20</span>,</span><br><span class="line">    features=FEATURES,</span><br><span class="line">    order_to_plot=features_rearranged</span><br><span class="line">)</span><br></pre></td></tr></table></figure>


<p><img src="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/images/post/BioAI/03/Segmentation_models%E5%AE%9E%E6%88%98_48_0.png" class="lazyload placeholder" data-srcset="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/images/post/BioAI/03/Segmentation_models%E5%AE%9E%E6%88%98_48_0.png" srcset="https://pic1.zhimg.com/v2-cd38920285d125be80b3eb504052c550_b.webp" alt="png"></p>
<h2 id="2-4-SegmentBorzoi"><a href="#2-4-SegmentBorzoi" class="headerlink" title="2.4 SegmentBorzoi"></a>2.4 SegmentBorzoi</h2><p>SegmentBorzoiåˆ©ç”¨äº†<a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41588-024-02053-6">Borzoi</a>ï¼Œå®ƒç§»é™¤äº†é¢„æµ‹å¤´ï¼Œå¹¶å°†å…¶æ›¿æ¢ä¸ºä¸€ä¸ªä¸€ç»´U-Netåˆ†å‰²å¤´ï¼Œä»¥<strong>é¢„æµ‹åºåˆ—ä¸­å¤šç§åŸºå› ç»„å…ƒç´ çš„ä½ç½®</strong>ã€‚</p>
<h3 id="2-4-1-ğŸ”-ç¤ºä¾‹"><a href="#2-4-1-ğŸ”-ç¤ºä¾‹" class="headerlink" title="2.4.1 ğŸ” ç¤ºä¾‹"></a>2.4.1 ğŸ” ç¤ºä¾‹</h3><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> haiku <span class="keyword">as</span> hk</span><br><span class="line"><span class="keyword">import</span> jax</span><br><span class="line"><span class="keyword">import</span> jax.numpy <span class="keyword">as</span> jnp</span><br><span class="line"><span class="keyword">import</span> numpy <span class="keyword">as</span> np</span><br><span class="line"></span><br><span class="line"><span class="keyword">from</span> nucleotide_transformer.borzoi.pretrained <span class="keyword">import</span> get_pretrained_segment_borzoi_model</span><br><span class="line"><span class="keyword">from</span> nucleotide_transformer.enformer.features <span class="keyword">import</span> FEATURES</span><br><span class="line"></span><br><span class="line"><span class="comment"># Initialize CPU as default JAX device. This makes the code robust to memory leakage on</span></span><br><span class="line"><span class="comment"># the devices.</span></span><br><span class="line">jax.config.update(<span class="string">&quot;jax_platform_name&quot;</span>, <span class="string">&quot;cpu&quot;</span>)</span><br><span class="line"></span><br><span class="line">backend = <span class="string">&quot;cpu&quot;</span></span><br><span class="line">devices = jax.devices(backend)</span><br><span class="line">num_devices = <span class="built_in">len</span>(devices)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Load model</span></span><br><span class="line">parameters, state, forward_fn, tokenizer, config = get_pretrained_segment_borzoi_model()</span><br><span class="line">forward_fn = hk.transform_with_state(forward_fn)</span><br><span class="line">apply_fn = jax.pmap(forward_fn.apply, devices=devices, donate_argnums=(<span class="number">0</span>,))</span><br><span class="line">random_key = jax.random.PRNGKey(seed=<span class="number">0</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Replicate over devices</span></span><br><span class="line">keys = jax.device_put_replicated(random_key, devices=devices)</span><br><span class="line">parameters = jax.device_put_replicated(parameters, devices=devices)</span><br><span class="line">state = jax.device_put_replicated(state, devices=devices)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Get data and tokenize it</span></span><br><span class="line">sequences = [<span class="string">&quot;A&quot;</span> * <span class="number">524_288</span>]</span><br><span class="line">tokens_ids = [b[<span class="number">1</span>] <span class="keyword">for</span> b <span class="keyword">in</span> tokenizer.batch_tokenize(sequences)]</span><br><span class="line">tokens = jnp.stack([jnp.asarray(tokens_ids, dtype=jnp.int32)] * num_devices, axis=<span class="number">0</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Infer</span></span><br><span class="line">outs, state = apply_fn(parameters, state, keys, tokens)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Obtain the logits over the genomic features</span></span><br><span class="line">logits = outs[<span class="string">&quot;logits&quot;</span>]</span><br><span class="line"><span class="comment"># Transform them on probabilities</span></span><br><span class="line">probabilities = np.asarray(jax.nn.softmax(logits, axis=-<span class="number">1</span>))[..., -<span class="number">1</span>]</span><br><span class="line"></span><br><span class="line"><span class="comment"># Get probabilities associated with intron</span></span><br><span class="line">idx_intron = FEATURES.index(<span class="string">&quot;intron&quot;</span>)</span><br><span class="line">probabilities_intron = probabilities[..., idx_intron]</span><br><span class="line"><span class="built_in">print</span>(<span class="string">f&quot;Intron probabilities shape: <span class="subst">&#123;probabilities_intron.shape&#125;</span>&quot;</span>)</span><br></pre></td></tr></table></figure>

<h3 id="2-4-2-å®Œæ•´ä»£ç "><a href="#2-4-2-å®Œæ•´ä»£ç " class="headerlink" title="2.4.2 å®Œæ•´ä»£ç "></a>2.4.2 å®Œæ•´ä»£ç </h3><p>ä¸‹è¿°ä»£ç å±•ç¤ºäº†å¦‚ä½•å¯¹ä¸€ä¸ª196608 bp ï¼ˆ524288bpï¼Ÿï¼‰çš„åºåˆ—è¿›è¡Œæ¨æ–­å¹¶ç»˜åˆ¶æ¦‚ç‡å›¾ã€‚æ¨¡å—å¯¼å…¥ã€åŸºå› ç»„ä¸‹è½½ã€ç»˜å›¾å‡½æ•°å¦‚å‰æ‰€è¿°ï¼Œä¸å†é‡å¤ã€‚</p>
<p><strong>ï¼ˆ1ï¼‰å®ä¾‹åŒ–SegmentBorzoiæ¨ç†å‡½æ•°</strong></p>
<p>ä»¥ä¸‹ä»£ç å…è®¸æ‚¨ä¸‹è½½SegmentBorzoiçš„æƒé‡ã€‚å®ƒä¼šè¿”å›æƒé‡å­—å…¸ã€haikuå‰å‘å‡½æ•°ã€åˆ†è¯å™¨å’Œé…ç½®å­—å…¸ã€‚</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">parameters, state, forward_fn, tokenizer, config = get_pretrained_segment_borzoi_model()</span><br></pre></td></tr></table></figure>

<pre><code>Downloading model&#39;s weights...
</code></pre>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">forward_fn = hk.transform_with_state(forward_fn)</span><br><span class="line">apply_fn = jax.pmap(forward_fn.apply, devices=devices, donate_argnums=(<span class="number">0</span>,))</span><br><span class="line"></span><br><span class="line"><span class="comment"># Put required quantities for the inference on the devices. This step is not</span></span><br><span class="line"><span class="comment"># reproduced in the second inference since the quantities will already be loaded</span></span><br><span class="line"><span class="comment"># on the devices !</span></span><br><span class="line">random_key = jax.random.PRNGKey(seed=<span class="number">0</span>)</span><br><span class="line">keys = jax.device_put_replicated(random_key, devices=devices)</span><br><span class="line">parameters = jax.device_put_replicated(parameters, devices=devices)</span><br><span class="line">state = jax.device_put_replicated(state, devices=devices)</span><br></pre></td></tr></table></figure>

<p><strong>ï¼ˆ2ï¼‰DNA åºåˆ—åˆ†è¯</strong></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">num_nucleotides = <span class="number">524_288</span></span><br><span class="line"></span><br><span class="line">idx_start = <span class="number">2650520</span></span><br><span class="line">idx_stop = idx_start + num_nucleotides</span><br><span class="line"></span><br><span class="line">sequences = [chr20[idx_start:idx_stop]]</span><br><span class="line">tokens_ids = [b[<span class="number">1</span>] <span class="keyword">for</span> b <span class="keyword">in</span> tokenizer.batch_tokenize(sequences)]</span><br><span class="line">tokens_str = [b[<span class="number">0</span>] <span class="keyword">for</span> b <span class="keyword">in</span> tokenizer.batch_tokenize(sequences)]</span><br><span class="line"></span><br><span class="line"><span class="comment"># This stacks the batch so that it is repeated across the devices. This is done</span></span><br><span class="line"><span class="comment"># in order to allow for replication even if one has more than one device.</span></span><br><span class="line"><span class="comment"># To take advantage of the multiple devices and infer different sequences on</span></span><br><span class="line"><span class="comment"># each of the devices, make sure to change this line into a reshape.</span></span><br><span class="line"><span class="comment"># a reshape</span></span><br><span class="line">tokens = jnp.stack([jnp.asarray(tokens_ids, dtype=jnp.int32)]*num_devices, axis=<span class="number">0</span>)</span><br><span class="line">tokens.shape</span><br><span class="line"></span><br></pre></td></tr></table></figure>


<pre><code>(1, 1, 524288)
</code></pre>
<p><strong>ï¼ˆ3ï¼‰å¯¹ç”Ÿæˆçš„æ‰¹æ¬¡è¿›è¡Œæ¨æ–­</strong></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Infer</span></span><br><span class="line">outs, state = apply_fn(parameters, state, keys, tokens) </span><br><span class="line"></span><br><span class="line"><span class="comment"># Obtain the logits over the genomic features</span></span><br><span class="line">logits = outs[<span class="string">&quot;logits&quot;</span>]</span><br><span class="line"><span class="comment"># Transform them on probabilities</span></span><br><span class="line">probabilities = np.asarray(jax.nn.softmax(logits, axis=-<span class="number">1</span>))[...,-<span class="number">1</span>]</span><br></pre></td></tr></table></figure>


<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">probabilities.shape</span><br></pre></td></tr></table></figure>


<pre><code>(1, 1, 196608, 14)
</code></pre>
<p><strong>ï¼ˆ4ï¼‰ç»˜åˆ¶è¯¥DNAåºåˆ—ä¸­14ä¸ªåŸºå› ç»„ç‰¹å¾çš„æ¦‚ç‡å›¾</strong></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">plot_features(</span><br><span class="line">    probabilities[<span class="number">0</span>, <span class="number">0</span>],</span><br><span class="line">    probabilities.shape[-<span class="number">2</span>],</span><br><span class="line">    fig_width=<span class="number">20</span>,</span><br><span class="line">    features=FEATURES,</span><br><span class="line">    order_to_plot=features_rearranged</span><br><span class="line">)</span><br></pre></td></tr></table></figure>


<p><img src="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/images/post/BioAI/03/Segmentation_models%E5%AE%9E%E6%88%98_61_0.png" class="lazyload placeholder" data-srcset="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/images/post/BioAI/03/Segmentation_models%E5%AE%9E%E6%88%98_61_0.png" srcset="https://pic1.zhimg.com/v2-cd38920285d125be80b3eb504052c550_b.webp" alt="png"></p>
<h1 id="å‚è€ƒæ–‡çŒ®-ğŸ“š"><a href="#å‚è€ƒæ–‡çŒ®-ğŸ“š" class="headerlink" title="å‚è€ƒæ–‡çŒ® ğŸ“š"></a>å‚è€ƒæ–‡çŒ® ğŸ“š</h1><ul>
<li>[1] de Almeida, B.P., Dalla-Torre, H., Richard, G. et al. Annotating the genome at single-nucleotide resolution with DNA foundation models. Nat Methods (2025). <a target="_blank" rel="noopener" href="https://doi.org/10.1038/s41592-025-02881-2">https://doi.org/10.1038/s41592-025-02881-2</a></li>
</ul>
<h1 id="åŠ å…³æ³¨"><a href="#åŠ å…³æ³¨" class="headerlink" title="åŠ å…³æ³¨"></a>åŠ å…³æ³¨</h1><table align="center"><tr>
  <td align="center"><img src="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/social/ç”Ÿä¿¡ä¹‹å·…å…¬ä¼—å·.jpg" class="lazyload placeholder" data-srcset="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/social/ç”Ÿä¿¡ä¹‹å·…å…¬ä¼—å·.jpg" srcset="https://pic1.zhimg.com/v2-cd38920285d125be80b3eb504052c550_b.webp" alt="ç”Ÿä¿¡ä¹‹å·…å¾®ä¿¡å…¬ä¼—å·" style="width: 100px;height: 100px;vertical-align: -20px;border-radius: 0%;margin-right: 0px;margin-bottom: 5px;align: center;"></td>
  <td align="center"><img src="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/social/å°ç¨‹åºç .png" class="lazyload placeholder" data-srcset="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/social/å°ç¨‹åºç .png" srcset="https://pic1.zhimg.com/v2-cd38920285d125be80b3eb504052c550_b.webp" alt="ç”Ÿä¿¡ä¹‹å·…å°ç¨‹åºç " style="width: 100px;height: 100px;vertical-align: -20px;border-radius: 0%;margin-left: 0px;margin-bottom: 5px;align: center"></td>
</tr></table>

      </div>
      <div class="post-tags-categories">
        
        <div class="tags">
          
            <a href="/tags/LLM/" class="">
              LLM
            </a>
          
            <a href="/tags/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF/" class="">
              ç”Ÿç‰©ä¿¡æ¯
            </a>
          
        </div>
        
      </div>
      
        <div class="copyright">
  <ul class="post-copyright">
    <li class="post-copyright-author">
    <strong>ä½œè€…:  </strong>Hualin Liu</a>
    </li>
    <li class="post-copyright-link">
    <strong>æ–‡ç« é“¾æ¥:  </strong>
    <a href="/post/a21.html" target="_blank" title="Segmentation modelså®æˆ˜">https://www.liaochenlanruo.fun/post/a21.html</a>
    </li>
    <li class="post-copyright-license">
      <strong>ç‰ˆæƒå£°æ˜:   </strong>
      æœ¬ç½‘ç«™æ‰€æœ‰æ–‡ç« é™¤ç‰¹åˆ«å£°æ˜å¤–,å‡é‡‡ç”¨ <a rel="license" href="https://creativecommons.org/licenses/by-nc-nd/4.0/" target="_blank" title="Attribution-NonCommercial-NoDerivatives 4.0 International (CC BY-NC-ND 4.0)">CC BY-NC-ND 4.0</a>
      è®¸å¯åè®®ã€‚è½¬è½½è¯·æ³¨æ˜å‡ºå¤„!
    </li>
  </ul>
<div>
      
    </article>
    <!-- ä¸Šä¸€ç¯‡æ–‡ç« å’Œä¸‹ä¸€ç¯‡æ–‡ç«  -->
    
      <!-- æ–‡ç« è¯¦æƒ…é¡µçš„ä¸Šä¸€é¡µå’Œä¸‹ä¸€é¡µ -->
<div class="post-nav">



  
  <div class="post-nav-prev post-nav-item">
    <div class="post-nav-img" style="background-size: cover; 
      background-position: center center;">
      <img class="lazyload lazyload placeholder" src="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/custom/bgs/thumb_69.webp" class="lazyload placeholder" data-srcset="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/custom/bgs/thumb_69.webp" srcset="https://pic1.zhimg.com/v2-cd38920285d125be80b3eb504052c550_b.webp" alt="">
    </div>
    <a href="/post/e149.html" class="post-nav-link">
      <div class="title">
        <i class="fas fa-angle-left"></i> ä¸Šä¸€ç¯‡:
        <div class="title-text">åŸºå› ç»„è¯•é‡‘çŸ³ï¼šåœ¨ä¸­å¿ƒæ³•åˆ™èƒŒæ™¯ä¸‹å¯¹åŸºå› ç»„è¯­è¨€æ¨¡å‹è¿›è¡ŒåŸºå‡†æµ‹è¯•</div>
      </div>
      
      <!-- <div class="content">
        2025å¹´6æœˆï¼Œé¦™æ¸¯ç§‘æŠ€å¤§å­¦è®¡ç®—æœºç§‘å­¦ä¸å·¥ç¨‹ç³»çš„Yihui Wangã€Zhiyuan CaiåŠHao Chenï¼ˆé€šè®¯ä½œè€…
      </div> -->
    </a>
  </div>



  
  <div class="post-nav-next post-nav-item">
    <div class="post-nav-img" style="background-size: cover; 
      background-position: center center;">
      <img class="lazyload lazyload placeholder" src="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/custom/bgs/thumb_227.webp" class="lazyload placeholder" data-srcset="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/custom/bgs/thumb_227.webp" srcset="https://pic1.zhimg.com/v2-cd38920285d125be80b3eb504052c550_b.webp" src="" alt="">
    </div>
    <a href="/post/ea14.html" class="post-nav-link">
      <div class="title">
        ä¸‹ä¸€ç¯‡: <i class="fas fa-angle-right"></i>
        <div class="title-text">ä½¿ç”¨DNAåŸºç¡€æ¨¡å‹åœ¨å•æ ¸è‹·é…¸åˆ†è¾¨ç‡ä¸Šæ³¨é‡ŠåŸºå› ç»„</div>
      </div>
      <!-- <div class="content">
        2025å¹´9æœˆï¼ŒInstaDeepï¼ˆè‹±å›½ä¼¦æ•¦ï¼‰ã€BioNTechï¼ˆå¾·å›½ç¾å› èŒ¨ï¼‰ã€å“¥æœ¬å“ˆæ ¹å¤§å­¦è®¡ç®—æœºç³»ç­‰æœºæ„çš„Bernar
      </div> -->
    </a>
  </div>

</div>

    
    

    <!-- æ‰“èµ -->
    
      <div id="appDonate" class="post-donate">
  <div id="donate_board" class="donate_bar center" ref="donate">
    <a id="btn_donate" class="btn_donate" href="javascript:;" title="æ‰“èµ" @click="showDialogDrawer()"></a>
  </div>
  <transition name="fade">
    <div 
      class="donate-box-mask"
      v-cloak 
      v-show="visible"
      @click="cancelDialogDrawer()"
    >
    </div>
  </transition>
  <transition name="bounce">
    <div class="donate-box" v-cloak v-show="visible">
      <div class="donate-box_close">
        <i class="fas fa-times" aria-hidden="true" @click="cancelDialogDrawer" pointer></i>
      </div>
      <div class="donate-box_title">
        <h4>
          ä½ çš„èµè¯†æ˜¯æˆ‘å‰è¿›çš„åŠ¨åŠ›
        </h4>
      </div>
      <div class="donate-box_tab">
        <div class="Alipay" pointer :class="{'active': tabActive === 'Alipay'}" @click="changeTabActive('Alipay')">
          æ”¯ä»˜å®
        </div>
        <div class="WeChatpay" pointer :class="{'active': tabActive === 'WeChatpay'}" @click="changeTabActive('WeChatpay')">
          å¾®ä¿¡
        </div>
      </div>
      <div class="donate-box_img">
        <div class="AlipayImg" v-show="tabActive === 'Alipay'">
          <img src="/medias/AliPay.jpg" class="lazyload placeholder" data-srcset="/medias/AliPay.jpg" srcset="https://pic1.zhimg.com/v2-cd38920285d125be80b3eb504052c550_b.webp" alt="æ”¯ä»˜å®æ‰“èµ" />
        </div> 
        <div class="WeChatpayImg" v-show="tabActive === 'WeChatpay'">
          <img src="/medias/reward-wepay.jpg" class="lazyload placeholder" data-srcset="/medias/reward-wepay.jpg" srcset="https://pic1.zhimg.com/v2-cd38920285d125be80b3eb504052c550_b.webp" alt="å¾®ä¿¡æ‰“èµ" />
        </div>
      </div>
    </div>
  </transition>
</div>

<script>
  var body = document.body || document.documentElement || window;
  var vm = new Vue({
    el: '#appDonate',
    data: {
      visible: false,
      tabActive: 'Alipay',
      top: 0,
    },
    computed: {
    },
    mounted() {
    },
    methods: {
      showDialogDrawer() {
        this.visible = true;
        // é˜²æ­¢é¡µé¢æ»šåŠ¨ï¼Œåªèƒ½è®©å¼¹æ¡†æ»šåŠ¨
        // function getScroll() {
        //   return {
        //     left: window.pageXOffset || document.documentElement.scrollLeft || document.body.scrollLeft || 0,
        //     top: window.pageYOffset || document.documentElement.scrollTop || document.body.scrollTop || 0
        //   };
        // }
        this.top = $(document).scrollTop() // or getScroll().top
        // console.log('aa', $('.main-content'));
        body.style.cssText = 'overflow: hidden;';
      },
      cancelDialogDrawer() {
        this.visible = false;
        body.removeAttribute('style');
        $(document).scrollTop(this.top)
      },
      changeTabActive(name) {
        this.tabActive = name;
      }
    },
    created() {}
  })
</script>
    

    <!-- åˆ†äº« -->
    
      <!-- https://github.com/overtrue/share.js -->
<!-- æ–‡ç« è¯¦æƒ…é¡µçš„åˆ†äº« -->
<div class="social-share" data-sites="twitter,facebook,google,qq,qzone,wechat,weibo,douban,linkedin" data-wechat-qrcode-helper="<p>å¾®ä¿¡æ‰«ä¸€æ‰«å³å¯åˆ†äº«ï¼</p>"></div>

<script defer src="/js/shareJs/social-share.min.js"></script>
</script>

<style>
  .social-share {
    margin: 20px 0;
  }
</style>


    
    
    <!-- è¯„è®º -->
    <!-- è¯„è®º -->

  <div id="myComment">
    
      <div id="gitment-container"></div>

    
  </div>


<!-- è¿˜éœ€è¦åœ¨åé¢è¿™ä¸ªåœ°å€é‡Œè®¾ç½®script, comment script in themes\hexo-theme-bamboo\layout\_partial\scripts\index.ejs -->


  </div>

  <!-- ç›®å½• -->
  <aside id='l_side'>
  
    
      <section class="widget side_blogger">
  <div class='content'>
    
      
        <a class='avatar flat-box rectangle' href='/about/'>
          <img src='https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/custom/head.jpg'/>
        </a>
      
    
    
      <div class='text'>
        
          <h2>äº†å°˜å…°è‹¥</h2>
        
        
          <p>Research Go</p>

        
        
          <p><span id="jinrishici-sentence">äº†å°˜å…°è‹¥çš„å°å‘</span></p>
          <script src="https://sdk.jinrishici.com/v2/browser/jinrishici.js" charset="utf-8"></script>
        
      </div>
    
    
      <div class="social-wrapper">
        
          
            <a href="/atom.xml"
              class="social fas fa-rss flat-btn"
              target="_blank"
              rel="external nofollow noopener noreferrer">
              
            </a>
          
        
          
            <a href="mailto:liaochenlanruo@webmail.hzau.edu.cn"
              class="social fas fa-envelope flat-btn"
              target="_blank"
              rel="external nofollow noopener noreferrer">
              
            </a>
          
        
          
            <a href="http://github.com/liaochenlanruo"
              class="social fab fa-github flat-btn"
              target="_blank"
              rel="external nofollow noopener noreferrer">
              
            </a>
          
        
          
            <a href="tencent://AddContact/?fromId=50&amp;fromSubId=1&amp;subcmd=all&amp;uin=549830082"
              class="social fab fa-qq flat-btn"
              target="_blank"
              rel="external nofollow noopener noreferrer">
              
            </a>
          
        
      </div>
    
  </div>
</section>

    
  
  
  
    
  

  <div class="layout_sticky">    
    
      
<section class="widget side_toc">
  
  <header>
    
      <i style="color: " class="fas fa-list fa-fw" aria-hidden="true"></i><span class='name' style="color: ">æœ¬æ–‡ç›®å½•</span>
    
  </header>


  <div class='content'>
    <div class="toc-main">
      <div class="toc-content">
        <!-- <ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#1-Segmentation-models-%E7%AE%80%E4%BB%8B"><span class="toc-text">1. Segmentation models ç®€ä»‹</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#2-%E5%A6%82%E4%BD%95%E4%BD%BF%E7%94%A8-%F0%9F%9A%80"><span class="toc-text">2. å¦‚ä½•ä½¿ç”¨ ğŸš€</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#2-1-%E5%AE%89%E8%A3%85%E5%B9%B6%E5%8A%A0%E8%BD%BD%E6%A8%A1%E5%9D%97"><span class="toc-text">2.1 å®‰è£…å¹¶åŠ è½½æ¨¡å—</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#2-2-SegmentNT"><span class="toc-text">2.2 SegmentNT</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#2-2-1-%F0%9F%94%8D-%E7%A4%BA%E4%BE%8B%E4%BB%A3%E7%A0%81"><span class="toc-text">2.2.1 ğŸ” ç¤ºä¾‹ä»£ç </span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#2-2-2-%E5%AE%8C%E6%95%B4%E4%BB%A3%E7%A0%81"><span class="toc-text">2.2.2 å®Œæ•´ä»£ç </span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#2-3-SegmentEnformer"><span class="toc-text">2.3 SegmentEnformer</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#2-3-1-%F0%9F%94%8D-%E7%A4%BA%E4%BE%8B"><span class="toc-text">2.3.1 ğŸ” ç¤ºä¾‹</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#2-3-2-%E5%AE%8C%E6%95%B4%E4%BB%A3%E7%A0%81"><span class="toc-text">2.3.2 å®Œæ•´ä»£ç </span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#2-4-SegmentBorzoi"><span class="toc-text">2.4 SegmentBorzoi</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#2-4-1-%F0%9F%94%8D-%E7%A4%BA%E4%BE%8B"><span class="toc-text">2.4.1 ğŸ” ç¤ºä¾‹</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#2-4-2-%E5%AE%8C%E6%95%B4%E4%BB%A3%E7%A0%81"><span class="toc-text">2.4.2 å®Œæ•´ä»£ç </span></a></li></ol></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%8F%82%E8%80%83%E6%96%87%E7%8C%AE-%F0%9F%93%9A"><span class="toc-text">å‚è€ƒæ–‡çŒ® ğŸ“š</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%8A%A0%E5%85%B3%E6%B3%A8"><span class="toc-text">åŠ å…³æ³¨</span></a></li></ol> -->
        <div class="toc"></div>
      </div>
    </div>
  </div>
</section>
<!-- æ‰‹æœºç«¯ç›®å½•æŒ‰é’® -->
<div id="toc-mobile-btn">
  <i class="fas fa-list-ul" aria-hidden="true"></i>
</div>

      
  <section class="widget side_recent_post">
    
  <header>
    
      <a style="color: " href='/tags/'><i class="fas fa-book fa-fw" aria-hidden="true"></i><span class='name'>æœ€æ–°æ–‡ç« </span></a>
    
  </header>


    <div class='content'>
      
      <!-- hashç®—æ³• -->
      
      <div class="aside-list">
        
          <div class="aside-list-item">
            
            
            

            <div class="post-img-box">
              <a href="/post/e149.html" class="post-img " style="background-size: cover; 
                background-position: center center;">
                <img class="lazyload lazyload placeholder" style="width:100%;height:100%;object-fit:cover;" data-src="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/custom/bgs/thumb_69.webp" class="lazyload placeholder" data-srcset="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/custom/bgs/thumb_69.webp" srcset="https://pic1.zhimg.com/v2-cd38920285d125be80b3eb504052c550_b.webp" alt="">
              </a>
            </div>
            <div class="post-date-title">
              <div>
                
                  <span class="post-date">11-16</span>
                
              </div>
              <a class="post-title" href="/post/e149.html">åŸºå› ç»„è¯•é‡‘çŸ³ï¼šåœ¨ä¸­å¿ƒæ³•åˆ™èƒŒæ™¯ä¸‹å¯¹åŸºå› ç»„è¯­è¨€æ¨¡å‹è¿›è¡ŒåŸºå‡†æµ‹è¯•</a>
            </div>
          </div>
        
          <div class="aside-list-item">
            
            
            

            <div class="post-img-box">
              <a href="/post/a21.html" class="post-img " style="background-size: cover; 
                background-position: center center;">
                <img class="lazyload lazyload placeholder" style="width:100%;height:100%;object-fit:cover;" data-src="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/custom/bgs/thumb_230.webp" class="lazyload placeholder" data-srcset="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/custom/bgs/thumb_230.webp" srcset="https://pic1.zhimg.com/v2-cd38920285d125be80b3eb504052c550_b.webp" alt="">
              </a>
            </div>
            <div class="post-date-title">
              <div>
                
                  <span class="post-date">11-16</span>
                
              </div>
              <a class="post-title" href="/post/a21.html">Segmentation modelså®æˆ˜</a>
            </div>
          </div>
        
          <div class="aside-list-item">
            
            
            

            <div class="post-img-box">
              <a href="/post/ea14.html" class="post-img " style="background-size: cover; 
                background-position: center center;">
                <img class="lazyload lazyload placeholder" style="width:100%;height:100%;object-fit:cover;" data-src="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/custom/bgs/thumb_227.webp" class="lazyload placeholder" data-srcset="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/custom/bgs/thumb_227.webp" srcset="https://pic1.zhimg.com/v2-cd38920285d125be80b3eb504052c550_b.webp" alt="">
              </a>
            </div>
            <div class="post-date-title">
              <div>
                
                  <span class="post-date">11-13</span>
                
              </div>
              <a class="post-title" href="/post/ea14.html">ä½¿ç”¨DNAåŸºç¡€æ¨¡å‹åœ¨å•æ ¸è‹·é…¸åˆ†è¾¨ç‡ä¸Šæ³¨é‡ŠåŸºå› ç»„</a>
            </div>
          </div>
        
          <div class="aside-list-item">
            
            
            

            <div class="post-img-box">
              <a href="/post/f0aa.html" class="post-img " style="background-size: cover; 
                background-position: center center;">
                <img class="lazyload lazyload placeholder" style="width:100%;height:100%;object-fit:cover;" data-src="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/custom/bgs/thumb_226.webp" class="lazyload placeholder" data-srcset="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/custom/bgs/thumb_226.webp" srcset="https://pic1.zhimg.com/v2-cd38920285d125be80b3eb504052c550_b.webp" alt="">
              </a>
            </div>
            <div class="post-date-title">
              <div>
                
                  <span class="post-date">11-11</span>
                
              </div>
              <a class="post-title" href="/post/f0aa.html">åŸºå› ç»„è¯­è¨€æ¨¡å‹çš„æœºé‡ä¸æŒ‘æˆ˜</a>
            </div>
          </div>
        
          <div class="aside-list-item">
            
            
            

            <div class="post-img-box">
              <a href="/post/81e3.html" class="post-img " style="background-size: cover; 
                background-position: center center;">
                <img class="lazyload lazyload placeholder" style="width:100%;height:100%;object-fit:cover;" data-src="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/custom/bgs/thumb_225.webp" class="lazyload placeholder" data-srcset="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/custom/bgs/thumb_225.webp" srcset="https://pic1.zhimg.com/v2-cd38920285d125be80b3eb504052c550_b.webp" alt="">
              </a>
            </div>
            <div class="post-date-title">
              <div>
                
                  <span class="post-date">11-11</span>
                
              </div>
              <a class="post-title" href="/post/81e3.html">AI åœ¨ç”Ÿç‰©ä¿¡æ¯å­¦çš„æ–¹æ³•é©æ–°ä¸åº”ç”¨å…¨æ™¯</a>
            </div>
          </div>
        
      </div>
    </div>
  </section>

    
  </div>
</aside>

  <!-- å›¾ç‰‡æ”¾å¤§ Wrap images with fancybox support -->
  <script defer src="/js/wrapImage.js"></script>
</div>

<!-- æ–‡ç« è¯¦æƒ…é¡µèƒŒæ™¯å›¾ -->
<div id="appBgSwiper" style="position: fixed;left: 0;top: 0;width: 100%;height: 100%;z-index: -2;"
	:style="{'background-color': bgColor ? bgColor : 'transparent'}">
	<transition-group tag="ul" :name="names">
		<li v-for='(image,index) in img' :key='index' v-show="index === mark" class="bg-swiper-box">
			<img :src="image" class="bg-swiper-img no-lazy">
		</li>
	</transition-group>
</div>
<script>
	var vm = new Vue({
		el: '#appBgSwiper',
		data: {
			names: '' || 'fade' || 'fade', // translate-fade fade
			mark: 0,
			img: [],
			bgColor: '',
			time: null
		},
		methods: {   //æ·»åŠ æ–¹æ³•
			change(i, m) {
				if (i > m) {
					// this.names = 'fade';
				} else if (i < m) {
					// this.names = 'fade';
				} else {
					return;
				}
				this.mark = i;
			},
			prev() {
				// this.names = 'fade';
				this.mark--;
				if (this.mark === -1) {
					this.mark = 3;
					return
				}
			},
			next() {
				// this.names = 'fade';
				this.mark++;
				if (this.mark === this.img.length) {
					this.mark = 0;
					return
				}
			},
			autoPlay() {
				// this.names = 'fade';
				this.mark++;
				if (this.mark === this.img.length) {
					this.mark = 0;
					return
				}
			},
			play() {
				let bgImgDelay = '' || '180000'
				let delay = parseInt(bgImgDelay) || 180000;
				this.time = setInterval(this.autoPlay, delay);
			},
			enter() {
				clearInterval(this.time);
			},
			leave() {
				this.play();
			}
		},
		created() {
			this.play()
		},
		beforeDestroy() {
			clearInterval(this.time);
		},
		mounted() {
			let prop = '' || '';
			let isImg = prop.includes('.bmp') || prop.includes('.jpg') || prop.includes('.png') || prop.includes('.tif') || prop.includes('.gif') || prop.includes('.pcx') || prop.includes('.tga') || prop.includes('.exif') || prop.includes('.fpx') || prop.includes('.psd') || prop.includes('.cdr') || prop.includes('.pcd') || prop.includes('.dxf') || prop.includes('.ufo') || prop.includes('.eps') || prop.includes('.ai') || prop.includes('.raw') || prop.includes('.WMF') || prop.includes('.webp') || prop.includes('.jpeg') || prop.includes('http://') || prop.includes('https://')
			if (isImg) {
				let img = prop.split(',');
				let configRoot = '/'
				let arrImg = [];
				img.forEach(el => {
					var Expression = /http(s)?:\/\/([\w-]+\.)+[\w-]+(\/[\w- .\/?%&=]*)?/;
					var objExp = new RegExp(Expression);

					if (objExp.test(el)) {
						// http or https
						arrImg.push(el);
					} else {
						// éhttp or httpså¼€å¤´
						// æœ¬åœ°æ–‡ä»¶
						let firstStr = el.charAt(0);
						if (firstStr == '/') {
							el = el.substr(1); // åˆ é™¤ç¬¬ä¸€ä¸ªå­—ç¬¦ '/',å› ä¸º configRootæœ€åä¸€ä¸ªå­—ç¬¦ä¸º /
						}
						el = configRoot + el;
						arrImg.push(el);
					}
				})
				this.img = arrImg;
			} else {
				this.bgColor = prop;
			}
		}
	})
</script>

<style>
	.bg-swiper-box {
		position: absolute;
		display: block;
		width: 100%;
		height: 100%;
	}

	.bg-swiper-img {
		object-fit: cover;
		width: 100%;
		height: 100%;
	}
</style>


  <script>
  // https://github.com/theme-next/hexo-theme-next/blob/master/layout/_third-party/math/mathjax.swig
  if (typeof MathJax === 'undefined') {
    window.MathJax = {
      loader: {
        source: {
          '[tex]/amsCd': '[tex]/amscd',
          '[tex]/AMScd': '[tex]/amscd'
        }
      },
      tex: {
        inlineMath: {'[+]': [['$', '$']]},
        tags: 'ams'
      },
      options: {
        renderActions: {
          findScript: [10, doc => {
            document.querySelectorAll('script[type^="math/tex"]').forEach(node => {
              const display = !!node.type.match(/; *mode=display/);
              const math = new doc.options.MathItem(node.textContent, doc.inputJax[0], display);
              const text = document.createTextNode('');
              node.parentNode.replaceChild(text, node);
              math.start = {node: text, delim: '', n: 0};
              math.end = {node: text, delim: '', n: 0};
              doc.math.push(math);
            });
          }, '', false],
          insertedScript: [200, () => {
            document.querySelectorAll('mjx-container').forEach(node => {
              let target = node.parentNode;
              if (target.nodeName.toLowerCase() === 'li') {
                target.parentNode.classList.add('has-jax');
              }
            });
          }, '', false]
        }
      }
    };
    (function () {
      var script = document.createElement('script');
      script.src = 'https://unpkg.com/mathjax@3.0/es5/tex-mml-chtml.js';
      script.defer = true;
      document.head.appendChild(script);
    })();
  } else {
    // æ–‡ç« ç« èŠ‚æ ‡é¢˜ä¸èƒ½ä¸º â€œMathJaxâ€ ï¼Œå¦åˆ™ä¼šæŠ¥é”™ã€‚
    MathJax.startup.document.state(0);
    MathJax.texReset();
    MathJax.typeset();
  }
</script>




  <script>
  function loadMermaid() {
    if (document.getElementsByClassName('mermaid').length) {
      if (window.mermaidJsLoad) mermaid.init()
      else {
        loadScript('https://unpkg.com/mermaid/dist/mermaid.min.js').then(() => {
          window.mermaidJsLoad = true
          mermaid.initialize({
            theme: 'default',
          })
          if ('true') {
            mermaid.init();
          }
        })
      }
    }
  };
  document.addEventListener("DOMContentLoaded", function () {
    loadMermaid();
  })

  document.addEventListener('pjax:complete', function () {
    loadMermaid();
  })
  
</script>


      </main>
    </div>

    <!-- é¡µè„š -->
    
  
  
    <!-- åº•éƒ¨é±¼å„¿è·³åŠ¨æ•ˆæœï¼Œä¾èµ–äºjquery-->
<div id="j-fish-skip" style=" position: relative;height: 153px;width: auto;"></div>
<script defer>
  var RENDERER = {
    POINT_INTERVAL: 5,
    FISH_COUNT: 3,
    MAX_INTERVAL_COUNT: 50,
    INIT_HEIGHT_RATE: .5,
    THRESHOLD: 50,
    FISH_COLOR: '',
    init: function () {
      this.setFishColor(); this.setParameters(), this.reconstructMethods(), this.setup(), this.bindEvent(), this.render()
    },
    setFishColor: function () {
      let isDark = JSON.parse(localStorage.getItem('dark')) || JSON.parse('false');
      if (isDark) {
        this.FISH_COLOR = '#222'; // æš—é»‘è‰²ï¼Œæœ‰æ—¶é—´æŠŠè¿™æ•´æˆä¸€ä¸ªå˜é‡
      } else {
        this.FISH_COLOR = '' || 'rgba(66, 185, 133, 0.8)';
      }
    },
    setParameters: function () {
      this.$window = $(window), this.$container = $("#j-fish-skip"), this.$canvas = $("<canvas />"), this.context = this.$canvas.appendTo(this.$container).get(0).getContext("2d"), this.points = [], this.fishes = [], this.watchIds = []
    },
    createSurfacePoints: function () {
      var t = Math.round(this.width / this.POINT_INTERVAL);
      this.pointInterval = this.width / (t - 1), this.points.push(new SURFACE_POINT(this, 0));
      for (var i = 1; i < t; i++) {
        var e = new SURFACE_POINT(this, i * this.pointInterval),
          h = this.points[i - 1];
        e.setPreviousPoint(h), h.setNextPoint(e), this.points.push(e)
      }
    },
    reconstructMethods: function () {
      this.watchWindowSize = this.watchWindowSize.bind(this), this.jdugeToStopResize = this.jdugeToStopResize.bind(this), this.startEpicenter = this.startEpicenter.bind(this), this.moveEpicenter = this.moveEpicenter.bind(this), this.reverseVertical = this.reverseVertical.bind(this), this.render = this.render.bind(this)
    },
    setup: function () {
      this.points.length = 0, this.fishes.length = 0, this.watchIds.length = 0, this.intervalCount = this.MAX_INTERVAL_COUNT, this.width = this.$container.width(), this.height = this.$container.height(), this.fishCount = this.FISH_COUNT * this.width / 500 * this.height / 500, this.$canvas.attr({
        width: this.width,
        height: this.height
      }), this.reverse = !1, this.fishes.push(new FISH(this)), this.createSurfacePoints()
    },
    watchWindowSize: function () {
      this.clearTimer(), this.tmpWidth = this.$window.width(), this.tmpHeight = this.$window.height(), this.watchIds.push(setTimeout(this.jdugeToStopResize, this.WATCH_INTERVAL))
    },
    clearTimer: function () {
      for (; this.watchIds.length > 0;) clearTimeout(this.watchIds.pop())
    },
    jdugeToStopResize: function () {
      var t = this.$window.width(),
        i = this.$window.height(),
        e = t == this.tmpWidth && i == this.tmpHeight;
      this.tmpWidth = t, this.tmpHeight = i, e && this.setup()
    },
    bindEvent: function () {
      this.$window.on("resize", this.watchWindowSize), this.$container.on("mouseenter", this.startEpicenter), this.$container.on("mousemove", this.moveEpicenter)
    },
    getAxis: function (t) {
      var i = this.$container.offset();
      return {
        x: t.clientX - i.left + this.$window.scrollLeft(),
        y: t.clientY - i.top + this.$window.scrollTop()
      }
    },
    startEpicenter: function (t) {
      this.axis = this.getAxis(t)
    },
    moveEpicenter: function (t) {
      var i = this.getAxis(t);
      this.axis || (this.axis = i), this.generateEpicenter(i.x, i.y, i.y - this.axis.y), this.axis = i
    },
    generateEpicenter: function (t, i, e) {
      if (!(i < this.height / 2 - this.THRESHOLD || i > this.height / 2 + this.THRESHOLD)) {
        var h = Math.round(t / this.pointInterval);
        h < 0 || h >= this.points.length || this.points[h].interfere(i, e)
      }
    },
    reverseVertical: function () {
      this.reverse = !this.reverse;
      for (var t = 0, i = this.fishes.length; t < i; t++) this.fishes[t].reverseVertical()
    },
    controlStatus: function () {
      for (var t = 0, i = this.points.length; t < i; t++) this.points[t].updateSelf();
      for (t = 0, i = this.points.length; t < i; t++) this.points[t].updateNeighbors();
      this.fishes.length < this.fishCount && 0 == --this.intervalCount && (this.intervalCount = this.MAX_INTERVAL_COUNT, this.fishes.push(new FISH(this)))
    },
    render: function () {
      requestAnimationFrame(this.render), this.controlStatus(), this.context.clearRect(0, 0, this.width, this.height), this.context.fillStyle = this.FISH_COLOR;
      for (var t = 0, i = this.fishes.length; t < i; t++) this.fishes[t].render(this.context);
      this.context.save(), this.context.globalCompositeOperation = "xor", this.context.beginPath(), this.context.moveTo(0, this.reverse ? 0 : this.height);
      for (t = 0, i = this.points.length; t < i; t++) this.points[t].render(this.context);
      this.context.lineTo(this.width, this.reverse ? 0 : this.height), this.context.closePath(), this.context.fill(), this.context.restore()
    }
  },
  SURFACE_POINT = function (t, i) {
    this.renderer = t, this.x = i, this.init()
  };
  SURFACE_POINT.prototype = {
    SPRING_CONSTANT: .03,
    SPRING_FRICTION: .9,
    WAVE_SPREAD: .3,
    ACCELARATION_RATE: .01,
    init: function () {
      this.initHeight = this.renderer.height * this.renderer.INIT_HEIGHT_RATE, this.height = this.initHeight, this.fy = 0, this.force = {
        previous: 0,
        next: 0
      }
    },
    setPreviousPoint: function (t) {
      this.previous = t
    },
    setNextPoint: function (t) {
      this.next = t
    },
    interfere: function (t, i) {
      this.fy = this.renderer.height * this.ACCELARATION_RATE * (this.renderer.height - this.height - t >= 0 ? -1 : 1) * Math.abs(i)
    },
    updateSelf: function () {
      this.fy += this.SPRING_CONSTANT * (this.initHeight - this.height), this.fy *= this.SPRING_FRICTION, this.height += this.fy
    },
    updateNeighbors: function () {
      this.previous && (this.force.previous = this.WAVE_SPREAD * (this.height - this.previous.height)), this.next && (this.force.next = this.WAVE_SPREAD * (this.height - this.next.height))
    },
    render: function (t) {
      this.previous && (this.previous.height += this.force.previous, this.previous.fy += this.force.previous), this.next && (this.next.height += this.force.next, this.next.fy += this.force.next), t.lineTo(this.x, this.renderer.height - this.height)
    }
  };
  var FISH = function (t) {
    this.renderer = t, this.init()
  };
  FISH.prototype = {
    GRAVITY: .4,
    init: function () {
      this.direction = Math.random() < .5, this.x = this.direction ? this.renderer.width + this.renderer.THRESHOLD : -this.renderer.THRESHOLD, this.previousY = this.y, this.vx = this.getRandomValue(4, 10) * (this.direction ? -1 : 1), this.renderer.reverse ? (this.y = this.getRandomValue(1 * this.renderer.height / 10, 4 * this.renderer.height / 10), this.vy = this.getRandomValue(2, 5), this.ay = this.getRandomValue(.05, .2)) : (this.y = this.getRandomValue(6 * this.renderer.height / 10, 9 * this.renderer.height / 10), this.vy = this.getRandomValue(-5, -2), this.ay = this.getRandomValue(-.2, -.05)), this.isOut = !1, this.theta = 0, this.phi = 0
    },
    getRandomValue: function (t, i) {
      return t + (i - t) * Math.random()
    },
    reverseVertical: function () {
      this.isOut = !this.isOut, this.ay *= -1
    },
    controlStatus: function (t) {
      this.previousY = this.y, this.x += this.vx, this.y += this.vy, this.vy += this.ay, this.renderer.reverse ? this.y > this.renderer.height * this.renderer.INIT_HEIGHT_RATE ? (this.vy -= this.GRAVITY, this.isOut = !0) : (this.isOut && (this.ay = this.getRandomValue(.05, .2)), this.isOut = !1) : this.y < this.renderer.height * this.renderer.INIT_HEIGHT_RATE ? (this.vy += this.GRAVITY, this.isOut = !0) : (this.isOut && (this.ay = this.getRandomValue(-.2, -.05)), this.isOut = !1), this.isOut || (this.theta += Math.PI / 20, this.theta %= 2 * Math.PI, this.phi += Math.PI / 30, this.phi %= 2 * Math.PI), this.renderer.generateEpicenter(this.x + (this.direction ? -1 : 1) * this.renderer.THRESHOLD, this.y, this.y - this.previousY), (this.vx > 0 && this.x > this.renderer.width + this.renderer.THRESHOLD || this.vx < 0 && this.x < -this.renderer.THRESHOLD) && this.init()
    },
    render: function (t) {
      t.save(), t.translate(this.x, this.y), t.rotate(Math.PI + Math.atan2(this.vy, this.vx)), t.scale(1, this.direction ? 1 : -1), t.beginPath(), t.moveTo(-30, 0), t.bezierCurveTo(-20, 15, 15, 10, 40, 0), t.bezierCurveTo(15, -10, -20, -15, -30, 0), t.fill(), t.save(), t.translate(40, 0), t.scale(.9 + .2 * Math.sin(this.theta), 1), t.beginPath(), t.moveTo(0, 0), t.quadraticCurveTo(5, 10, 20, 8), t.quadraticCurveTo(12, 5, 10, 0), t.quadraticCurveTo(12, -5, 20, -8), t.quadraticCurveTo(5, -10, 0, 0), t.fill(), t.restore(), t.save(), t.translate(-3, 0), t.rotate((Math.PI / 3 + Math.PI / 10 * Math.sin(this.phi)) * (this.renderer.reverse ? -1 : 1)), t.beginPath(), this.renderer.reverse ? (t.moveTo(5, 0), t.bezierCurveTo(10, 10, 10, 30, 0, 40), t.bezierCurveTo(-12, 25, -8, 10, 0, 0)) : (t.moveTo(-5, 0), t.bezierCurveTo(-10, -10, -10, -30, 0, -40), t.bezierCurveTo(12, -25, 8, -10, 0, 0)), t.closePath(), t.fill(), t.restore(), t.restore(), this.controlStatus(t)
    }
  }, $(function () {
    RENDERER.init()
    $('.dark').click(function () {
      setTimeout(() => {
        RENDERER.setFishColor();
        RENDERER.context.fill();
      });
    })
  });
</script>
  
  <div class="footer bg-color">
    <div class="footer-main">
      
        
          <div class="link">
            
              
                <a href="/atom.xml" class="social">
                  
                    <i class="fas fa-rss" aria-hidden="true"></i>
                  
                </a>
              
            
              
                <a href="mailto:liaochenlanruo@webmail.hzau.edu.cn" class="social">
                  
                    <i class="fas fa-envelope" aria-hidden="true"></i>
                  
                </a>
              
            
              
                <a target="_blank" rel="noopener" href="https://github.com/liaochenlanruo" class="social">
                  
                    <i class="fab fa-github" aria-hidden="true"></i>
                  
                </a>
              
            
              
                <a href="tencent://AddContact/?fromId=50&amp;fromSubId=1&amp;subcmd=all&amp;uin=549830082" class="social">
                  
                    <i class="fab fa-qq" aria-hidden="true"></i>
                  
                </a>
              
            
          </div>
        
      
        
          <div class="footer-copyright">
            <p>Copyright Â© 2018 - 2024 <a target="_blank" rel="noopener" href="https://github.com/liaochenlanruo">liaochenlanruo</a> | Powered by <a target="_blank" rel="noopener" href="https://hexo.io/zh-cn/docs/">Hexo</a> | Theme <a target="_blank" rel="noopener" href="https://github.com/yuang01/theme">Bamboo</a> </p>

          </div>
        
      
        
          
            <!-- ä¸è’œå­ç»Ÿè®¡ -->
            <!-- ä¸è’œå­ç»Ÿè®¡ -->
<span id="busuanzi_container_site_pv">
      <i class="fas fa-eye" aria-hidden="true"></i>æœ¬ç«™æ€»è®¿é—®é‡ï¼š<span id="busuanzi_value_site_pv"></span> æ¬¡
</span>
<span class="post-meta-divider">|</span>
<span id="busuanzi_container_site_uv" style='display:none'>
      <i class="fas fa-users" aria-hidden="true"></i>æœ¬ç«™è®¿å®¢æ•°ï¼š<span id="busuanzi_value_site_uv"></span> äºº
</span>

          
        
      
        
          <div class="footer-custom">
            
              <div><p><a target="_blank" rel="noopener" href="http://beian.miit.gov.cn/">é„‚ICPå¤‡2021021941å·</a>    <img src="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/logo/%E5%A4%87%E6%A1%88%E5%9B%BE%E6%A0%87.png" class="lazyload placeholder" data-srcset="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/logo/%E5%A4%87%E6%A1%88%E5%9B%BE%E6%A0%87.png" srcset="https://pic1.zhimg.com/v2-cd38920285d125be80b3eb504052c550_b.webp" alt="å¤‡æ¡ˆå›¾æ ‡"> <a target="_blank" rel="noopener" href="http://www.beian.gov.cn/portal/registerSystemInfo?recordcode=42011102004725">é„‚å…¬ç½‘å®‰å¤‡ 42011102004725å·</a></p>
</div>
            
          </div>
        
      
    </div>
  </div>



    <!-- æ¸²æŸ“æš—é»‘æŒ‰é’® -->
    
      <div class="dark" onclick="toggleDarkMode()">
  <div class="dark-content">
    <i class="fas" id="darkIcon" aria-hidden="true"></i>
  </div>
</div>

<script defer>
  $(function() {
    // åˆå§‹åŒ–æš—é»‘æ¨¡å¼çŠ¶æ€
    let isDark = JSON.parse(localStorage.getItem('dark')) || JSON.parse('false');
    updateDarkModeIcon(isDark);
  });

  function toggleDarkMode() {
    const isDark = $(document.body).hasClass('darkModel');
    $(document.body).toggleClass('darkModel');
    localStorage.setItem('dark', !isDark);
    updateDarkModeIcon(!isDark);
  }

  function updateDarkModeIcon(isDark) {
    const iconElement = document.getElementById('darkIcon');
    if (isDark) {
      iconElement.classList.remove('fa-moon');
      iconElement.classList.add('fa-lightbulb');
    } else {
      iconElement.classList.remove('fa-lightbulb');
      iconElement.classList.add('fa-moon');
    }
  }
</script>

    
    <!-- æ¸²æŸ“å›åˆ°é¡¶éƒ¨æŒ‰é’® -->
    
      <div class="goTop top-btn-color" pointer>
  <i class="fas fa-arrow-up" aria-hidden="true"></i>
</div>
<script defer src="/js/goTop.js"></script>

    
    <!-- æ¸²æŸ“å·¦ä¸‹è§’éŸ³ä¹æ’­æ”¾å™¨ -->
    
      <link rel="stylesheet" href="/js/aplayer/APlayer@1.10.1.min.css">
<style>
.aplayer .aplayer-lrc p {
  
  font-size: 12px;
  font-weight: 700;
  line-height: 16px !important;
}

.aplayer .aplayer-lrc p.aplayer-lrc-current {
  
  font-size: 15px;
  color: #42b983;
}


.aplayer.aplayer-fixed.aplayer-narrow .aplayer-body {
  left: -66px !important;
}

.aplayer.aplayer-fixed.aplayer-narrow .aplayer-body:hover {
  left: 0px !important;
}


</style>
<meting-js  
  class=""
  server="tencent"
  type="playlist"
  id="8062553743"
  fixed='true'
  autoplay='false'
  theme='#42b983'
  loop='all'
  order='random'
  preload='auto'
  volume='0.7'
  list-folded='true'
>
</meting-js>

<!-- <style>
  #aplayer {
    position: fixed;
    left: 0;
    bottom: 300px;
  }
</style> -->
<script defer src="https://unpkg.com/aplayer@1.10.1/dist/APlayer.min.js"></script>
<script defer src="https://unpkg.com/meting@2/dist/Meting.min.js"></script>
    

    <!-- å›¾ç‰‡æ”¾å¤§ -->
    
      <script src="https://unpkg.com/@fancyapps/ui@5.0/dist/fancybox/fancybox.umd.js"></script>
    

    <!-- ç™¾åº¦è§£æ -->
    <!-- Baidu Analytics -->

<script async>
    var _hmt = _hmt || [];
    (function () {
        var hm = document.createElement("script");
        hm.src = "https://hm.baidu.com/hm.js?aea6c1471d6299bdd1c87a0836f18675";
        var s = document.getElementsByTagName("script")[0];
        s.parentNode.insertBefore(hm, s);
    })();
</script>

    <!-- Baidu Push -->

<script async>
    (function () {
        var bp = document.createElement('script');
        var curProtocol = window.location.protocol.split(':')[0];
        if (curProtocol === 'https') {
            bp.src = 'https://zz.bdstatic.com/linksubmit/push.js';
        } else {
            bp.src = 'http://push.zhanzhang.baidu.com/push.js';
        }
        var s = document.getElementsByTagName("script")[0];
        s.parentNode.insertBefore(bp, s);
    })();
</script>

    
    <!-- èƒŒæ™¯å½©å¸¦ -->
    
      <script async type="text/javascript" size="100" alpha='0.4' zIndex="-1" src="/js/ribbon.min.js"></script>
    

    <script src="/js/utils/index.js"></script>
    <script src="/js/app.js"></script>
    
    <!-- æ–‡ç« ç›®å½•æ‰€éœ€js -->
<!-- <link href="/js/tocbot/tocbot.css" rel="stylesheet">
<script src="/js/tocbot/tocbot.min.js"></script> -->

<script src="https://cdnjs.cloudflare.com/ajax/libs/tocbot/4.18.2/tocbot.min.js"></script>


<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/tocbot/4.18.2/tocbot.css">

<script>
  var headerEl = 'h1, h2, h3, h4',  //headers 
    content = '.post-detail',//æ–‡ç« å®¹å™¨
    idArr = {};  //æ ‡é¢˜æ•°ç»„ä»¥ç¡®å®šæ˜¯å¦å¢åŠ ç´¢å¼•id
  //add #id
  var option = {
    // Where to render the table of contents.
    tocSelector: '.toc',
    // Where to grab the headings to build the table of contents.
    contentSelector: content,
    // Which headings to grab inside of the contentSelector element.
    headingSelector: headerEl,
    scrollSmooth: true,
    scrollSmoothOffset: -70,
    // headingsOffset: -($(window).height() * 0.4 - 45),
    headingsOffset: -($(window).height() * 0.4 - 70),
    // positionFixedSelector: '.toc-main',
    // positionFixedClass: 'is-position-fixed',
    fixedSidebarOffset: 'auto',
    activeLinkClass: 'is-active-link',
    orderedList: true,
    collapseDepth: 20,
    // onClick: function (e) {},
  }
  if ($('.toc').length > 0) {

    $(content).children(headerEl).each(function () {
      //å»é™¤ç©ºæ ¼ä»¥åŠå¤šä½™æ ‡ç‚¹
      var headerId = $(this).text().replace(/[\s|\~|`|\!|\@|\#|\$|\%|\^|\&|\*|\(|\)|\_|\+|\=|\||\|\[|\]|\{|\}|\;|\:|\"|\'|\,|\<|\.|\>|\/|\?|\ï¼š|\ï¼Œ|\ã€‚]/g, '');

      headerId = headerId.toLowerCase();
      if (idArr[headerId]) {
        //idå·²ç»å­˜åœ¨
        $(this).attr('id', headerId + '-' + idArr[headerId]);
        idArr[headerId]++;
      }
      else {
        //idæœªå­˜åœ¨
        idArr[headerId] = 1;
        $(this).attr('id', headerId);
      }
    });

    document.addEventListener("DOMContentLoaded", function () {
      tocbot.init(option);
      mobileTocClick();
    });

  }

  window.tocScrollFn = function () {
    return bamboo.throttle(function () {
      findHeadPosition();
    }, 100)()
  }
  window.addEventListener('scroll', tocScrollFn);

  const findHeadPosition = function (top) {
    if ($('.toc-list').length <= 0) {
      return false;
    }
    setTimeout(() => {  // or DOMContentLoaded 
      autoScrollToc();
    }, 0);
  }

  const autoScrollToc = function () {
    const $activeItem = document.querySelector('.is-active-link');
    const $cardToc = document.querySelector('.toc-content');
    const activePosition = $activeItem.getBoundingClientRect().top
    const sidebarScrollTop = $cardToc.scrollTop
    if (activePosition > (document.documentElement.clientHeight - 100)) {
      $cardToc.scrollTop = sidebarScrollTop + 150
    }
    if (activePosition < 150) {
      $cardToc.scrollTop = sidebarScrollTop - 150
    }
  }

  document.addEventListener('pjax:send', function () {
    if ($('.toc').length) {
      tocbot.destroy();
    }
  });

  document.addEventListener('pjax:complete', function () {
    if ($('.toc').length) {
      tocbot.init(option);
      mobileTocClick();
    }
  });
  
  // æ‰‹æœºç«¯tocæŒ‰é’®ç‚¹å‡»å‡ºç°ç›®å½•
  const mobileTocClick = function () {
    const $cardTocLayout = document.getElementsByClassName('side_toc')[0];
    const $cardToc = $cardTocLayout.getElementsByClassName('toc-content')[0];
    let right = '45px';
    if (window.innerWidth >= 551 && window.innerWidth <= 992) {
      right = '100px'
    }
    const mobileToc = {
      open: () => {
        $cardTocLayout.style.cssText = 'animation: toc-open .3s; opacity: 1; right: ' + right
      },

      close: () => {
        $cardTocLayout.style.animation = 'toc-close .2s'
        setTimeout(() => {
          $cardTocLayout.style.cssText = "opacity:''; animation: ''; right: ''"
        }, 100)
      }
    }
    document.getElementById('toc-mobile-btn').addEventListener('click', () => {
      if (window.getComputedStyle($cardTocLayout).getPropertyValue('opacity') === '0') mobileToc.open()
      else mobileToc.close()
    })

    $cardToc.addEventListener('click', (e) => {
      if (window.innerWidth < 992) { // å°äº992pxçš„æ—¶å€™
        mobileToc.close()
      }
    })
  }
</script>

<style>
  /* .is-position-fixed {
    position: sticky !important;
    top: 74px;
  }

  .toc-main ul {
    counter-reset: show-list;
  }

  .toc-main ul li::before {
    content: counter(item)".";
    display: block;
    position: absolute;
    left: 12px;
    top: 0;
  } */
</style>
 

<!-- è®¾ç½®å¯¼èˆªèƒŒæ™¯ -->
<script>
  let setHeaderClass = () => {
    const nav = $('#navHeader');
    const navTop = nav.outerHeight();
    const winTop = $(window).scrollTop();
    if(winTop > navTop) {
      nav.addClass('header-bg-color');
    }
    else {
      nav.removeClass('header-bg-color');
    }
  };

  let scrollCollect = () => {
    return bamboo.throttle(function (e) {
      setHeaderClass();
    }, 200)()
  }

  let initHeaderBg = () => {
    setHeaderClass();
  }

  setHeaderClass();
  window.addEventListener('scroll', scrollCollect);

  document.addEventListener('pjax:send', function () {
    window.removeEventListener('scroll', scrollCollect)
  })
  document.addEventListener('pjax:complete', function () {
    window.addEventListener('scroll', scrollCollect);
    setHeaderClass();
  })
</script> 

<!-- æ¸²æŸ“issuesæ ‡ç­¾é‡Œçš„å†…å®¹ -->
<script>
  function loadIssuesJS() {
    if ($(".post-detail").find(".issues-api").length == 0) {
      return;
    } 
    loadScript('/js/issues/index.js');
  };
  $(function () {
    loadIssuesJS();
  });
  document.addEventListener('pjax:complete', function () {
    if (typeof IssuesAPI == "undefined") {
      loadIssuesJS();
    }
  })
</script>

<!-- æ¸²æŸ“è¿œç¨‹jsonåŠ è½½çš„å›¾ç‰‡æ ‡ç­¾(getPhotoOnline)é‡Œçš„å†…å®¹ -->
<script>
  function loadPhotoOnlineJS() {
    if ($(".post-detail").find(".getJsonPhoto-api").length == 0) {
      return;
    } 
    loadScript('/js/getPhotoOnline/index.js');
  };
  $(function () {
    loadPhotoOnlineJS();
  });
  document.addEventListener('pjax:complete', function () {
    if (typeof getPhotoJson == "undefined") {
      loadPhotoOnlineJS();
    }
  })
</script>

<!-- æ¸²æŸ“è¿œç¨‹jsonåŠ è½½çš„talkæ ‡ç­¾(getTalkOnline)é‡Œçš„å†…å®¹ -->
<script>
  function loadTalkOnlineJS() {
    if ($(".post-detail").find(".getJsonTalk-api").length == 0) {
      return;
    } 
    loadScript('https://cdnjs.cloudflare.com/ajax/libs/waterfall.js/1.0.2/waterfall.min.js'); // ç€‘å¸ƒæµæ’ä»¶ï¼Œhttps://raphamorim.io/waterfall.js/
    loadScript('/js/getTalkOnline/index.js');
  };
  $(function () {
    loadTalkOnlineJS();
  });
  document.addEventListener('pjax:complete', function () {
    if (typeof getTalkJson == "undefined") {
      loadTalkOnlineJS();
    }
  })
</script>

<!-- æ¸²æŸ“è¿œç¨‹jsonåŠ è½½çš„site-cardæ ‡ç­¾(getSiteOnline)é‡Œçš„å†…å®¹ -->
<script>
  function loadSiteOnlineJS() {
    if ($(".post-detail").find(".getJsonSite-api").length == 0) {
      return;
    } 
    loadScript('/js/getSiteOnline/index.js');
  };
  $(function () {
    loadSiteOnlineJS();
  });
  document.addEventListener('pjax:complete', function () {
    if (typeof getSiteJson == "undefined") {
      loadSiteOnlineJS();
    }
  })
</script>

<!-- è¾“å…¥æ¡†æ‰“å­—ç‰¹æ•ˆ -->
<!-- è¾“å…¥æ¡†æ‰“å­—ç‰¹æ•ˆ -->

  <script src="/js/activate-power-mode.js"></script>
  <script>
    POWERMODE.colorful = true;  // æ‰“å¼€éšæœºé¢œè‰²ç‰¹æ•ˆ
    POWERMODE.shake = false;    // å…³é—­è¾“å…¥æ¡†æŠ–åŠ¨
    document.body.addEventListener('input', POWERMODE);//ç›‘å¬æ‰“å­—äº‹ä»¶
  </script>


<!-- markdownä»£ç ä¸€é”®å¤åˆ¶åŠŸèƒ½ -->

  <link rel="stylesheet" href="https://unpkg.com/v-plugs-ayu/lib/ayu.css">
  <script src="https://unpkg.com/v-plugs-ayu/lib/ayu.umd.min.js"></script>
  <script src="/js/clipboard/clipboard.min.js"></script>
  <div id="appCopy">
  </div>
  <script data-pjax>
    var vm = new Vue({
      el: '#appCopy',
      data: {
      },
      computed: {
      },
      mounted() {
        const that = this;
        var copy = 'å¤åˆ¶';
        /* code */
        var initCopyCode = function () {
          var copyHtml = '';
          copyHtml += '<button class="btn-copy" data-clipboard-snippet="" style="position:absolute;top:0;right:0;z-index:1;">';
          copyHtml += '<i class="fas fa-copy"></i><span>' + copy + '</span>';
          copyHtml += '</button>';
          $(".post-detail pre").not('.gutter pre').wrap("<div class='codeBox' style='position:relative;width:100%;'></div>")
          $(".post-detail pre").not('.gutter pre').before(copyHtml);
          new ClipboardJS('.btn-copy', {
            target: function (trigger) {
              return trigger.nextElementSibling;
            }
          });
        }
        initCopyCode();
        $('.btn-copy').unbind('click').bind('click', function () {
          doSomething();
        })
        $(document).unbind('keypress').bind('keypress', function (e) {
          if (e.ctrlKey && e.keyCode == 67) {
            doSomething();
          }
        })

        function doSomething() {
          that.$notify({
            title: "æˆåŠŸ",
            content: "ä»£ç å·²å¤åˆ¶ï¼Œè¯·éµå®ˆç›¸å…³æˆæƒåè®®ã€‚",
            type: 'success'
          })
        }
      },
      methods: {
      },
      created() { }
    })
  </script>
  

<!-- å›¾ç‰‡æ‡’åŠ è½½ -->
<script defer src="https://unpkg.com/vanilla-lazyload@17.1.0/dist/lazyload.min.js"></script>
<script>
  // https://www.npmjs.com/package/vanilla-lazyload
  // Set the options globally
  // to make LazyLoad self-initialize
  window.lazyLoadOptions = {
    elements_selector: ".lazyload",
    threshold: 0
  };
  // Listen to the initialization event
  // and get the instance of LazyLoad
  window.addEventListener(
    "LazyLoad::Initialized",
    function (event) {
      window.lazyLoadInstance = event.detail.instance;
    },
    false
  );
  document.addEventListener('DOMContentLoaded', function () {
    lazyLoadInstance.update();
  });
  document.addEventListener('pjax:complete', function () {
    lazyLoadInstance.update();
  });
</script>


<!-- å¡ç‰‡æ»šåŠ¨åŠ¨ç”» -->
   

<!-- è¯„è®ºæ‰€éœ€js -->

  
        <script type="text/javascript">
  var utteranceComment = {};

  function check_utterance() {
    let isDark = JSON.parse(localStorage.getItem('dark')) || JSON.parse('false');
    if (isDark) {
      utteranceComment.Theme = 'github-dark';
    } else {
      utteranceComment.Theme = 'github-light';
    }

    return document.getElementById("gitment-container");
  }
  comment_el = '#gitment-container';
  load_utterance = function () {
    if ($(comment_el).length) {
      // åŒ¿åå‡½æ•°ï¼Œé˜²æ­¢æ±¡æŸ“å…¨å±€å˜é‡
      const HEAD = check_utterance();

      var utterances = document.createElement('script');
      utterances.type = 'text/javascript';
      utterances.async = true;
      utterances.setAttribute('issue-term', 'pathname')
      utterances.setAttribute('theme', utteranceComment.Theme)
      utterances.setAttribute('repo', 'liaochenlanruo/liaochenlanruo.github.io')
      utterances.crossorigin = 'anonymous';
      utterances.src = 'https://utteranc.es/client.js';
      // content æ˜¯è¦æ’å…¥è¯„è®ºçš„åœ°æ–¹
      document.getElementById('gitment-container').appendChild(utterances);

    }
  }

  function dark_utterance() {
    const HEAD = check_utterance();
    if (!HEAD) return;
    const message = {
      type: 'set-theme',
      theme: utteranceComment.Theme
    };
    const utteranceIframe = document.querySelector('iframe');
    utteranceIframe.contentWindow.postMessage(message, 'https://utteranc.es');
  }

  $(document).ready(load_utterance);
  document.addEventListener('pjax:complete', function () {
    load_utterance();
  });

  $('.dark').click(function () {
    setTimeout(() => {
      dark_utterance();
    });
  })

</script>

<style>
  .utterances {
    max-width: inherit !important;
  }
</style>
      


<!-- é¼ æ ‡ç‚¹å‡»ç‰¹æ•ˆ -->
<!-- çˆ±å¿ƒç‚¹å‡» -->

  
    <script async src="/js/cursor/fireworks.js"></script>
  




  <script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js" data-pjax></script>


<!-- è½®æ’­å›¾æ ‡ç­¾ -->
<script>
  var bambooSwiperTag = {};
  function load_swiper() {
    if (!document.querySelectorAll(".post-swiper-container")[0]) return;
    loadCSS("https://unpkg.com/swiper@6/swiper-bundle.min.css")
    loadScript("https://unpkg.com/swiper@6/swiper-bundle.min.js").then(() => {
      pjax_swiper();
    });
  }

  load_swiper();

  function pjax_swiper() {
    bambooSwiperTag.swiper = new Swiper('.post-swiper-container', {
      slidesPerView: 'auto',
      spaceBetween: 8,
      centeredSlides: true,
      loop: true,
      autoplay: true ? {
        delay: 3000,
        stopOnLastSlide: false,
        disableOnInteraction: false,
      } : false,
      pagination: {
        el: '.swiper-pagination',
        clickable: true,
      },
      navigation: {
        nextEl: '.swiper-button-next',
        prevEl: '.swiper-button-prev',
      },
      on:{
        init: function(){
          swiperAnimateCache(this); //éšè—åŠ¨ç”»å…ƒç´  
          swiperAnimate(this); //åˆå§‹åŒ–å®Œæˆå¼€å§‹åŠ¨ç”»
        }, 
        slideChangeTransitionEnd: function(){ 
          swiperAnimate(this); //æ¯ä¸ªslideåˆ‡æ¢ç»“æŸæ—¶ä¹Ÿè¿è¡Œå½“å‰slideåŠ¨ç”»
          //this.slides.eq(this.activeIndex).find('.ani').removeClass('ani'); åŠ¨ç”»åªå±•ç°ä¸€æ¬¡ï¼Œå»é™¤aniç±»å
        } 
      }
    });
  }

  document.addEventListener('pjax:complete', function () {
    if (!document.querySelectorAll(".post-swiper-container")[0]) return;
    if (typeof bambooSwiperTag.swiper === "undefined") {
      load_swiper();
    } else {
      pjax_swiper();
    }
  });
</script>
    <!-- pjax -->
    

<!-- pjax -->


  <script src="/js/pjax@0.2.8/index.js"></script>
  
    <div class="pjax-animate">
  
    <div class="loading-circle"><div id="loader-circle"></div></div>
    <script>
      window.ShowLoading = function() {
        $(".loading-circle").css("display", "block");
      };
      window.HideLoading = function() {
        $(".loading-circle").css("display", "none");
      }
    </script>
  
	<script>
    document.addEventListener('pjax:complete', function () {
      window.HideLoading();
    })
    document.addEventListener('pjax:send', function () {
      window.ShowLoading();
    })
    document.addEventListener('pjax:error', function () {
      window.HideLoading();
    })
	</script>
</div>

  

  <script>
    var pjax = new Pjax({
      elements: 'a[href]:not([href^="#"]):not([href="javascript:void(0)"]):not([no-pjax])',   // æ‹¦æˆªæ­£å¸¸å¸¦é“¾æ¥çš„ a æ ‡ç­¾
      selectors: ["#pjax-container","title"],                                   // æ ¹æ®å®é™…éœ€è¦ç¡®è®¤é‡è½½åŒºåŸŸ
      cacheBust: false,   // url åœ°å€è¿½åŠ æ—¶é—´æˆ³ï¼Œç”¨ä»¥é¿å…æµè§ˆå™¨ç¼“å­˜
      timeout: 5000
    });

    document.addEventListener('pjax:send', function (e) {

      try {
        var currentUrl = window.location.pathname;
        var targetUrl = e.triggerElement.href;
        var banUrl = [""];
        if (banUrl[0] != "") {
          banUrl.forEach(item => {
            if(currentUrl.indexOf(item) != -1 || targetUrl.indexOf(item) != -1) {
              window.location.href = targetUrl;
            }
          });
        }
      } catch (error) {}

      $(window).unbind('resize');
      $(window).unbind('scroll');
      $(document).unbind('scroll');
      $(document).unbind('click');
      $('body').unbind('click');

    })
    
    document.addEventListener('pjax:complete', function () {
      $('script[data-pjax], .pjax-reload script').each(function () {
        $(this).parent().append($(this).remove());
      });
    });

    document.addEventListener('pjax:error', function (e) {
      window.location.href = e.triggerElement.href;
    })
    
    // åˆ·æ–°ä¸ä»é¡¶éƒ¨å¼€å§‹
    document.addEventListener("DOMContentLoaded", function () {
      history.scrollRestoration = 'auto';
    })
  </script>



  </body>
</html>