<!DOCTYPE html>
<html>
  <!-- meta/link... -->
  



<head>
  <meta http-equiv="content-type" content="text/html; charset=utf-8">
  <meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0" name="viewport">
  <!-- Global site tag (gtag.js) - Google Analytics -->

<script async src="https://www.googletagmanager.com/gtag/js?id=UA-139249935-1"></script>
<script>
    window.dataLayer = window.dataLayer || [];
    function gtag() {
        dataLayer.push(arguments);
    }

    gtag('js', new Date());
    gtag('config', 'UA-139249935-1');
</script>


  <title>宏基因组分析流程及代码 | 了尘兰若的小坑</title>

  <link rel="icon" type="image/x-icon, image/vnd.microsoft.icon" href="/favicon.ico">
  <link rel="stylesheet" href="https://at.alicdn.com/t/font_1911880_c1nvbyezg17.css">
  <link href="https://unpkg.com/@fortawesome/fontawesome-free/css/all.min.css" rel="stylesheet">
  <link href="/js/swiper/swiper@5.4.1.min.css" rel="stylesheet">
  
  
    <script>
        var themeModelId = '';
        if (themeModelId) {
            localStorage.setItem('modelId', themeModelId);
        }
    </script>
    
    <script async src="https://cdn.jsdelivr.net/gh/yuang01/live2d-widget@latest/autoload.js"></script>
    <script>
        var live2dOpen = eval('true') || false;
        if (!live2dOpen) {
            localStorage.setItem('waifu-display', 1609323474481);
        }
    </script>
  
  
  
<link rel="stylesheet" href="/css/animate.min.css">

  
<link rel="stylesheet" href="/css/style.css">

  
  
    <link href="https://unpkg.com/@fancyapps/ui@5.0/dist/fancybox/fancybox.css" rel="stylesheet">
  
  
    
<link rel="stylesheet" href="/js/shareJs/share.min.css">

  
  <style>
        @media (max-width: 992px) {
            #waifu {
                display: none;
            }
        }
    </style>
    <script defer src="//cdn.bootcss.com/pace/1.0.2/pace.min.js"></script>
    <link href="//cdn.bootcss.com/pace/1.0.2/themes/pink/pace-theme-flash.css" rel="stylesheet">

    
    
    <!-- 依赖于jquery和vue -->
    <script src="https://unpkg.com/jquery@3.5.1/dist/jquery.min.js"></script>
    <script src="https://unpkg.com/vue@2.6.11/dist/vue.min.js"></script>

    <!-- import link -->
    
        
            
        
            
        
    
    <!-- import script -->
    
        
            
        
            
        
    

<meta name="generator" content="Hexo 6.3.0"><link rel="alternate" href="/atom.xml" title="了尘兰若的小坑" type="application/atom+xml">
</head>

  
  <!-- 预加载动画 -->
  
  
  <div class="preloader_2" id="loader">
  <div class="loader"></div>
</div>

  <script>
    var endLoading = function () {
      document.body.style.overflow = 'auto';
      document.getElementById('loader').classList.add("loaded");
    }
    window.addEventListener('DOMContentLoaded', endLoading);
  </script>


  <body>
    <!-- 判断是否为暗黑风格 -->
    <!-- 判断是否为黑夜模式 -->
<script defer>
  let isDark = JSON.parse(localStorage.getItem('dark')) || JSON.parse('false');

  if (isDark) {
    $(document.body).addClass('darkModel');
  }
</script>

    <!-- 需要在上面加载的js -->
    <script>
  function loadScript(src, cb) {
    return new Promise(resolve => {
      setTimeout(function () {
        var HEAD = document.getElementsByTagName("head")[0] || document.documentElement;
        var script = document.createElement("script");
        script.setAttribute("type", "text/javascript");
        if (cb) {
          if (JSON.stringify(cb)) {
            for (let p in cb) {
              if (p == "onload") {
                script[p] = () => {
                  cb[p]()
                  resolve()
                }
              } else {
                script[p] = cb[p]
                script.onload = resolve
              }
            }
          } else {
            script.onload = () => {
              cb()
              resolve()
            };
          }
        } else {
          script.onload = resolve
        }
        script.setAttribute("src", src);
        HEAD.appendChild(script);
      });
    });
  }

  //https://github.com/filamentgroup/loadCSS
  var loadCSS = function (src) {
    return new Promise(resolve => {
      setTimeout(function () {
        var link = document.createElement('link');
        link.rel = "stylesheet";
        link.href = src;
        link.onload = resolve;
        document.getElementsByTagName("head")[0].appendChild(link);
      });
    });
  };

</script> 

<!-- 轮播图所需要的js -->
<script src="/js/swiper/swiper.min.js"></script>
<script src="/js/swiper/vue-awesome-swiper.js"></script>
<script src="/js/swiper/swiper.animate1.0.3.min.js"></script>

<script type="text/javascript">
  Vue.use(window.VueAwesomeSwiper)
</script>


  <script src="/js/vue-typed-js/index.js"></script>


<!-- 首页的公告滚动插件的js需要重新加载 -->
<script src="/js/vue-seamless-scroll/index.js"></script>

<!-- 打字机效果js -->
<script src="https://unpkg.com/typed.js@2.0.11"></script>


    <div id="safearea">
      <main class="main" id="pjax-container">
        <!-- 头部导航 -->
        
<header class="header  " 
  id="navHeader"
  style="position: fixed;
  left: 0; top: 0; z-index: 10;width: 100%;"
>
  <div class="header-content">
    <div class="bars">
      <div id="appDrawer" class="sidebar-image">
  <div class="drawer-box-icon">
    <i class="fas fa-bars" aria-hidden="true" @click="showDialogDrawer"></i>
  </div>
  
  <transition name="fade">
    <div class="drawer-box_mask" v-cloak style="display: none;" v-show="visible" @click.self="cancelDialogDrawer">
    </div>
  </transition>
  <div class="drawer-box" :class="{'active': visible}">
    <div class="drawer-box-head bg-color">
      <img class="drawer-box-head_logo lazyload placeholder" src="/medias/head.jpg" class="lazyload placeholder" data-srcset="/medias/head.jpg" srcset="https://pic1.zhimg.com/v2-cd38920285d125be80b3eb504052c550_b.webp" alt="logo">
      <h3 class="drawer-box-head_title">了尘兰若的小坑</h3>
      <h5 class="drawer-box-head_desc">分享微生物生物信息学分析方法，欢迎加入QQ群交流945751012，不接受群内广告！</h5>
    </div>
    
    <div class="drawer-box-content">
      <ul class="drawer-box-content_menu">
        
          
            <li class="drawer-box-content_item" style="position: relative;">
              
                <a href="/" class="drawer-menu-item-link">
                  
                    <i class="fas fa-home" aria-hidden="true"></i>
                  
                  <span class="name">首页</span>
                </a>
              
            </li>
          
            <li class="drawer-box-content_item" style="position: relative;">
              
                <a href="javascript:;" class="drawer-menu-item-link has-children" @click="openOrCloseMenu(1)">
                  <span>
                    
                      <i class="fas fa-archive"></i>
                    
                    <span class="name">归档</span>
                  </span>
                  <i class="fas fa-chevron-left arrow " :class="{'icon-rotate': isOpen(1)}" aria-hidden="true"></i>
                </a>
                <ul class="drawer-sub-menu" v-if="isOpen(1)">
                  
                  <li>
                    <a href="/archives">
                      
                      <i class="fas fa-archive" style="margin-top: -20px;"></i>
                      
                      <span>Archives</span>
                    </a>
                  </li>
                  
                  <li>
                    <a href="/categories">
                      
                      <i class="fas fa-bookmark" style="margin-top: -20px;"></i>
                      
                      <span>Categories</span>
                    </a>
                  </li>
                  
                  <li>
                    <a href="/tags">
                      
                      <i class="fas fa-tags" style="margin-top: -20px;"></i>
                      
                      <span>Tags</span>
                    </a>
                  </li>
                  
                </ul>
              
            </li>
          
            <li class="drawer-box-content_item" style="position: relative;">
              
                <a href="javascript:;" class="drawer-menu-item-link has-children" @click="openOrCloseMenu(2)">
                  <span>
                    
                      <i class="fas fa-user"></i>
                    
                    <span class="name">关于</span>
                  </span>
                  <i class="fas fa-chevron-left arrow " :class="{'icon-rotate': isOpen(2)}" aria-hidden="true"></i>
                </a>
                <ul class="drawer-sub-menu" v-if="isOpen(2)">
                  
                  <li>
                    <a href="/about">
                      
                      <i class="fas fa-user-secret" style="margin-top: -20px;"></i>
                      
                      <span>About</span>
                    </a>
                  </li>
                  
                  <li>
                    <a href="/me">
                      
                      <i class="fas fa-user-md" style="margin-top: -20px;"></i>
                      
                      <span>我</span>
                    </a>
                  </li>
                  
                  <li>
                    <a href="/skip/box">
                      
                      <i class="fas fa-box-open" style="margin-top: -20px;"></i>
                      
                      <span>魔盒</span>
                    </a>
                  </li>
                  
                </ul>
              
            </li>
          
            <li class="drawer-box-content_item" style="position: relative;">
              
                <a href="javascript:;" class="drawer-menu-item-link has-children" @click="openOrCloseMenu(3)">
                  <span>
                    
                      <i class="fas fa-link"></i>
                    
                    <span class="name">更多</span>
                  </span>
                  <i class="fas fa-chevron-left arrow " :class="{'icon-rotate': isOpen(3)}" aria-hidden="true"></i>
                </a>
                <ul class="drawer-sub-menu" v-if="isOpen(3)">
                  
                  <li>
                    <a href="/gallery">
                      
                      <i class="fas fa-camera-retro" style="margin-top: -20px;"></i>
                      
                      <span>图库</span>
                    </a>
                  </li>
                  
                  <li>
                    <a target="_blank" rel="noopener" href="https://liaochenlanruo.fun/reference/">
                      
                      <i class="fab fa-unity" style="margin-top: -20px;"></i>
                      
                      <span>速查表</span>
                    </a>
                  </li>
                  
                  <li>
                    <a href="/friends">
                      
                      <i class="fas fa-link" style="margin-top: -20px;"></i>
                      
                      <span>友情连接</span>
                    </a>
                  </li>
                  
                </ul>
              
            </li>
          
            <li class="drawer-box-content_item" style="position: relative;">
              
                <a target="_blank" rel="noopener" href="https://liaochenlanruo.gitee.io" class="drawer-menu-item-link">
                  
                    <i class="fas fa-external-link" aria-hidden="true"></i>
                  
                  <span class="name">English</span>
                </a>
              
            </li>
          
            <li class="drawer-box-content_item" style="position: relative;">
              
                <a target="_blank" rel="noopener" href="https://www.travellings.cn/go-by-clouds.html" class="drawer-menu-item-link">
                  
                    <i class="fas fa-subway" aria-hidden="true"></i>
                  
                  <span class="name">开往</span>
                </a>
              
            </li>
          
        
        
          <li class="drawer-box-content_item">
            <a target="_blank" rel="noopener" href="http://github.com/liaochenlanruo">
              <i class="fas fa-github" aria-hidden="true"></i>
              <span>Github</span>
            </a>
          </li>
        
      </ul>
    </div>
  </div>
</div>

<script>
  var body = document.body || document.documentElement || window;
  var vm = new Vue({
    el: '#appDrawer',
    data: {
      visible: false,
      top: 0,
      openArr: [],
    },
    computed: {
    },
    mounted() {
    },
    methods: {
      isOpen(index) {
        if (this.openArr.includes(index)) {
          return true;
        } else {
          return false;
        }
      },
      openOrCloseMenu(curIndex) {
        const index = this.openArr.indexOf(curIndex);
        if (index !== -1) {
          this.openArr.splice(index, 1);
        } else {
          this.openArr.push(curIndex);
        }
      },
      showDialogDrawer() {
        this.visible = true;
        // 防止页面滚动，只能让弹框滚动
        this.top = $(document).scrollTop()
        body.style.cssText = 'width: 100%; height: 100%;overflow: hidden;';
      },
      cancelDialogDrawer() {
        this.visible = false;
        body.removeAttribute('style');
        $(document).scrollTop(this.top)
      }
    },
    created() {}
  })
</script>

    </div>
    <div class="blog-title" id="author-avatar">
      
        <div class="avatar">
          <img src="/medias/head.jpg" class="lazyload placeholder" data-srcset="/medias/head.jpg" srcset="https://pic1.zhimg.com/v2-cd38920285d125be80b3eb504052c550_b.webp" alt="logo">
        </div>
      
      <a href="/" class="logo">了尘兰若的小坑</a>
    </div>
    <nav class="navbar">
      <ul class="menu">
        
          
            <li class="menu-item" style="position: relative;">
              
                <a href="/" class="menu-item-link" title="首页">
                  
                    <i class="fas fa-home" aria-hidden="true"></i>
                  
                  <span class="name">首页</span>
                </a>
              
            </li>
          
            <li class="menu-item" style="position: relative;">
              
                <a href="javascript:;" class="menu-item-link" title="归档">
                  
                    <i class="fas fa-archive"></i>
                  
                  <span class="name">归档</span>
                  <i class="fas fa-chevron-down arrow" aria-hidden="true"></i>
                </a>
                <ul class="sub-menu">
                  
                  <li>
                    <a href="/archives">
                      
                      <i class="fas fa-archive" style="margin-top: -20px;"></i>
                      
                      <span>Archives</span>
                    </a>
                  </li>
                  
                  <li>
                    <a href="/categories">
                      
                      <i class="fas fa-bookmark" style="margin-top: -20px;"></i>
                      
                      <span>Categories</span>
                    </a>
                  </li>
                  
                  <li>
                    <a href="/tags">
                      
                      <i class="fas fa-tags" style="margin-top: -20px;"></i>
                      
                      <span>Tags</span>
                    </a>
                  </li>
                  
                </ul>
              
            </li>
          
            <li class="menu-item" style="position: relative;">
              
                <a href="javascript:;" class="menu-item-link" title="关于">
                  
                    <i class="fas fa-user"></i>
                  
                  <span class="name">关于</span>
                  <i class="fas fa-chevron-down arrow" aria-hidden="true"></i>
                </a>
                <ul class="sub-menu">
                  
                  <li>
                    <a href="/about">
                      
                      <i class="fas fa-user-secret" style="margin-top: -20px;"></i>
                      
                      <span>About</span>
                    </a>
                  </li>
                  
                  <li>
                    <a href="/me">
                      
                      <i class="fas fa-user-md" style="margin-top: -20px;"></i>
                      
                      <span>我</span>
                    </a>
                  </li>
                  
                  <li>
                    <a href="/skip/box">
                      
                      <i class="fas fa-box-open" style="margin-top: -20px;"></i>
                      
                      <span>魔盒</span>
                    </a>
                  </li>
                  
                </ul>
              
            </li>
          
            <li class="menu-item" style="position: relative;">
              
                <a href="javascript:;" class="menu-item-link" title="更多">
                  
                    <i class="fas fa-link"></i>
                  
                  <span class="name">更多</span>
                  <i class="fas fa-chevron-down arrow" aria-hidden="true"></i>
                </a>
                <ul class="sub-menu">
                  
                  <li>
                    <a href="/gallery">
                      
                      <i class="fas fa-camera-retro" style="margin-top: -20px;"></i>
                      
                      <span>图库</span>
                    </a>
                  </li>
                  
                  <li>
                    <a target="_blank" rel="noopener" href="https://liaochenlanruo.fun/reference/">
                      
                      <i class="fab fa-unity" style="margin-top: -20px;"></i>
                      
                      <span>速查表</span>
                    </a>
                  </li>
                  
                  <li>
                    <a href="/friends">
                      
                      <i class="fas fa-link" style="margin-top: -20px;"></i>
                      
                      <span>友情连接</span>
                    </a>
                  </li>
                  
                </ul>
              
            </li>
          
            <li class="menu-item" style="position: relative;">
              
                <a target="_blank" rel="noopener" href="https://liaochenlanruo.gitee.io" class="menu-item-link" title="English">
                  
                    <i class="fas fa-external-link" aria-hidden="true"></i>
                  
                  <span class="name">English</span>
                </a>
              
            </li>
          
            <li class="menu-item" style="position: relative;">
              
                <a target="_blank" rel="noopener" href="https://www.travellings.cn/go-by-clouds.html" class="menu-item-link" title="开往">
                  
                    <i class="fas fa-subway" aria-hidden="true"></i>
                  
                  <span class="name">开往</span>
                </a>
              
            </li>
          
        
      </ul>
      
      
        <div id="appSearch">
  <div class="search"  @click="showDialog()"><i class="fas fa-search" aria-hidden="true"></i></div>
  <transition name="fade">
    <div class="message-box_wrapper" style="display: none;" v-cloak v-show="dialogVisible" @click.self="cancelDialogVisible()">
      <div class="message-box animated bounceInDown">
        <h2>
          <span>
            <i class="fas fa-search" aria-hidden="true"></i>
            <span class="title">本地搜索</span>
          </span>
          <i class="fas fa-times close" pointer style="float:right;" aria-hidden="true" @click.self="cancelDialogVisible()"></i>
        </h2>
        <form class="site-search-form">
          <input type="text"
            placeholder="请输入关键字"
            id="local-search-input" 
            @click="getSearchFile()"
            class="st-search-input"
            v-model="searchInput"
          />
        </form>
        <div class="result-wrapper">
          <div id="local-search-result" class="local-search-result-cls"></div>
        </div>
      </div>
    </div>
  </transition>
</div>
<script src="/js/local_search.js"></script>
<script>
  var body = document.body || document.documentElement || window;
  var vm = new Vue({
    el: '#appSearch',
    data: {
      dialogVisible: false,
      searchInput: '',
      top: 0,
    },
    computed: {
    },
    mounted() {
      window.addEventListener('pjax:complete', () => {
        this.cancelDialogVisible();
      })
    },
    methods: {
      showDialog() {
        this.dialogVisible = true;
        // 防止页面滚动，只能让弹框滚动
        this.top = $(document).scrollTop()
        body.style.cssText = 'overflow: hidden;';
      },
      getSearchFile() {
        if (!this.searchInput) {
          getSearchFile("/search.xml");
        }
      },
      cancelDialogVisible() {
        this.dialogVisible = false;
        body.removeAttribute('style');
        $(document).scrollTop(this.top)
      },
    },
    created() {}
  })
</script>
<!-- 解决刷新页面闪烁问题，可以在元素上添加display: none, 或者用vue.extend方法，详情：https://blog.csdn.net/qq_31393401/article/details/81017912 -->
<!-- 下面是搜索基本写法 -->
<!-- <script type="text/javascript" id="local.search.active">
  var inputArea = document.querySelector("#local-search-input");
  inputArea.onclick   = function(){ getSearchFile(); this.onclick = null }
  inputArea.onkeydown = function(){ if(event.keyCode == 13) return false }
</script> -->

      

    </nav>
  </div>
  
    <a target="_blank" rel="noopener" href="http://github.com/liaochenlanruo" class="github-corner color-primary" aria-label="View source on GitHub"><svg width="60" height="60" viewBox="0 0 250 250" style="fill:#fff; position: absolute; top: 0; border: 0; right: 0;" aria-hidden="true"><path d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"></path><path d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5,78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5,87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2" fill="currentColor" style="transform-origin: 130px 106px;" class="octo-arm"></path><path d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3,112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z" fill="currentColor" class="octo-body"></path></svg></a><style>.github-corner:hover .octo-arm{animation:octocat-wave 560ms ease-in-out}@keyframes octocat-wave{0%,100%{transform:rotate(0)}20%,60%{transform:rotate(-25deg)}40%,80%{transform:rotate(10deg)}}@media (max-width:500px){.github-corner:hover .octo-arm{animation:none}.github-corner .octo-arm{animation:octocat-wave 560ms ease-in-out}}</style>
  
  
    <div id="he-plugin-simple"></div>
    <script defer>
      WIDGET = {
        CONFIG: {
          "modules": "012",
          "background": 5,
          "tmpColor": "4A4A4A",
          "tmpSize": 16,
          "cityColor": "4A4A4A",
          "citySize": 16,
          "aqiSize": 16,
          "weatherIconSize": 24,
          "alertIconSize": 18,
          "padding": "10px 10px 10px 10px",
          "shadow": "1",
          "language": "auto",
          "borderRadius": 5,
          "fixed": "false",
          "vertical": "middle",
          "horizontal": "center",
          "key": "2784dd3fcb1e4f0f9a9b579bf69641f2"
        }
      }
    </script>
    <script defer src="https://widget.qweather.net/simple/static/js/he-simple-common.js?v=2.0"></script> 
    
</header>
        <!-- 内容区域 -->
        
<!-- prismjs 代码高亮 -->




<div class="bg-dark-floor" style="position: fixed;left: 0;top: 0;width: 100%;height: 100%;z-index: -1;"></div>


  <!-- 文章详情页顶部图片和标题 -->




<div class="post-detail-header" id="thumbnail_canvas" style="background-repeat: no-repeat; background-size: cover; 
  background-position: center center;position: relative;background-image:url('https://fastly.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/custom/bgs/thumb_41.webp')">
  <div class="post-detail-header-mask"></div>
  <canvas id="header_canvas"style="position:absolute;bottom:0;pointer-events:none;"></canvas>
  
  <div class="post-detail-header_info-box">
    <div class="title-box">
      <span class="title">
        宏基因组分析流程及代码
      </span>
    </div>
    
    
      
        <span class="post-detail-header_date">
          <i class="fas fa-calendar"></i> 发表于：2021-01-19 |
        </span>
      

      
        <span class="post-detail-header_categories">
          <i class="iconfont iconbookmark1"></i> 分类：
          
            <a href="/categories/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF/" class="post-detail-header_category">
              生物信息
            </a>
          
        </span>
      

      
        <div class="post-detail-header_wordcount">
          <span class="totalcount">
            <i class="fas fa-file-text-o"></i> 字数统计: 6.4k |
          </span>
  
          <span class="min2read">
            <i class="fas fa-clock"></i> 阅读时长: 36分钟 |
          </span>
  
          
            <span class="reading">
              <i class="fas fa-eye"></i> 阅读量：<span id="busuanzi_value_page_pv"></span>
            </span>
          
        </div>
      
    
  </div>
  
  
    <script defer src="/js/bubble/bubble.js"></script>
  
</div>





<div class="post-detail-content post-row" 
  style="padding-top: 0px;">
  <div class="main-content">
    <article class="post post-detail">
      <div class="post-content">
        <p>本文阐述宏基因组物种分类、组装、bining、基因预测及注释……</p>
<span id="more"></span>

<h1 id="A、软件列表及安装"><a href="#A、软件列表及安装" class="headerlink" title="A、软件列表及安装"></a>A、软件列表及安装</h1><h2 id="A-1-分类相关"><a href="#A-1-分类相关" class="headerlink" title="A.1 分类相关"></a>A.1 分类相关</h2><h3 id="A-1-1-metaphlan-3-0"><a href="#A-1-1-metaphlan-3-0" class="headerlink" title="A.1.1 metaphlan 3.0"></a>A.1.1 metaphlan 3.0</h3><p>MetaPhlAn is a tool for profiling the composition of microbial communities from metagenomic shotgun sequencing data.</p>
<h4 id="a-安装主文件"><a href="#a-安装主文件" class="headerlink" title="a. 安装主文件"></a>a. 安装主文件</h4><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$ conda create -n metaphlan python=3.7 metaphlan</span><br><span class="line">$ conda activate metaphlan</span><br></pre></td></tr></table></figure>
<h4 id="b-安装数据库"><a href="#b-安装数据库" class="headerlink" title="b. 安装数据库"></a>b. 安装数据库</h4><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ metaphlan --install</span><br></pre></td></tr></table></figure>

<h4 id="c-安装依赖包"><a href="#c-安装依赖包" class="headerlink" title="c. 安装依赖包"></a>c. 安装依赖包</h4><ul>
<li>hclust<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ conda install -c bioconda hclust2</span><br></pre></td></tr></table></figure></li>
<li>R, vegan, ape <figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ conda install r-base r-vegan r-ape</span><br></pre></td></tr></table></figure></li>
<li>rbiom<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">$ R</span><br><span class="line">$ install.packages(<span class="string">&quot;rbiom&quot;</span>)</span><br><span class="line">$ quit()</span><br></pre></td></tr></table></figure></li>
</ul>
<h3 id="A-1-2-gtdbtk"><a href="#A-1-2-gtdbtk" class="headerlink" title="A.1.2 gtdbtk"></a>A.1.2 <a target="_blank" rel="noopener" href="https://ecogenomics.github.io/GTDBTk/">gtdbtk</a></h3><h4 id="a-Hardware-requirements"><a href="#a-Hardware-requirements" class="headerlink" title="a. Hardware requirements"></a>a. Hardware requirements</h4><table>
<thead>
<tr>
<th>Domain</th>
<th>Memory</th>
<th>Storage</th>
<th>Time</th>
</tr>
</thead>
<tbody><tr>
<td>Archaea</td>
<td>~8 GB</td>
<td>~27 GB</td>
<td>~1 hour &#x2F; 1,000 genomes @ 64 CPUs</td>
</tr>
<tr>
<td>Bacteria</td>
<td>~150 GB</td>
<td>~27 GB</td>
<td>~1 hour &#x2F; 1,000 genomes @ 64 CPUs</td>
</tr>
</tbody></table>
<h4 id="b-Install-GTDB-Tk-with-conda"><a href="#b-Install-GTDB-Tk-with-conda" class="headerlink" title="b. Install GTDB-Tk with conda"></a>b. Install GTDB-Tk with conda</h4><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ conda create -n gtdbtk -c conda-forge -c bioconda gtdbtk</span><br></pre></td></tr></table></figure>

<h4 id="c-GTDB-Tk-reference-data"><a href="#c-GTDB-Tk-reference-data" class="headerlink" title="c. GTDB-Tk reference data"></a>c. GTDB-Tk reference data</h4><ul>
<li><strong>Note that different versions of the GTDB release data may not run on all versions of GTDB-Tk, below are all supported versions:</strong></li>
</ul>
<table>
<thead>
<tr>
<th>GTDB Release</th>
<th>Minimum version</th>
<th>Maximum version</th>
</tr>
</thead>
<tbody><tr>
<td>R95</td>
<td>1.3.0</td>
<td>N&#x2F;A</td>
</tr>
<tr>
<td>R89</td>
<td>0.3.0</td>
<td>0.1.2</td>
</tr>
<tr>
<td>R86.2</td>
<td>0.2.1</td>
<td>0.2.2</td>
</tr>
<tr>
<td>R86</td>
<td>0.1.0</td>
<td>0.1.6</td>
</tr>
<tr>
<td>R83</td>
<td>0.0.6</td>
<td>0.0.7</td>
</tr>
</tbody></table>
<ul>
<li>Download the reference data</li>
</ul>
 <figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">$ wget https://data.ace.uq.edu.au/public/gtdb/data/releases/release95/95.0/auxillary_files/gtdbtk_r95_data.tar.gz</span><br><span class="line">  </span><br><span class="line">$ tar xvzf gtdbtk_r95_data.tar.gz</span><br></pre></td></tr></table></figure>
<p><strong>注意</strong>：GTDB-Tk requires an environment variable named GTDBTK_DATA_PATH to be set to the directory containing the unarchived GTDB-Tk reference data.</p>
<ul>
<li>You can automatically alias GTDBTK_DATA_PATH whenever the environment is activated by editing {gtdbtk environment path}&#x2F;etc&#x2F;conda&#x2F;activate.d&#x2F;gtdbtk.sh, e.g.:</li>
</ul>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Determine the GTDB-Tk environment path</span></span><br><span class="line">$ conda activate gtdbtk</span><br><span class="line">$ <span class="built_in">which</span> gtdbtk</span><br><span class="line"><span class="comment"># /miniconda3/envs/gtdbtk-1.3.0/bin/gtdbtk</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Edit the activate file</span></span><br><span class="line">$ <span class="built_in">echo</span> <span class="string">&quot;export GTDBTK_DATA_PATH=/path/to/release/package/&quot;</span> &gt; /miniconda3/envs/gtdbtk-1.3.0/etc/conda/activate.d/gtdbtk.sh</span><br></pre></td></tr></table></figure>

<h3 id="A-1-3-Kraken2"><a href="#A-1-3-Kraken2" class="headerlink" title="A.1.3 Kraken2"></a>A.1.3 <a target="_blank" rel="noopener" href="https://github.com/DerrickWood/kraken2">Kraken2</a></h3><h4 id="a-Hardware-requirements-1"><a href="#a-Hardware-requirements-1" class="headerlink" title="a. Hardware requirements"></a>a. Hardware requirements</h4><ul>
<li>Disk space: Construction of a Kraken 2 standard database requires approximately 100 GB of disk space. A test on 01 Jan 2018 of the default installation showed 42 GB of disk space was used to store the genomic library files, 26 GB was used to store the taxonomy information from NCBI, and 29 GB was used to store the Kraken 2 compact hash table.</li>
<li>Memory: To run efficiently, Kraken 2 requires enough free memory to hold the database (primarily the hash table) in RAM. While this can be accomplished with a ramdisk, Kraken 2 will by default load the database into process-local RAM; the –memory-mapping switch to kraken2 will avoid doing so. The default database size is 29 GB (as of Jan. 2018), and you will need slightly more than that in RAM if you want to build the default database.</li>
</ul>
<h4 id="b-Install-Kraken2-with-conda"><a href="#b-Install-Kraken2-with-conda" class="headerlink" title="b. Install Kraken2 with conda"></a>b. Install Kraken2 with conda</h4><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ conda create -n kraken2 kraken2</span><br></pre></td></tr></table></figure>
<h4 id="c-Build-the-database"><a href="#c-Build-the-database" class="headerlink" title="c. Build the database"></a>c. Build the database</h4><ul>
<li>下载数据库。找到一个存储空间比较大的目录并进入，运行如下命令，这里下载的数据库包括archaea，bacteria，plasmid，viral，fungi，protozoa，UniVec和UniVec_Core：<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">nohup</span> bash -c <span class="string">&#x27;for i in archaea bacteria plasmid viral fungi protozoa UniVec UniVec_Core; do kraken2-build --download-library $i --threads 24 --db db_name; done&#x27;</span> &amp;</span><br></pre></td></tr></table></figure></li>
<li>下载分类信息<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">nohup</span> kraken2-build --download-taxonomy --threads 24 --db db_name &amp;</span><br></pre></td></tr></table></figure></li>
<li>建立索引<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">nohup</span> kraken2-build --build --threads 24 --db db_name &amp;</span><br></pre></td></tr></table></figure>
<h4 id="d-序列分类"><a href="#d-序列分类" class="headerlink" title="d. 序列分类"></a>d. 序列分类</h4></li>
</ul>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">kraken2 --paired --threads 24 --unclassified-out unclassified<span class="comment">#.fq --classified-out classified#.fq --output outfile --confidence 0.5 --memory-mapping --use-names --report reportname --report-zero-counts --db $DBNAME reads1 read2</span></span><br></pre></td></tr></table></figure>
<h2 id="A-2-组装、Bining、质量评估"><a href="#A-2-组装、Bining、质量评估" class="headerlink" title="A.2 组装、Bining、质量评估"></a>A.2 组装、Bining、质量评估</h2><h3 id="A-2-1-metawrap"><a href="#A-2-1-metawrap" class="headerlink" title="A.2.1 metawrap"></a>A.2.1 metawrap</h3><p>MetaWRAP is a pipeline for genome-resolved metagenomic data analysis.</p>
<h4 id="a-安装主文件-1"><a href="#a-安装主文件-1" class="headerlink" title="a. 安装主文件"></a>a. 安装主文件</h4><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ conda create -n metawrap python=2.7 metaphlan</span><br></pre></td></tr></table></figure>
<h4 id="b-安装其他分析工具到metawrap环境中"><a href="#b-安装其他分析工具到metawrap环境中" class="headerlink" title="b. 安装其他分析工具到metawrap环境中"></a>b. 安装其他分析工具到metawrap环境中</h4><ul>
<li><strong>cd-hit</strong></li>
<li><strong>coverm</strong> DNA read coverage and relative abundance calculator focused on metagenomics applications</li>
<li><strong>bamm</strong> Metagenomics-focused BAM file manipulation</li>
<li><strong>unitem</strong> Ensemble binning strategies for combining the output of multiple binning methods</li>
<li><strong>humann2</strong> The HMP Unified Metabolic Analysis Network 2</li>
<li><strong><a target="_blank" rel="noopener" href="https://github.com/biobakery/biobakery/wiki/GraPhlAn">graphlan</a></strong> </li>
<li><strong><a target="_blank" rel="noopener" href="https://pypi.org/project/export2graphlan/">export2graphlan</a></strong> </li>
<li><strong><a target="_blank" rel="noopener" href="https://data.ace.uq.edu.au/public/CheckM_databases/">checkm database</a></strong><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">$ conda activeta metawrap</span><br><span class="line">$ conda install cd-hit coverm bamm unitem humann2 graphlan export2graphlan</span><br><span class="line"></span><br><span class="line"><span class="comment"># 找到一个合适的目录并cd进入以便存储checkm数据库</span></span><br><span class="line">$ wget https://data.ace.uq.edu.au/public/CheckM_databases/checkm_data_2015_01_16.tar.gz</span><br><span class="line">$ tar zxvf checkm_data_2015_01_16.tar.gz</span><br><span class="line">$ checkm data setRoot</span><br><span class="line"><span class="comment"># 随后输入数据库所在的full path</span></span><br></pre></td></tr></table></figure></li>
</ul>
<h2 id="A-3-非冗余基因功能注释"><a href="#A-3-非冗余基因功能注释" class="headerlink" title="A.3 非冗余基因功能注释"></a>A.3 非冗余基因功能注释</h2><h3 id="A-3-1-eggNOG-mapper"><a href="#A-3-1-eggNOG-mapper" class="headerlink" title="A.3.1 eggNOG-mapper"></a>A.3.1 eggNOG-mapper</h3><p>功能注释，uniref, eggNOG, KEGG, GO; CAZy; VFDB; CARD; TCDB; PHI。</p>
<h3 id="A-3-2-EnrichM"><a href="#A-3-2-EnrichM" class="headerlink" title="A.3.2 EnrichM"></a>A.3.2 EnrichM</h3><p><a target="_blank" rel="noopener" href="https://github.com/geronimp/enrichM">EnrichM</a> is a set of comparative genomics tools for large sets of metagenome assembled genomes (MAGs). The current functionality includes:</p>
<ul>
<li>A basic annotation pipeline for MAGs.</li>
<li>A pipeline to determine the metabolic pathways that are encoded by MAGs, using KEGG modules as a reference (although custom pathways can be specified)</li>
<li>A pipeline to identify genes or metabolic pathways that are enriched within and between user-defined groups of genomes (groups can be genomes that are related functionally, phylogenetically, recovered from different environments, etc).</li>
<li>To construct metabolic networks from annotated population genomes.</li>
<li>Construct random forest machine learning models from the functional composition of either MAGs, metagenomes or transcriptomes.</li>
<li>Apply these random forest machine learning models to classify new MAGs metagenomes.</li>
</ul>
<h4 id="a-安装主文件-2"><a href="#a-安装主文件-2" class="headerlink" title="a. 安装主文件"></a>a. 安装主文件</h4><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">$ conda create -n enrichm python=3</span><br><span class="line">$ conda install enrichm</span><br></pre></td></tr></table></figure>
<h4 id="b-安装数据库-1"><a href="#b-安装数据库-1" class="headerlink" title="b. 安装数据库"></a>b. 安装数据库</h4><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 约5.7 G</span></span><br><span class="line">$ enrichm data</span><br></pre></td></tr></table></figure>
<p><strong>报错</strong> ：</p>
<blockquote>
<p>Traceback (most recent call last):<br>  File “&#x2F;home&#x2F;hualin&#x2F;miniconda3&#x2F;envs&#x2F;enrichm&#x2F;lib&#x2F;python3.6&#x2F;site-packages&#x2F;enrichm&#x2F;data.py”, line 114, in do<br>    version_remote &#x3D; urllib.request.urlopen(self.ftp + self.VERSION).readline().strip().decode(“utf-8”)<br>AttributeError: module ‘urllib’ has no attribute ‘request’</p>
<p>During handling of the above exception, another exception occurred:</p>
<p>Traceback (most recent call last):<br>  File “&#x2F;home&#x2F;hualin&#x2F;miniconda3&#x2F;envs&#x2F;enrichm&#x2F;bin&#x2F;enrichm”, line 342, in &lt;module&gt;<br>    run.run_enrichm(args, sys.argv)<br>  File “&#x2F;home&#x2F;hualin&#x2F;miniconda3&#x2F;envs&#x2F;enrichm&#x2F;lib&#x2F;python3.6&#x2F;site-packages&#x2F;enrichm&#x2F;run.py”, line 288, in run_enrichm<br>    d.do(args.uninstall, args.dry)<br>  File “&#x2F;home&#x2F;hualin&#x2F;miniconda3&#x2F;envs&#x2F;enrichm&#x2F;lib&#x2F;python3.6&#x2F;site-packages&#x2F;enrichm&#x2F;data.py”, line 117, in do<br>    “Unable to find most current EnrichM database VERSION in ftp. Please complain at <a target="_blank" rel="noopener" href="https://github.com/geronimp/enrichM">https://github.com/geronimp/enrichM</a>“)<br>Exception: Unable to find most current EnrichM database VERSION in ftp. Please complain at <a target="_blank" rel="noopener" href="https://github.com/geronimp/enrichM">https://github.com/geronimp/enrichM</a></p>
</blockquote>
<p><strong>解决方案：将data.py中的’import urllib’替换为’import urllib.request’</strong></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ vim /home/hualin/miniconda3/envs/enrichm/lib/python3.6/site-packages/enrichm/data.py</span><br></pre></td></tr></table></figure>
<h4 id="c-Sepcifying-the-location-of-the-EnrichM-database"><a href="#c-Sepcifying-the-location-of-the-EnrichM-database" class="headerlink" title="c. Sepcifying the location of the EnrichM database"></a>c. Sepcifying the location of the EnrichM database</h4><p>If you would like to store the EnrichM database outside of your home directory, move you need to tell EnrichM where to look. To do this, export a BASH variable named “ENRICHM_DB”:</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ <span class="built_in">export</span> ENRICHM_DB=/path/to/database/ &gt;&gt;~/.bashrc</span><br></pre></td></tr></table></figure>

<p><strong>注意</strong> ：&#x2F;path&#x2F;to&#x2F;database&#x2F;根据实际情况而定！<br><strong>报错</strong>：</p>
<blockquote>
<p>$ enrichm<br>Traceback (most recent call last):<br>  File “&#x2F;home&#x2F;hualin&#x2F;miniconda3&#x2F;envs&#x2F;enrichm&#x2F;bin&#x2F;enrichm”, line 38, in &lt;module&gt;<br>    from enrichm.run import Run<br>  File “&#x2F;home&#x2F;hualin&#x2F;miniconda3&#x2F;envs&#x2F;enrichm&#x2F;lib&#x2F;python3.6&#x2F;site-packages&#x2F;enrichm&#x2F;run.py”, line 24, in &lt;module&gt;<br>    from enrichm.network_analyzer import NetworkAnalyser<br>  File “&#x2F;home&#x2F;hualin&#x2F;miniconda3&#x2F;envs&#x2F;enrichm&#x2F;lib&#x2F;python3.6&#x2F;site-packages&#x2F;enrichm&#x2F;network_analyzer.py”, line 22, in &lt;module&gt;<br>    from enrichm.network_builder import NetworkBuilder<br>  File “&#x2F;home&#x2F;hualin&#x2F;miniconda3&#x2F;envs&#x2F;enrichm&#x2F;lib&#x2F;python3.6&#x2F;site-packages&#x2F;enrichm&#x2F;network_builder.py”, line 24, in &lt;module&gt;<br>    from enrichm.databases import Databases<br>  File “&#x2F;home&#x2F;hualin&#x2F;miniconda3&#x2F;envs&#x2F;enrichm&#x2F;lib&#x2F;python3.6&#x2F;site-packages&#x2F;enrichm&#x2F;databases.py”, line 28, in &lt;module&gt;<br>    class Databases:<br>  File “&#x2F;home&#x2F;hualin&#x2F;miniconda3&#x2F;envs&#x2F;enrichm&#x2F;lib&#x2F;python3.6&#x2F;site-packages&#x2F;enrichm&#x2F;databases.py”, line 36, in Databases<br>    PICKLE_VERSION &#x3D; open(os.path.join(CUR_DATABASE_DIR, ‘VERSION’)).readline().strip()<br>FileNotFoundError: [Errno 2] No such file or directory: ‘&#x2F;new_data&#x2F;hualin&#x2F;db&#x2F;enrichm_database_v10&#x2F;26-11-2018&#x2F;VERSION’</p>
</blockquote>
<p><strong>Solve</strong>: 将下载的数据库文件全部复制一份到“<strong>26-11-2018</strong>”目录中，否则后续运行annotaton时会提示找不到数据库文件。</p>
<h1 id="B、数据分析"><a href="#B、数据分析" class="headerlink" title="B、数据分析"></a>B、数据分析</h1><h2 id="B-1-使用metaphlan从Reads中获取物种分类信息"><a href="#B-1-使用metaphlan从Reads中获取物种分类信息" class="headerlink" title="B.1 使用metaphlan从Reads中获取物种分类信息"></a>B.1 使用metaphlan从Reads中获取物种分类信息</h2><h3 id="Step-1-激活环境"><a href="#Step-1-激活环境" class="headerlink" title="Step 1. 激活环境"></a>Step 1. 激活环境</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ conda activate metaphlan</span><br></pre></td></tr></table></figure>
<h3 id="Step-2-对paired-end-Reads进行注释"><a href="#Step-2-对paired-end-Reads进行注释" class="headerlink" title="Step 2. 对paired-end Reads进行注释"></a>Step 2. 对paired-end Reads进行注释</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ metaphlan Reads1,Reads2 --input_type fastq --bowtie2out Str1.bowtie2.bz2 --<span class="built_in">nproc</span> 10 -o Str1_profiled.txt</span><br></pre></td></tr></table></figure>
<p>Reads1和Reads2分别代表双端测序得到的正向和反向数据；–input_type指定文件格式，我们拿到的下机数据一般为fastq格式；–bowtie2out参数将会保存运行产生的中间文件以便后续重新运行程序时作为输入文件；–nproc指定使用的线程数量；-o指定输出文件名。</p>
<h3 id="Step-3-汇总所有的结果文件"><a href="#Step-3-汇总所有的结果文件" class="headerlink" title="Step 3. 汇总所有的结果文件"></a>Step 3. 汇总所有的结果文件</h3><p>对所有的文件均运行Step 2 ，产生了多个输出文件（*_profiled.txt），可以将它们汇总到一个文件中（merged_abundance_table.txt），便于后续对多个样品进行相互比较。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ merge_metaphlan_tables.py *_profiled.txt &gt; merged_abundance_table.txt</span><br></pre></td></tr></table></figure>
<h3 id="Step-4-从合并的文件中提取种水平的分类"><a href="#Step-4-从合并的文件中提取种水平的分类" class="headerlink" title="Step 4. 从合并的文件中提取种水平的分类"></a>Step 4. 从合并的文件中提取种水平的分类</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ grep -E <span class="string">&quot;s__|clade&quot;</span> merged_abundance_table.txt | sed <span class="string">&#x27;s/^.*s__//g&#x27;</span> | <span class="built_in">cut</span> -f1,3-8 | sed -e <span class="string">&#x27;s/clade_name/body_site/g&#x27;</span> &gt; merged_abundance_table_species.txt</span><br></pre></td></tr></table></figure>

<h3 id="Step-5-绘制样本间种水平的热图"><a href="#Step-5-绘制样本间种水平的热图" class="headerlink" title="Step 5. 绘制样本间种水平的热图"></a>Step 5. 绘制样本间种水平的热图</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ hclust2.py -i merged_abundance_table_species.txt -o abundance_heatmap_species.png --ftop 50 --f_dist_f braycurtis --s_dist_f braycurtis --cell_aspect_ratio 0.5 -l --flabel_size 10 --slabel_size 10 --max_flabel_len 100 --max_slabel_len 100 --minv 0.1 --dpi 300</span><br></pre></td></tr></table></figure>
<h3 id="Step-6-计算样本间的unifrac距离"><a href="#Step-6-计算样本间的unifrac距离" class="headerlink" title="Step 6. 计算样本间的unifrac距离"></a>Step 6. 计算样本间的unifrac距离</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 下载依赖的tree文件和脚本,与要分析的文件放于同一目录下</span></span><br><span class="line">$ wget https://github.com/biobakery/MetaPhlAn/blob/master/metaphlan/utils/mpa_v30_CHOCOPhlAn_201901_species_tree.nwk</span><br><span class="line">$ wget https://github.com/biobakery/MetaPhlAn/blob/master/metaphlan/utils/calculate_unifrac.R</span><br><span class="line"></span><br><span class="line"><span class="comment"># 开始计算距离</span></span><br><span class="line">$ Rscript plot_unifrac.R merged_abundance_table.txt mpa_v30_CHOCOPhlAn_201901_species_tree.nwk unifrac_merged_abundance_table.txt</span><br></pre></td></tr></table></figure>
<h3 id="Step-7-绘制cladogram图"><a href="#Step-7-绘制cladogram图" class="headerlink" title="Step 7. 绘制cladogram图"></a>Step 7. 绘制cladogram图</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 激活依赖软件所在的环境</span></span><br><span class="line">$ conda activate metawrap</span><br><span class="line"></span><br><span class="line"><span class="comment"># 生成绘图所需的文件</span></span><br><span class="line">$ <span class="built_in">tail</span> -n +2 merged_abundance_table.txt | <span class="built_in">cut</span> -f1,3- &gt; merged_abundance_table_reformatted.txt</span><br><span class="line"></span><br><span class="line">$ export2graphlan.py --skip_rows 1 -i merged_abundance_table_reformatted.txt --tree merged_abundance.tree.txt --annotation merged_abundance.annot.txt --most_abundant 100 --abundance_threshold 1 --least_biomarkers 10 --annotations 5,6 --external_annotations 7 --min_clade_size 1</span><br><span class="line"></span><br><span class="line"><span class="comment"># 绘图</span></span><br><span class="line">$ graphlan_annotate.py --annot merged_abundance.annot.txt merged_abundance.tree.txt merged_abundance.xml</span><br><span class="line"></span><br><span class="line">$ graphlan.py --dpi 300 merged_abundance.xml merged_abundance.pdf --external_legends</span><br></pre></td></tr></table></figure>

<h2 id="B-2-使用metawrap对Reads进行组装"><a href="#B-2-使用metawrap对Reads进行组装" class="headerlink" title="B.2 使用metawrap对Reads进行组装"></a>B.2 使用metawrap对Reads进行组装</h2><h3 id="Step-1-激活环境-1"><a href="#Step-1-激活环境-1" class="headerlink" title="Step 1. 激活环境"></a>Step 1. 激活环境</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ conda activate metawrap</span><br></pre></td></tr></table></figure>
<h3 id="Step-2-组装"><a href="#Step-2-组装" class="headerlink" title="Step 2. 组装"></a>Step 2. 组装</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ metawrap assembly -1 Reads1 -2 Reads2 -o Assemble1 -m 300 -t 15 --metaspades</span><br></pre></td></tr></table></figure>
<details>
<summary>点击此处查看参数</summary>
Usage: metaWRAP assembly [options] -1 reads_1.fastq -2 reads_2.fastq -o output_dir

<ul>
<li>Options:<ul>
<li><p>   -1 STR          forward fastq reads</p>
</li>
<li><p>   -2 STR          reverse fastq reads</p>
</li>
<li><p>   -o STR          output directory</p>
</li>
<li><p>   -m INT          memory in GB (default&#x3D;24)</p>
</li>
<li><p>   -t INT          number of threads (defualt&#x3D;1)</p>
</li>
<li><p>   -l INT		minimum length of assembled contigs (default&#x3D;1000)</p>
</li>
<li><p>   --megahit	assemble with megahit (default)</p>
</li>
<li><p>--metaspades	assemble with metaspades instead of megahit (better results but slower and higher memory requirement)</p>
</li>
</ul>
</li>
</ul>
</details>

<p>Reads1和Reads2分别代表双端测序得到的正向和反向数据；-o指定输出目录，-m指定最大可用内存大小，超过设定值后程序会自动退出，建议设大一点，我10G的数据大概需要180G内存；-t指定线程数；–metaspades表示用metaspades进行组装，特别慢，但是组装结果相对好一些。</p>
<h2 id="B-3-使用metawrap对Contigs进行Bining"><a href="#B-3-使用metawrap对Contigs进行Bining" class="headerlink" title="B.3 使用metawrap对Contigs进行Bining"></a>B.3 使用metawrap对Contigs进行Bining</h2><h3 id="Step-1-激活环境-2"><a href="#Step-1-激活环境-2" class="headerlink" title="Step 1. 激活环境"></a>Step 1. 激活环境</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ conda activate metawrap</span><br></pre></td></tr></table></figure>
<h3 id="Step-2-Bining"><a href="#Step-2-Bining" class="headerlink" title="Step 2. Bining"></a>Step 2. Bining</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ metawrap binning -o Str1.INITIAL_BINNING -t 20 -m 200 --universal --run-checkm -a &lt;path of contigs&gt; --metabat2 --maxbin2 --concoct 解压的Reads1 解压的Reads2</span><br></pre></td></tr></table></figure>
<details>
<summary>点击此处查看参数</summary>
Usage: metaWRAP binning [options] -a assembly.fa -o output_dir readsA_1.fastq readsA_2.fastq ... [readsX_1.fastq readsX_2.fastq]

<ul>
<li><p>Note1: Make sure to provide all your separately replicate read files, not the joined file.</p>
</li>
<li><p>Note2: You may provide single end or interleaved reads as well with the use of the correct option</p>
</li>
<li><p>Note3: If the output already has the .bam alignments files from previous runs, the module will skip re-aligning the reads</p>
</li>
<li><p>Options:</p>
<ul>
<li>   -a STR          metagenomic assembly file</li>
<li>   -o STR          output directory</li>
<li>   -t INT          number of threads (default&#x3D;1)</li>
<li>   -m INT		amount of RAM available (default&#x3D;4)</li>
<li>   -l INT		minimum contig length to bin (default&#x3D;1000bp). Note: metaBAT will default to 1500bp minimum</li>
<li>   --metabat2      bin contigs with metaBAT2</li>
<li>   --metabat1	bin contigs with the original metaBAT</li>
<li>   --maxbin2	bin contigs with MaxBin2</li>
<li>   --concoct	bin contigs with CONCOCT</li>
<li>--universal	use universal marker genes instead of bacterial markers in MaxBin2 (improves Archaea binning)</li>
<li>   --run-checkm	immediately run CheckM on the bin results (requires 40GB+ of memory)</li>
<li>   --single-end	non-paired reads mode (provide *.fastq files)</li>
<li>   --interleaved	the input read files contain interleaved paired-end reads</details>
**注意避坑：** 这里的Reads1和Reads2需要提供解压缩后的Reads文件，不但要解压缩，还需要重命名，即后缀名必须为“\_clean\_1.fastq” 和 “\_clean\_2.fastq”，否则软件无法运行。-o指定输出目录；-t指定线程数；-m指定最大内存限制；--run-checkm表明即时检查Bining的质量；--metabat2 --maxbin2 --concoct 指定同时采用这三个分箱工具进行Bining；--universal指定MaxBin2采用universal marker 基因代替 bacterial markers  (improves Archaea binning)。</li>
</ul>
</li>
</ul>
<h3 id="Step-3-整合三种方法的bins-metabat2-maxbin2-concoct-结果"><a href="#Step-3-整合三种方法的bins-metabat2-maxbin2-concoct-结果" class="headerlink" title="Step 3. 整合三种方法的bins (metabat2, maxbin2, concoct) 结果"></a>Step 3. 整合三种方法的bins (metabat2, maxbin2, concoct) 结果</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ metawrap bin_refinement -o F01_BIN_REFINEMENT(输出目录) -t 20 -A str1.INITIAL_BINNING/metabat2_bins/ -B str1.INITIAL_BINNING/maxbin2_bins/ -C str1.INITIAL_BINNING/concoct_bins/ -c 70 -x 5</span><br></pre></td></tr></table></figure>
<details>
<summary>点击此处查看参数</summary>
Usage: metaWRAP bin_refinement [options] -o output_dir -A bin_folderA [-B bin_folderB -C bin_folderC]

<ul>
<li><p>Note: the contig names in different bin folders must be consistant (must come from the same assembly).</p>
</li>
<li><p>Options:</p>
<ul>
<li>-o STR          output directory</li>
<li>   -t INT          number of threads (default&#x3D;1)</li>
<li>   -m INT		memory available (default&#x3D;40)</li>
<li>   -c INT          完整度minimum % completion of bins [should be &gt;50%] (default&#x3D;70)</li>
<li>   -x INT          污染度maximum % contamination of bins that is acceptable (default&#x3D;10)</li>
<li>   -A STR		folder with metagenomic bins (files must have .fa or .fasta extension)</li>
<li>   -B STR		another folder with metagenomic bins</li>
<li>   -C STR		another folder with metagenomic bins</li>
<li>   --skip-refinement	dont use binning_refiner to come up with refined bins based on combinations of binner outputs</li>
<li>   --skip-checkm		dont run CheckM to assess bins</li>
<li>   --skip-consolidation	choose the best version of each bin from all bin refinement iteration</li>
<li>   --keep-ambiguous	for contigs that end up in more than one bin, keep them in all bins (default: keeps them only in the best bin)</li>
<li>   --remove-ambiguous	for contigs that end up in more than one bin, remove them in all bins (default: keeps them only in the best bin)</li>
<li>   --quick			adds –reduced_tree option to checkm, reducing runtime, especially with low memory</li>
</ul>
</li>
</ul>
</details>

<h3 id="Step-4-Blobology可视化bin"><a href="#Step-4-Blobology可视化bin" class="headerlink" title="Step 4. Blobology可视化bin"></a>Step 4. Blobology可视化bin</h3><p><strong>一个坑：</strong> metawrap安装的blast为2.6版本，只能用Version 4 的nt库。而最新的nt库为Version 5，v4已经不再维护了。因此需要更新metawrap安装环境中的blast至2.8.0及以上版本，这里无法通过‘conda updata blast’实现更新，而是需要下载新版本的blast可执行程序，覆盖metawrap环境中的blast程度们。 如果你采用的是默认版本的blast和V5的nt库，会得到报错：“BLAST Database error: Error: Not a valid version 4 database”.</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ metawrap blobology -a Result/Assemble/F01/final_assembly.fasta -t 20 -o F01.BLOBOLOGY --bins F01_BIN_REFINEMENT/metawrap_70_5_bins reads/F01_clean_1.fastq reads/F01_clean_2.fastq</span><br></pre></td></tr></table></figure>
<details>
<summary>点击此处查看参数</summary>
Usage: metaWRAP blobology [options] -a assembly.fasta -o output_dir readsA_1.fastq readsA_2.fastq [readsB_1.fastq readsB_2.fastq ... ]

<ul>
<li><p>Options:</p>
<ul>
<li>   -a STR		assembly fasta file</li>
<li>   -o STR          output directory</li>
<li>   -t INT          number of threads</li>
<li>   --subsamble 	INT	Number of contigs to run blobology on. Subsampling is randomized. (default&#x3D;ALL)</li>
<li>   --bins		STR	Folder containing bins. Contig names must match those of the assembly file. (default&#x3D;None)</li>
</ul>
</li>
</ul>
</details>

<h3 id="Step-5-Bins-定量"><a href="#Step-5-Bins-定量" class="headerlink" title="Step 5. Bins 定量"></a>Step 5. Bins 定量</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ metawrap quant_bins -b F01_BIN_REFINEMENT/metawrap_70_5_bins -t 8 -o F01.QUANT_BINS -a Result/Assemble/F01/final_assembly.fasta reads/F01_clean_1.fastq reads/F01_clean_2.fastq</span><br></pre></td></tr></table></figure>
<details>
<summary>点击此处查看参数</summary>
Usage: metaWRAP quant_bins [options] -b bins_folder -o output_dir -a assembly.fa readsA_1.fastq readsA_2.fastq ... [readsX_1.fastq readsX_2.fastq]

<ul>
<li><p>Options:</p>
<ul>
<li>   -b STR          folder containing draft genomes (bins) in fasta format</li>
<li>   -o STR          output directory</li>
<li>   -a STR		fasta file with entire metagenomic assembly (strongly recommended!)</li>
<li>   -t INT		number of threads</li>
</ul>
</li>
</ul>
</details>

<h3 id="Step-6-重组装"><a href="#Step-6-重组装" class="headerlink" title="Step 6. 重组装"></a>Step 6. 重组装</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">metawrap reassemble_bins -o F01.BIN_REASSEMBLY -1 reads/F01_clean_1.fastq -2 reads/F01_clean_2.fastq -t 20 -m 400 -c 70 -x 10 -b F01_BIN_REFINEMENT/metawrap_70_5_bins</span><br></pre></td></tr></table></figure>
<details>
<summary>点击此处查看参数</summary>
Usage: metaWRAP reassemble_bins [options] -o output_dir -b bin_folder -1 reads_1.fastq -2 reads_2.fastq

<ul>
<li><p>Options:</p>
<ul>
<li>   -b STR		folder with metagenomic bins</li>
<li>   -o STR		output directory</li>
<li>   -1 STR          forward reads to use for reassembly</li>
<li>   -2 STR          reverse reads to use for reassembly</li>
<li>   -t INT		number of threads (default&#x3D;1)</li>
<li>   -m INT		memory in GB (default&#x3D;40)</li>
<li>   -c INT		minimum desired bin completion % (default&#x3D;70)</li>
<li>   -x INT		maximum desired bin contamination % (default&#x3D;10)</li>
<li>   -l INT		minimum contig length to be included in reassembly (default&#x3D;500)</li>
<li>   --strict-cut-off	maximum allowed SNPs for strict read mapping (default&#x3D;2)</li>
<li>   --permissive-cut-off	maximum allowed SNPs for permissive read mapping (default&#x3D;5)</li>
<li>   --skip-checkm		dont run CheckM to assess bins</li>
<li>   --parallel		run spades reassembly in parallel, but only using 1 thread per bin</li>
</ul>
</li>
</ul>
</details>

<h2 id="B-4-MAGs注释"><a href="#B-4-MAGs注释" class="headerlink" title="B.4 MAGs注释"></a>B.4 MAGs注释</h2><h3 id="B-4-1-GTDB-TK-进行物种分类和注释，构建系统发育树"><a href="#B-4-1-GTDB-TK-进行物种分类和注释，构建系统发育树" class="headerlink" title="B.4.1 GTDB-TK 进行物种分类和注释，构建系统发育树"></a>B.4.1 GTDB-TK 进行物种分类和注释，构建系统发育树</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ conda activate gtdbtk</span><br></pre></td></tr></table></figure>

<h4 id="classify-wf——Classify-workflow-包括了Step-1-3"><a href="#classify-wf——Classify-workflow-包括了Step-1-3" class="headerlink" title="classify_wf——Classify workflow (包括了Step 1-3)"></a>classify_wf——Classify workflow (包括了Step 1-3)</h4><p>The classify workflow consists of three steps: identify, align, and classify.</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ time gtdbtk classify_wf --genome_dir metawrap_70_5_bins/ --out_dir classify_wf_output -x .fa --prefix F --cpus 6 -r --debug</span><br></pre></td></tr></table></figure>
<details>
<summary>点击此处查看参数</summary>
usage: gtdbtk classify_wf (--genome_dir GENOME_DIR | --batchfile BATCHFILE) --out_dir OUT_DIR [-x EXTENSION] [--min_perc_aa MIN_PERC_AA] [--prefix PREFIX] [--cpus CPUS]
                          [--pplacer_cpus PPLACER_CPUS] [--force] [--scratch_dir SCRATCH_DIR] [-r] [--min_af MIN_AF] [--debug] [-h]

<ul>
<li><p>mutually exclusive required arguments:</p>
<ul>
<li>--genome_dir GENOME_DIR<br>               directory containing genome files in FASTA format</li>
<li>--batchfile BATCHFILE<br>               path to file describing genomes - tab separated in 2 or 3 columns (FASTA file, genome ID, translation table [optional])</li>
</ul>
</li>
<li><p>required named arguments:</p>
<ul>
<li>--out_dir OUT_DIR     directory to output files</li>
</ul>
</li>
<li><p>optional arguments:</p>
<ul>
<li>-x, --extension EXTENSION<br>                extension of files to process, gz &#x3D; gzipped (default: fna)</li>
<li>--min_perc_aa MIN_PERC_AA<br>                exclude genomes that do not have at least this percentage of AA in the MSA (inclusive bound) (default: 10)</li>
<li>--prefix PREFIX       prefix for all output files (default: gtdbtk)</li>
<li>--cpus CPUS           number of CPUs to use (default: 1)</li>
<li>--pplacer_cpus PPLACER_CPUS<br>                use pplacer_cpus during placement (default: cpus)</li>
<li>--force               continue processing if an error occurs on a single genome (default: False)</li>
<li>--scratch_dir SCRATCH_DIR<br>                Reduce pplacer memory usage by writing to disk (slower).</li>
<li>-r, --recalculate_red<br>                recalculate RED values based on the reference tree and all added user genomes (default: False)</li>
<li>--min_af MIN_AF       minimum alignment fraction to consider closest genome (default: 0.65)</li>
<li>--debug               create intermediate files for debugging purposes (default: False)</li>
<li>-h, --help            show help message</details></li>
</ul>
</li>
</ul>
<h4 id="Step-1-identify——在基因组中寻找marker-genes"><a href="#Step-1-identify——在基因组中寻找marker-genes" class="headerlink" title="Step 1: identify——在基因组中寻找marker genes"></a>Step 1: identify——在基因组中寻找marker genes</h4><p>Translation table选择：use table 11 unless the coding density using table 4 is 5% higher than when using table 11 and the coding density under table 4 is &gt;70%.  GTDB-Tk 不会区分 tables 4和tables 5. 若用户清楚使用哪一个table，可以通过--batchfile指定。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ time gtdbtk identify --genome_dir metawrap_70_5_bins/ --out_dir identify_output --cpus 6 --prefix F --debug -x .fa --write_single_copy_genes</span><br></pre></td></tr></table></figure>

<h4 id="Step-2-align"><a href="#Step-2-align" class="headerlink" title="Step 2: align"></a>Step 2: align</h4><p>Create a multiple sequence alignment based on the AR122&#x2F;BAC120 marker set.<br>Time 3m50.019s</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ time gtdbtk align --identify_dir identify_output/ --out_dir align_output --cpus 3 --prefix F --debug</span><br></pre></td></tr></table></figure>

<h4 id="Step-3-classify"><a href="#Step-3-classify" class="headerlink" title="Step 3: classify"></a>Step 3: classify</h4><p>Determine taxonomic classification of genomes.<br>Time 118m9.386s</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ time gtdbtk classify --genome_dir metawrap_70_5_bins/ --align_dir align_output/ --out_dir classify_output --cpus 3 --prefix F --debug -x .fa -r</span><br></pre></td></tr></table></figure>
<p><strong>注意</strong>：如果内存较小，则加上“–scratch_dir”参数。</p>
<hr>
<h4 id="Step-4-export-msa-这一步不运行"><a href="#Step-4-export-msa-这一步不运行" class="headerlink" title="Step 4: export_msa (这一步不运行)"></a>Step 4: export_msa (这一步不运行)</h4><p>The export_msa will export the untrimmed archaeal or bacterial MSA used in the reference data.</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 古菌 0m1.503s</span></span><br><span class="line">$ time gtdbtk export_msa --domain arc --output msa_arc.faa --debug</span><br><span class="line"></span><br><span class="line"><span class="comment"># 细菌 0m16.679s</span></span><br><span class="line">$ time gtdbtk export_msa --domain bac --output msa_bac.faa --debug</span><br></pre></td></tr></table></figure>
<h4 id="Step-5-trim-msa-这一步不运行"><a href="#Step-5-trim-msa-这一步不运行" class="headerlink" title="Step 5: trim_msa (这一步不运行)"></a>Step 5: trim_msa (这一步不运行)</h4><p>The trim_msa command will trim a MSA given a user-specified mask file, or the archaeal&#x2F;bacterial mask present in the reference data.</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 古菌 0m10.675s</span></span><br><span class="line">$ time gtdbtk trim_msa --untrimmed_msa msa_arc.faa --output msa_arc_trim.faa --reference_mask arc --debug</span><br><span class="line"></span><br><span class="line"><span class="comment"># 细菌 3m28.793s</span></span><br><span class="line">$ time gtdbtk trim_msa --untrimmed_msa msa_bac.faa --output msa_bac_trim.faa --reference_mask bac --debug</span><br></pre></td></tr></table></figure>

<h4 id="Step-6-infer"><a href="#Step-6-infer" class="headerlink" title="Step 6: infer"></a>Step 6: infer</h4><p>Infer tree from multiple sequence alignment.</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#古菌 44m54.884s</span></span><br><span class="line">$ time gtdbtk infer --msa_file align_output/F.ar122.msa.fasta --out_dir infer_out_arc_F --prefix F --cpus 12 --debug</span><br><span class="line"></span><br><span class="line"><span class="comment">#细菌</span></span><br><span class="line">$ time gtdbtk infer --msa_file align_output/F.bac120.msa.fasta --out_dir infer_out_bac_F --prefix F --cpus 12 --debug</span><br></pre></td></tr></table></figure>
<details>
<summary>点击此处查看参数</summary>
usage: gtdbtk infer --msa_file MSA_FILE --out_dir OUT_DIR [--prot_model {JTT,WAG,LG}] [--no_support] [--gamma] [--prefix PREFIX] [--cpus CPUS] [--debug] [-h]

<ul>
<li><p>required named arguments:</p>
<ul>
<li>--msa_file MSA_FILE   multiple sequence alignment in FASTA format</li>
<li>--out_dir OUT_DIR     directory to output files</li>
</ul>
</li>
<li><p>optional arguments:</p>
<ul>
<li>--prot_model {JTT,WAG,LG}<br>                protein substitution model for tree inference (default: WAG)</li>
<li>--no_support          do not compute local support values using the Shimodaira-Hasegawa test (default: False)</li>
<li>--gamma               rescale branch lengths to optimize the Gamma20 likelihood (default: False)</li>
<li>--prefix PREFIX       prefix for all output files (default: gtdbtk)</li>
<li>--cpus CPUS           number of CPUs to use (default: 1)</li>
<li>--debug               create intermediate files for debugging purposes (default: False)</li>
<li>-h, --help            show help message</details></li>
</ul>
</li>
</ul>
<h4 id="Step-7-decorate"><a href="#Step-7-decorate" class="headerlink" title="Step 7: decorate"></a>Step 7: decorate</h4><p>Decorate a tree with the GTDB-Tk taxonomy.</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 古菌</span></span><br><span class="line">$ time gtdbtk decorate --input_tree infer_out_arc_F/F.unrooted.tree --output_tree F.decorate_unrooted_arc.tree --gtdbtk_classification_file classify_output/classify/F.ar122.summary.tsv --debug</span><br><span class="line"></span><br><span class="line"><span class="comment"># 细菌</span></span><br><span class="line">$ time gtdbtk decorate --input_tree infer_out_bac_F/F.unrooted.tree --output_tree F.decorate_unrooted_brc.tree --gtdbtk_classification_file classify_output/classify/F.bac120.summary.tsv --debug</span><br></pre></td></tr></table></figure>
<details>
<summary>点击此处查看参数</summary>
usage: gtdbtk decorate --input_tree INPUT_TREE --output_tree OUTPUT_TREE [--gtdbtk_classification_file GTDBTK_CLASSIFICATION_FILE] [--custom_taxonomy_file CUSTOM_TAXONOMY_FILE]
                       [--debug] [-h]

<ul>
<li><p>required named arguments:</p>
<ul>
<li>--input_tree INPUT_TREE<br>                path to the unrooted tree in Newick format</li>
<li>--output_tree OUTPUT_TREE<br>                path to output the tree</li>
</ul>
</li>
<li><p>optional arguments:</p>
<ul>
<li>--gtdbtk_classification_file GTDBTK_CLASSIFICATION_FILE<br>                file with GTDB-Tk classifications produced by the <code>classify</code> command</li>
<li>--custom_taxonomy_file CUSTOM_TAXONOMY_FILE<br>                file indicating custom taxonomy strings for user genomes, that should contain any genomes belonging to the outgroup</li>
<li>--debug               create intermediate files for debugging purposes (default: False)</li>
<li>-h, --help            show help message</details></li>
</ul>
</li>
</ul>
<h4 id="Step-8-root"><a href="#Step-8-root" class="headerlink" title="Step 8: root"></a>Step 8: root</h4><p>Root a tree using an outgroup.</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ time gtdbtk root --input_tree input.tree --outgroup_taxon p__Nanoarchaeota --output_tree output.tree --gtdbtk_classification_file &lt;file with GTDB-Tk classifications produced by the classify <span class="built_in">command</span>&gt; --debug</span><br></pre></td></tr></table></figure>
<details>
<summary>点击此处查看参数</summary>
usage: gtdbtk root --input_tree INPUT_TREE --outgroup_taxon OUTGROUP_TAXON --output_tree OUTPUT_TREE [--gtdbtk_classification_file GTDBTK_CLASSIFICATION_FILE]
                   [--custom_taxonomy_file CUSTOM_TAXONOMY_FILE] [--debug] [-h]

<ul>
<li><p>required named arguments:</p>
<ul>
<li>--input_tree INPUT_TREE<br>                path to the unrooted tree in Newick format</li>
<li>--outgroup_taxon OUTGROUP_TAXON<br>                taxon to use as outgroup (e.g., p__Patescibacteria or p__Altarchaeota)</li>
<li>--output_tree OUTPUT_TREE<br>                path to output the tree</li>
</ul>
</li>
<li><p>optional arguments:</p>
<ul>
<li>--gtdbtk_classification_file GTDBTK_CLASSIFICATION_FILE<br>                file with GTDB-Tk classifications produced by the <code>classify</code> command</li>
<li>--custom_taxonomy_file CUSTOM_TAXONOMY_FILE<br>                file indicating custom taxonomy strings for user genomes, that should contain any genomes belonging to the outgroup</li>
<li>--debug               create intermediate files for debugging purposes (default: False)</li>
<li>-h, --help            show help message</details></li>
</ul>
</li>
</ul>
<hr>
<h4 id="Step-9-ani-rep——计算ANI值"><a href="#Step-9-ani-rep——计算ANI值" class="headerlink" title="Step 9: ani_rep——计算ANI值"></a>Step 9: ani_rep——计算ANI值</h4><p>Compute the ANI of input genomes to all GTDB-Tk representative genomes.</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ time gtdbtk ani_rep --genome_dir metawrap_70_5_bins/ --out_dir ani_rep/ --cpus 6 -x .fa --prefix F --debug</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ gtdbtk de_novo_wf --genome_dir metawrap_70_5_bins/ --out_dir de_novo_wf --extension .fa --bacteria --outgroup_taxon p__Patescibacteria --prefix F --cpus 6 --debug</span><br></pre></td></tr></table></figure>
<details>
<summary>点击此处查看参数</summary>
usage: gtdbtk ani_rep (--genome_dir GENOME_DIR | --batchfile BATCHFILE) --out_dir OUT_DIR [--no_mash] [--mash_k MASH_K] [--mash_s MASH_S] [--mash_d MASH_D] [--mash_v MASH_V]
                      [--mash_db MASH_DB] [--min_af MIN_AF] [-x EXTENSION] [--prefix PREFIX] [--cpus CPUS] [--debug] [-h]

<ul>
<li><p>mutually exclusive required arguments:</p>
<ul>
<li><p>--genome_dir GENOME_DIR<br>                directory containing genome files in FASTA format</p>
</li>
<li><p>--batchfile BATCHFILE<br>                path to file describing genomes - tab separated in 2 or 3 columns (FASTA file, genome ID, translation table [optional])</p>
</li>
<li><p>- required named arguments:</p>
</li>
<li><p>--out_dir OUT_DIR     directory to output files</p>
</li>
</ul>
</li>
<li><p>optional Mash arguments:</p>
<ul>
<li>--no_mash             skip pre-filtering of genomes using Mash (default: False)</li>
<li>--mash_k MASH_K       k-mer size [1-32] (default: 16)</li>
<li>--mash_s MASH_S       maximum number of non-redundant hashes (default: 5000)</li>
<li>--mash_d MASH_D       maximum distance to keep [0-1] (default: 0.1)</li>
<li>--mash_v MASH_V       maximum p-value to keep [0-1] (default: 1.0)</li>
<li>--mash_db MASH_DB     path to save&#x2F;read (if exists) the Mash reference sketch database (.msh)</li>
</ul>
</li>
<li><p>optional FastANI arguments:</p>
<ul>
<li>--min_af MIN_AF       minimum alignment fraction to consider closest genome (default: 0.65)</li>
</ul>
</li>
<li><p>optional arguments:</p>
<ul>
<li>-x, --extension EXTENSION<br>                extension of files to process, gz &#x3D; gzipped (default: fna)</li>
<li>--prefix PREFIX       prefix for all output files (default: gtdbtk)</li>
<li>--cpus CPUS           number of CPUs to use (default: 1)</li>
<li>--debug               create intermediate files for debugging purposes (default: False)</li>
<li>-h, --help            show help message </details></li>
</ul>
</li>
</ul>
<h3 id="B-4-2-EnrichM-注释"><a href="#B-4-2-EnrichM-注释" class="headerlink" title="B.4.2 EnrichM 注释"></a>B.4.2 EnrichM 注释</h3><h4 id="Step-1-annotate"><a href="#Step-1-annotate" class="headerlink" title="Step 1. annotate"></a>Step 1. annotate</h4><p>基因组注释管道，可以使用dbCAN将基因组与KO, PFAM, TIGRFAM和CAZY 数据库进行比对。结果将为每个基因组产生一个.gff文件，并生成每个注释类型的频率矩阵（frequency matrix），其中行是注释IDs，列是基因组。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ enrichm annotate --genome_directory metawrap_70_5_bins --output EnrichM_annotate --force --ko --ko_hmm --pfam --tigrfam --clusters --orthologs --cazy --ec --threads 30 --parallel 8 --suffix .fa --count_domains --chunk_number 8</span><br></pre></td></tr></table></figure>
<details>
<summary>点击此处查看参数</summary>

<ul>
<li><p>Output options:</p>
<ul>
<li>--output OUTPUT       Output directory</li>
<li>--force               Overwrite previous run</li>
</ul>
</li>
<li><p>Input options (Use one):</p>
<ul>
<li>--genome_files GENOME_FILES [GENOME_FILES …]<br>      Space separated list of genomes to annotate</li>
<li>--genome_directory GENOME_DIRECTORY<br>      Directory containing genomes to annotate</li>
<li>--protein_files PROTEIN_FILES [PROTEIN_FILES …]<br>      Space separated list of .faa files of genomes to annotate. Protein files need to be generated by prodigal.</li>
<li>--protein_directory PROTEIN_DIRECTORY<br>      Directory containing .faa files of genomes to annotate. Protein files need to be generated by prodigal.</li>
</ul>
</li>
<li><p>Annotations options:</p>
<ul>
<li>--ko                  Annotate with KO ids</li>
<li>--ko_hmm              Annotate with KO ids</li>
<li>--pfam                Annotate with Pfam HMMs</li>
<li>--tigrfam             Annotate with TIGRFAM HMMs</li>
<li>--clusters            Annotate with cluster ids</li>
<li>--orthologs           Annotate with ortholog ids</li>
<li>--cazy                Annotate with dbCAN HMMs</li>
<li>--ec                  Annotate with EC ids</li>
</ul>
</li>
<li><p>Cutoff options:</p>
<ul>
<li>--cut_ga              For PFAM and TIGRfam searches: use profiles GA gathering cutoffs to set all thresholding</li>
<li>--cut_nc              For PFAM and TIGRfam searches: use profiles NC noise cutoffs to set all thresholding</li>
<li>--cut_tc              For PFAM and TIGRfam searches: use profiles TC trusted cutoffs to set all thresholding</li>
<li>--cut_ko              For KO HMM annotation searches: use cutoffs defined by KEGG to maximise F-score.</li>
<li>--evalue EVALUE       Use this evalue cutoff to filter false positives (default: 1e-05)</li>
<li>--bit BIT             Use this bit score cutoff to filter false positives (default: 0)</li>
<li>--id ID               Use this percent identity cutoff to filter false positives (default: 0.3)</li>
<li>--aln_query ALN_QUERY<br>      This fraction of the query must align to the reference (default: 0.7)</li>
<li>--aln_reference ALN_REFERENCE<br>      This fraction of the reference must align to the query (default: 0.7)</li>
<li>--c C                 When clustering, use matches above this fraction of aligned (covered) query and target residues (default: 0.7)</li>
</ul>
</li>
<li><p>Runtime options:</p>
<ul>
<li>--threads THREADS     Use this number of threads when annotating with BLAST and HMMsearch (default: 1)</li>
<li>--parallel PARALLEL   Run this number of jobs in parallel when annotating with HMMsearch (default: 5)</li>
<li>--inflation INFLATION<br>     Inflation factor to use when calling clusters in ortholog (default &#x3D; 5.0)</li>
<li>--suffix SUFFIX       Treat files ending with this suffix within the –genome_directory as genomes (default: .fna for –genome_directory and .faa for )</li>
<li>--light               Don’t store metadata for genome files (can’t use enrichM compare downstream, default&#x3D;False)</li>
<li>--count_domains       Fill the frequency matrix with the total number of times an annotation was detected (for example, when one domain more than once within a protein), rather than the count of proteins with with that annotation</li>
<li>--chunk_number CHUNK_NUMBER<br>      Split loading of genomes into this many chunks (default &#x3D; 4)</li>
<li>--chunk_max CHUNK_MAX<br>      Maximum number of genomes to load per chunk (default &#x3D; 2500)</li>
</ul>
</li>
</ul>
</details>

<h4 id="Step-2-classify"><a href="#Step-2-classify" class="headerlink" title="Step 2. classify"></a>Step 2. classify</h4><p>Determine what pathways a genome encodes. Classify quickly reads in KO annotations in the form of a matrix (KO IDs as rows, genomes as columns) and determines which KEGG modules are complete. Annotation matrices can be generated using the annotate function.</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">$ enrichm classify --output EnrichM_classify/ko_hmm_frequency_table --aggregate --genome_and_annotation_matrix EnrichM_annotate/ko_hmm_frequency_table.tsv --cutoff 0</span><br><span class="line"></span><br><span class="line">$ enrichm classify --output EnrichM_classify/ko_frequency_table --aggregate --genome_and_annotation_matrix EnrichM_annotate/ko_frequency_table.tsv --cutoff 0</span><br><span class="line"></span><br><span class="line"><span class="comment"># 以下6个命令可不必运行</span></span><br><span class="line">$ enrichm classify --output EnrichM_classify/cazy_frequency_table --aggregate --genome_and_annotation_matrix EnrichM_annotate/cazy_frequency_table.tsv --cutoff 0</span><br><span class="line"></span><br><span class="line">$ enrichm classify --output EnrichM_classify/cluster_frequency_table --aggregate --genome_and_annotation_matrix EnrichM_annotate/cluster_frequency_table.tsv --cutoff 0</span><br><span class="line"></span><br><span class="line">$ enrichm classify --output EnrichM_classify/ec_frequency_table --aggregate --genome_and_annotation_matrix EnrichM_annotate/ec_frequency_table.tsv --cutoff 0</span><br><span class="line"></span><br><span class="line">$ enrichm classify --output EnrichM_classify/ortholog_frequency_table --aggregate --genome_and_annotation_matrix EnrichM_annotate/ortholog_frequency_table.tsv --cutoff 0</span><br><span class="line"></span><br><span class="line">$ enrichm classify --output EnrichM_classify/pfam_frequency_table --aggregate --genome_and_annotation_matrix EnrichM_annotate/pfam_frequency_table.tsv --cutoff 0</span><br><span class="line"></span><br><span class="line">$ enrichm classify --output EnrichM_classify/tigrfam_frequency_table --aggregate --genome_and_annotation_matrix EnrichM_annotate/tigrfam_frequency_table.tsv --cutoff 0</span><br></pre></td></tr></table></figure>
<details>
<summary>点击此处查看参数</summary>

<ul>
<li><p>Output options:</p>
<ul>
<li>--output OUTPUT       Output directory</li>
<li>--force               Overwrite previous run</li>
</ul>
</li>
<li><p>Input options:</p>
<ul>
<li>--genome_and_annotation_matrix GENOME_AND_ANNOTATION_MATRIX<br>      Path to file containing a genome annotation matrix</li>
<li>--custom_modules CUSTOM_MODULES<br>      Tab separated file containing module name, definition as the columns</li>
<li>--module_rules_json MODULE_RULES_JSON<br>       json file specifying rules to interpret the annotation and guide module annotation</li>
<li>--gff_files GFF_FILES<br>      GFF files for the genomes being classified.</li>
</ul>
</li>
<li><p>Cutoff options:</p>
<ul>
<li>--cutoff CUTOFF       Output only modules with greater than this percent of the requied KO groups (default &#x3D; print all modules)</li>
</ul>
</li>
<li><p>Runtime options:</p>
<ul>
<li>--aggregate           Calculate the abundance of each pathway within each genome&#x2F;sample (column)</details></li>
</ul>
</li>
</ul>
<h4 id="Step-3-enrichment"><a href="#Step-3-enrichment" class="headerlink" title="Step 3. enrichment"></a>Step 3. enrichment</h4><p>Enrichment will read in KO or PFAM annotations in the form of a matrix (IDs as rows, genomes as columns) and a metadata file that separates genomes into groups to compare, and will run some basic stats to determine the enrichment of modules or pfam clans between and within the groups.</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ enrichm enrichment --output EnrichM_enrichment/ --metadata genome.list --annotation_matrix EnrichM_annotate/ko_frequency_table.tsv --force</span><br></pre></td></tr></table></figure>
<details>
<summary>点击此处查看参数</summary>

<ul>
<li><p>- Output options:</p>
<ul>
<li>--output OUTPUT       Output directory</li>
<li>--force               Overwrite previous run</li>
</ul>
</li>
<li><p>Input options:</p>
<ul>
<li>--annotate_output ANNOTATE_OUTPUT<br>     Output directory provided by enrichm annotate</li>
<li>--metadata METADATA   Metadata file with two columns, the first with the genome name, the second with the groupings to compare.</li>
<li>--annotation_matrix ANNOTATION_MATRIX<br>     Annotation matrix to compare.</li>
<li>--gff_files GFF_FILES [GFF_FILES …]<br>     Gff files for genomes to compare.</li>
<li>--abundance 基因组丰度矩阵</li>
<li>--abundance_metadata ABUNDANCE_METADATA<br>     Metadata grouping abundance samples.</li>
<li>--transcriptome TRANSCRIPTOME  基因组丰度矩阵</li>
<li>--transcriptome_metadata TRANSCRIPTOME_METADATA<br>     Metadata grouping abundance samples.</li>
</ul>
</li>
<li><p>Genome Taxonomy DataBase (GTDB) options:</p>
<ul>
<li>--batchfile BATCHFILE<br>     metadata file to compare with.</li>
</ul>
</li>
<li><p>Runtime options:</p>
<ul>
<li>--pval_cutoff PVAL_CUTOFF<br>     Only output results with a p-value below a this cutoff (default&#x3D;0.05).</li>
<li>--proportions_cutoff PROPORTIONS_CUTOFF<br>     Proportion enrichment cutoff.</li>
<li>--threshold THRESHOLD<br>     The threshold to control for in false discovery rate of familywise error rate.</li>
<li>--multi_test_correction MULTI_TEST_CORRECTION<br>     The form of mutiple test correction to use. Uses the statsmodel module and consequently has all of its options.<br>                Default: Benjamini-Hochberg FDR (fdr_bh)<br>                Options: Bonferroni (b)<br>                     Sidak (s)<br>                     Holm (h)<br>                     Holm-Sidak (hs)<br>                     Simes-Hochberg (sh)<br>                     Hommel (ho)<br>                     FDR Benjamini-Yekutieli (fdr_by)<br>                     FDR 2-stage Benjamini-Hochberg (fdr_tsbh)<br>                     FDR 2-stage Benjamini-Krieger-Yekutieli (fdr_tsbky)<br>                     FDR adaptive Gavrilov-Benjamini-Sarkar (fdr_gbs))</li>
<li>--processes PROCESSES  采用多少个进程进行富集分析</li>
<li>--allow_negative_values  允许输入的矩阵中有负值</li>
<li>--ko                    Compare KO ids (annotated with DIAMOND)</li>
<li>--ko_hmm          Compare KO ids (annotated with HMMs)</li>
<li>--pfam                Compare Pfam ids</li>
<li>--tigrfam             Compare TIGRFAM ids</li>
<li>--cluster              Compare cluster ids</li>
<li>--ortholog            Compare ortholog ids</li>
<li>--cazy                 Compare dbCAN ids</li>
<li>--ec                     Compare EC ids</li>
<li>--range RANGE         Base pair range to search for synteny within. Default &#x3D; 2500.</li>
<li>--subblock_size SUBBLOCK_SIZE<br>                Number of genes clustered in a row to be reported. Default &#x3D; 2.</li>
<li>--operon_mismatch_cutoff OPERON_MISMATCH_CUTOFF<br>                Number of allowed mismatches when searching for operons across genomes. Defaul</li>
<li>--operon_match_score_cutoff OPERON_MATCH_SCORE_CUTOFF<br>                Score cutoff for operon matches</li>
</ul>
</li>
</ul>
</details>

<h4 id="Step-4-pathway"><a href="#Step-4-pathway" class="headerlink" title="Step 4. pathway"></a>Step 4. pathway</h4><p>Pathway reads in a KO matrix and generates a Cytoscape-readable metabolic network and metadata file. Only reactions that are possible given the KOs present in the input matrix are shown, and the modules and reactions that are included in the output can be customized（<strong>报错</strong>）.</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ enrichm pathway --matrix EnrichM_annotate/ko_frequency_table.tsv --genome_metadata genome.list --output EnrichM_pathway --abundance EnrichM_enrichment/F01_vs_F02_ivg_results.cdf.tsv --abundance_metadata genome.list --force</span><br></pre></td></tr></table></figure>
<details>
<summary>点击此处查看参数</summary>

<ul>
<li><p>Input options:</p>
<ul>
<li>--matrix KO矩阵. 必须提供</li>
<li>--genome_metadata GENOME_METADATA<br>       Metadata文件包含两列，第一列为基因组名字，第二列with the groupings to compare.</li>
<li>--abundance 丰度矩阵</li>
<li>--abundance_metadata ABUNDANCE_METADATA<br>      Metadata文件包含两列，第一列为基因组名字，第二列with the groupings to compare.</li>
<li>--tpm_values TPM_VALUES<br>       DetectM产生的TPM values</li>
<li>--tpm_metadata TPM_METADATA<br>       Metadata文件包含两列，第一列为基因组名字，第二列with the groupings to compare.</li>
<li>--metabolome METABOLOME<br>       Metabolome CID matrix.</li>
</ul>
</li>
<li><p>Logging options:</p>
<ul>
<li>--log LOG             Output logging information to this file.</li>
<li>--verbosity VERBOSITY<br>     Level of verbosity (1 - 5 - default &#x3D; 4) 5 &#x3D; Very verbose, 1 &#x3D; Silent</li>
</ul>
</li>
<li><p>Output options:</p>
<ul>
<li>--output             输出路径</li>
<li>--force               覆盖之前输出的结果</li>
</ul>
</li>
<li><p>Pathway options:</p>
<ul>
<li>--limit LIMIT [LIMIT …]<br>     USE ONLY these reactions, or reactions within this pathway or module (space separated list).</li>
<li>--filter FILTER [FILTER …]<br>     Do not use these reactions, or reactions within this pathway or module (space separated list).</li>
<li>--enrichment_output ENRICHMENT_OUTPUT<br>      Supply an enrichment output to integrate the results into the output network.</li>
</ul>
</li>
</ul>
</details>

<h4 id="Step-5-explore"><a href="#Step-5-explore" class="headerlink" title="Step 5. explore"></a>Step 5. explore</h4><p>Explore is similar to pathway, but rather than generating a specified pathway it will start from a given query compound ID, and explore the possible reactions that use that compound given the enzymes present in the input KO matrix.</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ </span><br></pre></td></tr></table></figure>
<details>
<summary>点击此处查看参数</summary>

<ul>
<li><p>Input options: </p>
<ul>
<li>--matrix MATRIX       KO矩阵. 必须提供</li>
<li>--genome_metadata GENOME_METADATA<br>      Metadata文件包含两列，第一列为基因组名字，第二列with the groupings to compare.</li>
<li>--abundance 丰度矩阵</li>
<li>--abundance_metadata ABUNDANCE_METADATA<br>     Metadata文件包含两列，第一列为基因组名字，第二列with the groupings to compare..</li>
<li>--tpm_values TPM_VALUES<br>     DetectM产生的TPM values</li>
<li>--tpm_metadata TPM_METADATA<br>     Metadata文件包含两列，第一列为基因组名字，第二列with the groupings to compare..</li>
<li>--metabolome METABOLOME<br>     Metabolome CID matrix.</li>
</ul>
</li>
<li><p>Logging options:</p>
<ul>
<li>--log LOG             Output logging information to this file.</li>
<li>--verbosity VERBOSITY<br>      Level of verbosity (1 - 5 - default &#x3D; 4) 5 &#x3D; Very verbose, 1 &#x3D; Silent</li>
</ul>
</li>
<li><p>Output options:</p>
<ul>
<li>--output             输出路径</li>
<li>--force               覆盖之前输出的结果</li>
</ul>
</li>
<li><p>Query options:</p>
<ul>
<li>--queries QUERIES     A file containing the KEGG ids of the compounds from which to start in the metabolic network</li>
<li>--depth DEPTH         Number of steps to take into the metabolic network</details></li>
</ul>
</li>
</ul>
<h1 id="C-Metagenome-functional-profiling"><a href="#C-Metagenome-functional-profiling" class="headerlink" title="C Metagenome functional profiling"></a>C Metagenome functional profiling</h1><ul>
<li>从宏基因组组装的contigs中预测基因——prodigal -p meta模式</li>
<li>Metagenome-assembled genes which were not included in the MAGs were subjected to taxonomic classification using <a target="_blank" rel="noopener" href="https://github.com/bioinformatics-centre/kaiju">Kaiju</a></li>
<li>eggNOG-mapper比对<a target="_blank" rel="noopener" href="https://doi.org/10.1093/nar/gkv1248">eggnog</a>数据库</li>
<li>HMMER 比对<a target="_blank" rel="noopener" href="https://doi.org/10.1093/nar/gky995">Pfam</a></li>
<li><a target="_blank" rel="noopener" href="https://www.kegg.jp/blastkoala/">GhostKOALA </a>和<a target="_blank" rel="noopener" href="https://www.genome.jp/tools/kofamkoala/">KofamKOALA (v1.0.0)</a>比对<a target="_blank" rel="noopener" href="https://doi.org/10.1093/nar/28.1.27">KEGG</a></li>
<li>BLASTP比对<a target="_blank" rel="noopener" href="https://doi.org/10.1093/nar/gkv1103">TCDB</a></li>
<li><a target="_blank" rel="noopener" href="http://bcb.unl.edu/dbCAN2/">dbCAN2 (v2.0.1)</a>比对<a target="_blank" rel="noopener" href="https://doi.org/10.1093/nar/gkn663">CAZy</a></li>
<li>BLASTP比对<a target="_blank" rel="noopener" href="https://www.ebi.ac.uk/merops/submit_searches.shtml">MEROPS </a></li>
<li><a target="_blank" rel="noopener" href="https://github.com/Arkadiy-Garber/FeGenie">FeGenie</a>检测Iron-related genes</li>
<li>Fe-containing domains were characterized using <a target="_blank" rel="noopener" href="https://doi.org/10.1073/pnas.0605798103">Superfamily (v1.75)</a>.</li>
<li>砷呼吸和抗性基因挖掘<a target="_blank" rel="noopener" href="https://github.com/ShadeLab/PAPER_Dunivin_meta_arsenic/tree/master/HMM_search">https://github.com/ShadeLab/PAPER_Dunivin_meta_arsenic&#x2F;tree&#x2F;master&#x2F;HMM_search</a>，模型下载<a target="_blank" rel="noopener" href="https://github.com/ShadeLab/PAPER_Dunivin_meta_arsenic/tree/master/gene_targeted_assembly/gene_resource">https://github.com/ShadeLab/PAPER_Dunivin_meta_arsenic/tree/master/gene_targeted_assembly/gene_resource</a></li>
</ul>
<h1 id="参考资料："><a href="#参考资料：" class="headerlink" title="参考资料："></a>参考资料：</h1><ul>
<li><a target="_blank" rel="noopener" href="https://github.com/biobakery/MetaPhlAn/wiki/MetaPhlAn-3.0">MetaPhlAn 3.0 tutorial</a></li>
<li><a target="_blank" rel="noopener" href="https://zouhua.top/archives/9d8099c8.html">MetaPhlAn 3.0: 宏基因组物种组成分析软件</a></li>
<li><a target="_blank" rel="noopener" href="https://github.com/biobakery/biobakery/wiki/GraPhlAn">GraPhlAn Tutorial</a></li>
<li></li>
</ul>

      </div>
      <div class="post-tags-categories">
        
        <div class="tags">
          
            <a href="/tags/%E5%AE%8F%E5%9F%BA%E5%9B%A0%E7%BB%84/" class="">
              宏基因组
            </a>
          
        </div>
        
      </div>
      
        <div class="copyright">
  <ul class="post-copyright">
    <li class="post-copyright-author">
    <strong>作者:  </strong>Hualin Liu</a>
    </li>
    <li class="post-copyright-link">
    <strong>文章链接:  </strong>
    <a href="/post/ab1b.html" target="_blank" title="宏基因组分析流程及代码">https://www.liaochenlanruo.fun/post/ab1b.html</a>
    </li>
    <li class="post-copyright-license">
      <strong>版权声明:   </strong>
      本网站所有文章除特别声明外,均采用 <a rel="license" href="https://creativecommons.org/licenses/by-nc-nd/4.0/" target="_blank" title="Attribution-NonCommercial-NoDerivatives 4.0 International (CC BY-NC-ND 4.0)">CC BY-NC-ND 4.0</a>
      许可协议。转载请注明出处!
    </li>
  </ul>
<div>
      
    </article>
    <!-- 上一篇文章和下一篇文章 -->
    
      <!-- 文章详情页的上一页和下一页 -->
<div class="post-nav">



  
  <div class="post-nav-prev post-nav-item">
    <div class="post-nav-img" style="background-size: cover; 
      background-position: center center;">
      <img class="lazyload lazyload placeholder" src="https://fastly.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/custom/bgs/thumb_68.webp" class="lazyload placeholder" data-srcset="https://fastly.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/custom/bgs/thumb_68.webp" srcset="https://pic1.zhimg.com/v2-cd38920285d125be80b3eb504052c550_b.webp" alt="">
    </div>
    <a href="/post/4cf4.html" class="post-nav-link">
      <div class="title">
        <i class="fas fa-angle-left"></i> 上一篇:
        <div class="title-text">Use microeco分析扩增子数据</div>
      </div>
      
      <!-- <div class="content">
        本文阐述使用microeco分析扩增子数据……


1. 安装microeco1234567# If devtools 
      </div> -->
    </a>
  </div>



  
  <div class="post-nav-next post-nav-item">
    <div class="post-nav-img" style="background-size: cover; 
      background-position: center center;">
      <img class="lazyload lazyload placeholder" src="https://fastly.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/custom/bgs/thumb_80.webp" class="lazyload placeholder" data-srcset="https://fastly.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/custom/bgs/thumb_80.webp" srcset="https://pic1.zhimg.com/v2-cd38920285d125be80b3eb504052c550_b.webp" src="" alt="">
    </div>
    <a href="/post/9d1b.html" class="post-nav-link">
      <div class="title">
        下一篇: <i class="fas fa-angle-right"></i>
        <div class="title-text">Linux骚操作</div>
      </div>
      <!-- <div class="content">
        记录一些平时遇到的Linux问题及解决方法！


骚操作1：查杀僵尸进程通过top命令发现了僵尸进程
这里的aragor
      </div> -->
    </a>
  </div>

</div>

    
    

    <!-- 打赏 -->
    
      <div id="appDonate" class="post-donate">
  <div id="donate_board" class="donate_bar center" ref="donate">
    <a id="btn_donate" class="btn_donate" href="javascript:;" title="打赏" @click="showDialogDrawer()"></a>
  </div>
  <transition name="fade">
    <div 
      class="donate-box-mask"
      v-cloak 
      v-show="visible"
      @click="cancelDialogDrawer()"
    >
    </div>
  </transition>
  <transition name="bounce">
    <div class="donate-box" v-cloak v-show="visible">
      <div class="donate-box_close">
        <i class="fas fa-times" aria-hidden="true" @click="cancelDialogDrawer" pointer></i>
      </div>
      <div class="donate-box_title">
        <h4>
          你的赏识是我前进的动力
        </h4>
      </div>
      <div class="donate-box_tab">
        <div class="Alipay" pointer :class="{'active': tabActive === 'Alipay'}" @click="changeTabActive('Alipay')">
          支付宝
        </div>
        <div class="WeChatpay" pointer :class="{'active': tabActive === 'WeChatpay'}" @click="changeTabActive('WeChatpay')">
          微信
        </div>
      </div>
      <div class="donate-box_img">
        <div class="AlipayImg" v-show="tabActive === 'Alipay'">
          <img src="/medias/AliPay.jpg" class="lazyload placeholder" data-srcset="/medias/AliPay.jpg" srcset="https://pic1.zhimg.com/v2-cd38920285d125be80b3eb504052c550_b.webp" alt="支付宝打赏" />
        </div> 
        <div class="WeChatpayImg" v-show="tabActive === 'WeChatpay'">
          <img src="/medias/reward-wepay.jpg" class="lazyload placeholder" data-srcset="/medias/reward-wepay.jpg" srcset="https://pic1.zhimg.com/v2-cd38920285d125be80b3eb504052c550_b.webp" alt="微信打赏" />
        </div>
      </div>
    </div>
  </transition>
</div>

<script>
  var body = document.body || document.documentElement || window;
  var vm = new Vue({
    el: '#appDonate',
    data: {
      visible: false,
      tabActive: 'Alipay',
      top: 0,
    },
    computed: {
    },
    mounted() {
    },
    methods: {
      showDialogDrawer() {
        this.visible = true;
        // 防止页面滚动，只能让弹框滚动
        // function getScroll() {
        //   return {
        //     left: window.pageXOffset || document.documentElement.scrollLeft || document.body.scrollLeft || 0,
        //     top: window.pageYOffset || document.documentElement.scrollTop || document.body.scrollTop || 0
        //   };
        // }
        this.top = $(document).scrollTop() // or getScroll().top
        // console.log('aa', $('.main-content'));
        body.style.cssText = 'overflow: hidden;';
      },
      cancelDialogDrawer() {
        this.visible = false;
        body.removeAttribute('style');
        $(document).scrollTop(this.top)
      },
      changeTabActive(name) {
        this.tabActive = name;
      }
    },
    created() {}
  })
</script>
    

    <!-- 分享 -->
    
      <!-- https://github.com/overtrue/share.js -->
<!-- 文章详情页的分享 -->
<div class="social-share" data-sites="twitter,facebook,google,qq,qzone,wechat,weibo,douban,linkedin" data-wechat-qrcode-helper="<p>微信扫一扫即可分享！</p>"></div>

<script defer src="/js/shareJs/social-share.min.js"></script>
</script>

<style>
  .social-share {
    margin: 20px 0;
  }
</style>


    
    
    <!-- 评论 -->
    <!-- 评论 -->

  <div id="myComment">
    
      <div id="gitment-container"></div>

    
  </div>


<!-- 还需要在后面这个地址里设置script, comment script in themes\hexo-theme-bamboo\layout\_partial\scripts\index.ejs -->


  </div>

  <!-- 目录 -->
  <aside id='l_side'>
  
    
      <section class="widget side_blogger">
  <div class='content'>
    
      
        <a class='avatar flat-box rectangle' href='/about/'>
          <img src='https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/custom/head.jpg'/>
        </a>
      
    
    
      <div class='text'>
        
          <h2>了尘兰若</h2>
        
        
          <p>Research Go</p>

        
        
          <p><span id="jinrishici-sentence">了尘兰若的小坑</span></p>
          <script src="https://sdk.jinrishici.com/v2/browser/jinrishici.js" charset="utf-8"></script>
        
      </div>
    
    
      <div class="social-wrapper">
        
          
            <a href="/atom.xml"
              class="social fas fa-rss flat-btn"
              target="_blank"
              rel="external nofollow noopener noreferrer">
              
            </a>
          
        
          
            <a href="mailto:liaochenlanruo@webmail.hzau.edu.cn"
              class="social fas fa-envelope flat-btn"
              target="_blank"
              rel="external nofollow noopener noreferrer">
              
            </a>
          
        
          
            <a href="http://github.com/liaochenlanruo"
              class="social fab fa-github flat-btn"
              target="_blank"
              rel="external nofollow noopener noreferrer">
              
            </a>
          
        
          
            <a href="tencent://AddContact/?fromId=50&amp;fromSubId=1&amp;subcmd=all&amp;uin=549830082"
              class="social fab fa-qq flat-btn"
              target="_blank"
              rel="external nofollow noopener noreferrer">
              
            </a>
          
        
      </div>
    
  </div>
</section>

    
  
  
  
    
  

  <div class="layout_sticky">    
    
      
<section class="widget side_toc">
  
  <header>
    
      <i style="color: " class="fas fa-list fa-fw" aria-hidden="true"></i><span class='name' style="color: ">本文目录</span>
    
  </header>


  <div class='content'>
    <div class="toc-main">
      <div class="toc-content">
        <!-- <ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#A%E3%80%81%E8%BD%AF%E4%BB%B6%E5%88%97%E8%A1%A8%E5%8F%8A%E5%AE%89%E8%A3%85"><span class="toc-text">A、软件列表及安装</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#A-1-%E5%88%86%E7%B1%BB%E7%9B%B8%E5%85%B3"><span class="toc-text">A.1 分类相关</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#A-1-1-metaphlan-3-0"><span class="toc-text">A.1.1 metaphlan 3.0</span></a><ol class="toc-child"><li class="toc-item toc-level-4"><a class="toc-link" href="#a-%E5%AE%89%E8%A3%85%E4%B8%BB%E6%96%87%E4%BB%B6"><span class="toc-text">a. 安装主文件</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#b-%E5%AE%89%E8%A3%85%E6%95%B0%E6%8D%AE%E5%BA%93"><span class="toc-text">b. 安装数据库</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#c-%E5%AE%89%E8%A3%85%E4%BE%9D%E8%B5%96%E5%8C%85"><span class="toc-text">c. 安装依赖包</span></a></li></ol></li><li class="toc-item toc-level-3"><a class="toc-link" href="#A-1-2-gtdbtk"><span class="toc-text">A.1.2 gtdbtk</span></a><ol class="toc-child"><li class="toc-item toc-level-4"><a class="toc-link" href="#a-Hardware-requirements"><span class="toc-text">a. Hardware requirements</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#b-Install-GTDB-Tk-with-conda"><span class="toc-text">b. Install GTDB-Tk with conda</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#c-GTDB-Tk-reference-data"><span class="toc-text">c. GTDB-Tk reference data</span></a></li></ol></li><li class="toc-item toc-level-3"><a class="toc-link" href="#A-1-3-Kraken2"><span class="toc-text">A.1.3 Kraken2</span></a><ol class="toc-child"><li class="toc-item toc-level-4"><a class="toc-link" href="#a-Hardware-requirements-1"><span class="toc-text">a. Hardware requirements</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#b-Install-Kraken2-with-conda"><span class="toc-text">b. Install Kraken2 with conda</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#c-Build-the-database"><span class="toc-text">c. Build the database</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#d-%E5%BA%8F%E5%88%97%E5%88%86%E7%B1%BB"><span class="toc-text">d. 序列分类</span></a></li></ol></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#A-2-%E7%BB%84%E8%A3%85%E3%80%81Bining%E3%80%81%E8%B4%A8%E9%87%8F%E8%AF%84%E4%BC%B0"><span class="toc-text">A.2 组装、Bining、质量评估</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#A-2-1-metawrap"><span class="toc-text">A.2.1 metawrap</span></a><ol class="toc-child"><li class="toc-item toc-level-4"><a class="toc-link" href="#a-%E5%AE%89%E8%A3%85%E4%B8%BB%E6%96%87%E4%BB%B6-1"><span class="toc-text">a. 安装主文件</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#b-%E5%AE%89%E8%A3%85%E5%85%B6%E4%BB%96%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7%E5%88%B0metawrap%E7%8E%AF%E5%A2%83%E4%B8%AD"><span class="toc-text">b. 安装其他分析工具到metawrap环境中</span></a></li></ol></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#A-3-%E9%9D%9E%E5%86%97%E4%BD%99%E5%9F%BA%E5%9B%A0%E5%8A%9F%E8%83%BD%E6%B3%A8%E9%87%8A"><span class="toc-text">A.3 非冗余基因功能注释</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#A-3-1-eggNOG-mapper"><span class="toc-text">A.3.1 eggNOG-mapper</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#A-3-2-EnrichM"><span class="toc-text">A.3.2 EnrichM</span></a><ol class="toc-child"><li class="toc-item toc-level-4"><a class="toc-link" href="#a-%E5%AE%89%E8%A3%85%E4%B8%BB%E6%96%87%E4%BB%B6-2"><span class="toc-text">a. 安装主文件</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#b-%E5%AE%89%E8%A3%85%E6%95%B0%E6%8D%AE%E5%BA%93-1"><span class="toc-text">b. 安装数据库</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#c-Sepcifying-the-location-of-the-EnrichM-database"><span class="toc-text">c. Sepcifying the location of the EnrichM database</span></a></li></ol></li></ol></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#B%E3%80%81%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90"><span class="toc-text">B、数据分析</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#B-1-%E4%BD%BF%E7%94%A8metaphlan%E4%BB%8EReads%E4%B8%AD%E8%8E%B7%E5%8F%96%E7%89%A9%E7%A7%8D%E5%88%86%E7%B1%BB%E4%BF%A1%E6%81%AF"><span class="toc-text">B.1 使用metaphlan从Reads中获取物种分类信息</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#Step-1-%E6%BF%80%E6%B4%BB%E7%8E%AF%E5%A2%83"><span class="toc-text">Step 1. 激活环境</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#Step-2-%E5%AF%B9paired-end-Reads%E8%BF%9B%E8%A1%8C%E6%B3%A8%E9%87%8A"><span class="toc-text">Step 2. 对paired-end Reads进行注释</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#Step-3-%E6%B1%87%E6%80%BB%E6%89%80%E6%9C%89%E7%9A%84%E7%BB%93%E6%9E%9C%E6%96%87%E4%BB%B6"><span class="toc-text">Step 3. 汇总所有的结果文件</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#Step-4-%E4%BB%8E%E5%90%88%E5%B9%B6%E7%9A%84%E6%96%87%E4%BB%B6%E4%B8%AD%E6%8F%90%E5%8F%96%E7%A7%8D%E6%B0%B4%E5%B9%B3%E7%9A%84%E5%88%86%E7%B1%BB"><span class="toc-text">Step 4. 从合并的文件中提取种水平的分类</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#Step-5-%E7%BB%98%E5%88%B6%E6%A0%B7%E6%9C%AC%E9%97%B4%E7%A7%8D%E6%B0%B4%E5%B9%B3%E7%9A%84%E7%83%AD%E5%9B%BE"><span class="toc-text">Step 5. 绘制样本间种水平的热图</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#Step-6-%E8%AE%A1%E7%AE%97%E6%A0%B7%E6%9C%AC%E9%97%B4%E7%9A%84unifrac%E8%B7%9D%E7%A6%BB"><span class="toc-text">Step 6. 计算样本间的unifrac距离</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#Step-7-%E7%BB%98%E5%88%B6cladogram%E5%9B%BE"><span class="toc-text">Step 7. 绘制cladogram图</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#B-2-%E4%BD%BF%E7%94%A8metawrap%E5%AF%B9Reads%E8%BF%9B%E8%A1%8C%E7%BB%84%E8%A3%85"><span class="toc-text">B.2 使用metawrap对Reads进行组装</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#Step-1-%E6%BF%80%E6%B4%BB%E7%8E%AF%E5%A2%83-1"><span class="toc-text">Step 1. 激活环境</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#Step-2-%E7%BB%84%E8%A3%85"><span class="toc-text">Step 2. 组装</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#B-3-%E4%BD%BF%E7%94%A8metawrap%E5%AF%B9Contigs%E8%BF%9B%E8%A1%8CBining"><span class="toc-text">B.3 使用metawrap对Contigs进行Bining</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#Step-1-%E6%BF%80%E6%B4%BB%E7%8E%AF%E5%A2%83-2"><span class="toc-text">Step 1. 激活环境</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#Step-2-Bining"><span class="toc-text">Step 2. Bining</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#Step-3-%E6%95%B4%E5%90%88%E4%B8%89%E7%A7%8D%E6%96%B9%E6%B3%95%E7%9A%84bins-metabat2-maxbin2-concoct-%E7%BB%93%E6%9E%9C"><span class="toc-text">Step 3. 整合三种方法的bins (metabat2, maxbin2, concoct) 结果</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#Step-4-Blobology%E5%8F%AF%E8%A7%86%E5%8C%96bin"><span class="toc-text">Step 4. Blobology可视化bin</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#Step-5-Bins-%E5%AE%9A%E9%87%8F"><span class="toc-text">Step 5. Bins 定量</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#Step-6-%E9%87%8D%E7%BB%84%E8%A3%85"><span class="toc-text">Step 6. 重组装</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#B-4-MAGs%E6%B3%A8%E9%87%8A"><span class="toc-text">B.4 MAGs注释</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#B-4-1-GTDB-TK-%E8%BF%9B%E8%A1%8C%E7%89%A9%E7%A7%8D%E5%88%86%E7%B1%BB%E5%92%8C%E6%B3%A8%E9%87%8A%EF%BC%8C%E6%9E%84%E5%BB%BA%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2%E6%A0%91"><span class="toc-text">B.4.1 GTDB-TK 进行物种分类和注释，构建系统发育树</span></a><ol class="toc-child"><li class="toc-item toc-level-4"><a class="toc-link" href="#classify-wf%E2%80%94%E2%80%94Classify-workflow-%E5%8C%85%E6%8B%AC%E4%BA%86Step-1-3"><span class="toc-text">classify_wf——Classify workflow (包括了Step 1-3)</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#Step-1-identify%E2%80%94%E2%80%94%E5%9C%A8%E5%9F%BA%E5%9B%A0%E7%BB%84%E4%B8%AD%E5%AF%BB%E6%89%BEmarker-genes"><span class="toc-text">Step 1: identify——在基因组中寻找marker genes</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#Step-2-align"><span class="toc-text">Step 2: align</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#Step-3-classify"><span class="toc-text">Step 3: classify</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#Step-4-export-msa-%E8%BF%99%E4%B8%80%E6%AD%A5%E4%B8%8D%E8%BF%90%E8%A1%8C"><span class="toc-text">Step 4: export_msa (这一步不运行)</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#Step-5-trim-msa-%E8%BF%99%E4%B8%80%E6%AD%A5%E4%B8%8D%E8%BF%90%E8%A1%8C"><span class="toc-text">Step 5: trim_msa (这一步不运行)</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#Step-6-infer"><span class="toc-text">Step 6: infer</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#Step-7-decorate"><span class="toc-text">Step 7: decorate</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#Step-8-root"><span class="toc-text">Step 8: root</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#Step-9-ani-rep%E2%80%94%E2%80%94%E8%AE%A1%E7%AE%97ANI%E5%80%BC"><span class="toc-text">Step 9: ani_rep——计算ANI值</span></a></li></ol></li><li class="toc-item toc-level-3"><a class="toc-link" href="#B-4-2-EnrichM-%E6%B3%A8%E9%87%8A"><span class="toc-text">B.4.2 EnrichM 注释</span></a><ol class="toc-child"><li class="toc-item toc-level-4"><a class="toc-link" href="#Step-1-annotate"><span class="toc-text">Step 1. annotate</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#Step-2-classify"><span class="toc-text">Step 2. classify</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#Step-3-enrichment"><span class="toc-text">Step 3. enrichment</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#Step-4-pathway"><span class="toc-text">Step 4. pathway</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#Step-5-explore"><span class="toc-text">Step 5. explore</span></a></li></ol></li></ol></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#C-Metagenome-functional-profiling"><span class="toc-text">C Metagenome functional profiling</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%8F%82%E8%80%83%E8%B5%84%E6%96%99%EF%BC%9A"><span class="toc-text">参考资料：</span></a></li></ol> -->
        <div class="toc"></div>
      </div>
    </div>
  </div>
</section>
<!-- 手机端目录按钮 -->
<div id="toc-mobile-btn">
  <i class="fas fa-list-ul" aria-hidden="true"></i>
</div>

      
  <section class="widget side_recent_post">
    
  <header>
    
      <a style="color: " href='/tags/'><i class="fas fa-book fa-fw" aria-hidden="true"></i><span class='name'>最新文章</span></a>
    
  </header>


    <div class='content'>
      
      <!-- hash算法 -->
      
      <div class="aside-list">
        
          <div class="aside-list-item">
            
            
            

            <div class="post-img-box">
              <a href="/post/f0aa.html" class="post-img " style="background-size: cover; 
                background-position: center center;">
                <img class="lazyload lazyload placeholder" style="width:100%;height:100%;object-fit:cover;" data-src="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/custom/bgs/thumb_226.webp" class="lazyload placeholder" data-srcset="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/custom/bgs/thumb_226.webp" srcset="https://pic1.zhimg.com/v2-cd38920285d125be80b3eb504052c550_b.webp" alt="">
              </a>
            </div>
            <div class="post-date-title">
              <div>
                
                  <span class="post-date">11-11</span>
                
              </div>
              <a class="post-title" href="/post/f0aa.html">基因组语言模型的机遇与挑战</a>
            </div>
          </div>
        
          <div class="aside-list-item">
            
            
            

            <div class="post-img-box">
              <a href="/post/81e3.html" class="post-img " style="background-size: cover; 
                background-position: center center;">
                <img class="lazyload lazyload placeholder" style="width:100%;height:100%;object-fit:cover;" data-src="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/custom/bgs/thumb_225.webp" class="lazyload placeholder" data-srcset="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/custom/bgs/thumb_225.webp" srcset="https://pic1.zhimg.com/v2-cd38920285d125be80b3eb504052c550_b.webp" alt="">
              </a>
            </div>
            <div class="post-date-title">
              <div>
                
                  <span class="post-date">11-11</span>
                
              </div>
              <a class="post-title" href="/post/81e3.html">AI 在生物信息学的方法革新与应用全景</a>
            </div>
          </div>
        
          <div class="aside-list-item">
            
            
            

            <div class="post-img-box">
              <a href="/post/81e2.html" class="post-img " style="background-size: cover; 
                background-position: center center;">
                <img class="lazyload lazyload placeholder" style="width:100%;height:100%;object-fit:cover;" data-src="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/custom/bgs/thumb_224.webp" class="lazyload placeholder" data-srcset="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/custom/bgs/thumb_224.webp" srcset="https://pic1.zhimg.com/v2-cd38920285d125be80b3eb504052c550_b.webp" alt="">
              </a>
            </div>
            <div class="post-date-title">
              <div>
                
                  <span class="post-date">11-11</span>
                
              </div>
              <a class="post-title" href="/post/81e2.html">BioAI 专辑：解读 AI 重塑生物信息学研究逻辑</a>
            </div>
          </div>
        
          <div class="aside-list-item">
            
            
            

            <div class="post-img-box">
              <a href="/post/7972.html" class="post-img " style="background-size: cover; 
                background-position: center center;">
                <img class="lazyload lazyload placeholder" style="width:100%;height:100%;object-fit:cover;" data-src="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/custom/bgs/thumb_227.webp" class="lazyload placeholder" data-srcset="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/custom/bgs/thumb_227.webp" srcset="https://pic1.zhimg.com/v2-cd38920285d125be80b3eb504052c550_b.webp" alt="">
              </a>
            </div>
            <div class="post-date-title">
              <div>
                
                  <span class="post-date">10-19</span>
                
              </div>
              <a class="post-title" href="/post/7972.html">在Ubuntu中配置Python文字识别环境（基于Tesseract OCR）</a>
            </div>
          </div>
        
          <div class="aside-list-item">
            
            
            

            <div class="post-img-box">
              <a href="/post/81e1.html" class="post-img " style="background-size: cover; 
                background-position: center center;">
                <img class="lazyload lazyload placeholder" style="width:100%;height:100%;object-fit:cover;" data-src="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/custom/bgs/thumb_223.webp" class="lazyload placeholder" data-srcset="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/custom/bgs/thumb_223.webp" srcset="https://pic1.zhimg.com/v2-cd38920285d125be80b3eb504052c550_b.webp" alt="">
              </a>
            </div>
            <div class="post-date-title">
              <div>
                
                  <span class="post-date">10-19</span>
                
              </div>
              <a class="post-title" href="/post/81e1.html">CentOS 7 升级 GCC 教程（以 GCC 9 为例）</a>
            </div>
          </div>
        
      </div>
    </div>
  </section>

    
  </div>
</aside>

  <!-- 图片放大 Wrap images with fancybox support -->
  <script defer src="/js/wrapImage.js"></script>
</div>

<!-- 文章详情页背景图 -->
<div id="appBgSwiper" style="position: fixed;left: 0;top: 0;width: 100%;height: 100%;z-index: -2;"
	:style="{'background-color': bgColor ? bgColor : 'transparent'}">
	<transition-group tag="ul" :name="names">
		<li v-for='(image,index) in img' :key='index' v-show="index === mark" class="bg-swiper-box">
			<img :src="image" class="bg-swiper-img no-lazy">
		</li>
	</transition-group>
</div>
<script>
	var vm = new Vue({
		el: '#appBgSwiper',
		data: {
			names: '' || 'fade' || 'fade', // translate-fade fade
			mark: 0,
			img: [],
			bgColor: '',
			time: null
		},
		methods: {   //添加方法
			change(i, m) {
				if (i > m) {
					// this.names = 'fade';
				} else if (i < m) {
					// this.names = 'fade';
				} else {
					return;
				}
				this.mark = i;
			},
			prev() {
				// this.names = 'fade';
				this.mark--;
				if (this.mark === -1) {
					this.mark = 3;
					return
				}
			},
			next() {
				// this.names = 'fade';
				this.mark++;
				if (this.mark === this.img.length) {
					this.mark = 0;
					return
				}
			},
			autoPlay() {
				// this.names = 'fade';
				this.mark++;
				if (this.mark === this.img.length) {
					this.mark = 0;
					return
				}
			},
			play() {
				let bgImgDelay = '' || '180000'
				let delay = parseInt(bgImgDelay) || 180000;
				this.time = setInterval(this.autoPlay, delay);
			},
			enter() {
				clearInterval(this.time);
			},
			leave() {
				this.play();
			}
		},
		created() {
			this.play()
		},
		beforeDestroy() {
			clearInterval(this.time);
		},
		mounted() {
			let prop = '' || '';
			let isImg = prop.includes('.bmp') || prop.includes('.jpg') || prop.includes('.png') || prop.includes('.tif') || prop.includes('.gif') || prop.includes('.pcx') || prop.includes('.tga') || prop.includes('.exif') || prop.includes('.fpx') || prop.includes('.psd') || prop.includes('.cdr') || prop.includes('.pcd') || prop.includes('.dxf') || prop.includes('.ufo') || prop.includes('.eps') || prop.includes('.ai') || prop.includes('.raw') || prop.includes('.WMF') || prop.includes('.webp') || prop.includes('.jpeg') || prop.includes('http://') || prop.includes('https://')
			if (isImg) {
				let img = prop.split(',');
				let configRoot = '/'
				let arrImg = [];
				img.forEach(el => {
					var Expression = /http(s)?:\/\/([\w-]+\.)+[\w-]+(\/[\w- .\/?%&=]*)?/;
					var objExp = new RegExp(Expression);

					if (objExp.test(el)) {
						// http or https
						arrImg.push(el);
					} else {
						// 非http or https开头
						// 本地文件
						let firstStr = el.charAt(0);
						if (firstStr == '/') {
							el = el.substr(1); // 删除第一个字符 '/',因为 configRoot最后一个字符为 /
						}
						el = configRoot + el;
						arrImg.push(el);
					}
				})
				this.img = arrImg;
			} else {
				this.bgColor = prop;
			}
		}
	})
</script>

<style>
	.bg-swiper-box {
		position: absolute;
		display: block;
		width: 100%;
		height: 100%;
	}

	.bg-swiper-img {
		object-fit: cover;
		width: 100%;
		height: 100%;
	}
</style>




  <script>
  function loadMermaid() {
    if (document.getElementsByClassName('mermaid').length) {
      if (window.mermaidJsLoad) mermaid.init()
      else {
        loadScript('https://unpkg.com/mermaid/dist/mermaid.min.js').then(() => {
          window.mermaidJsLoad = true
          mermaid.initialize({
            theme: 'default',
          })
          if ('true') {
            mermaid.init();
          }
        })
      }
    }
  };
  document.addEventListener("DOMContentLoaded", function () {
    loadMermaid();
  })

  document.addEventListener('pjax:complete', function () {
    loadMermaid();
  })
  
</script>


      </main>
    </div>

    <!-- 页脚 -->
    
  
  
    <!-- 底部鱼儿跳动效果，依赖于jquery-->
<div id="j-fish-skip" style=" position: relative;height: 153px;width: auto;"></div>
<script defer>
  var RENDERER = {
    POINT_INTERVAL: 5,
    FISH_COUNT: 3,
    MAX_INTERVAL_COUNT: 50,
    INIT_HEIGHT_RATE: .5,
    THRESHOLD: 50,
    FISH_COLOR: '',
    init: function () {
      this.setFishColor(); this.setParameters(), this.reconstructMethods(), this.setup(), this.bindEvent(), this.render()
    },
    setFishColor: function () {
      let isDark = JSON.parse(localStorage.getItem('dark')) || JSON.parse('false');
      if (isDark) {
        this.FISH_COLOR = '#222'; // 暗黑色，有时间把这整成一个变量
      } else {
        this.FISH_COLOR = '' || 'rgba(66, 185, 133, 0.8)';
      }
    },
    setParameters: function () {
      this.$window = $(window), this.$container = $("#j-fish-skip"), this.$canvas = $("<canvas />"), this.context = this.$canvas.appendTo(this.$container).get(0).getContext("2d"), this.points = [], this.fishes = [], this.watchIds = []
    },
    createSurfacePoints: function () {
      var t = Math.round(this.width / this.POINT_INTERVAL);
      this.pointInterval = this.width / (t - 1), this.points.push(new SURFACE_POINT(this, 0));
      for (var i = 1; i < t; i++) {
        var e = new SURFACE_POINT(this, i * this.pointInterval),
          h = this.points[i - 1];
        e.setPreviousPoint(h), h.setNextPoint(e), this.points.push(e)
      }
    },
    reconstructMethods: function () {
      this.watchWindowSize = this.watchWindowSize.bind(this), this.jdugeToStopResize = this.jdugeToStopResize.bind(this), this.startEpicenter = this.startEpicenter.bind(this), this.moveEpicenter = this.moveEpicenter.bind(this), this.reverseVertical = this.reverseVertical.bind(this), this.render = this.render.bind(this)
    },
    setup: function () {
      this.points.length = 0, this.fishes.length = 0, this.watchIds.length = 0, this.intervalCount = this.MAX_INTERVAL_COUNT, this.width = this.$container.width(), this.height = this.$container.height(), this.fishCount = this.FISH_COUNT * this.width / 500 * this.height / 500, this.$canvas.attr({
        width: this.width,
        height: this.height
      }), this.reverse = !1, this.fishes.push(new FISH(this)), this.createSurfacePoints()
    },
    watchWindowSize: function () {
      this.clearTimer(), this.tmpWidth = this.$window.width(), this.tmpHeight = this.$window.height(), this.watchIds.push(setTimeout(this.jdugeToStopResize, this.WATCH_INTERVAL))
    },
    clearTimer: function () {
      for (; this.watchIds.length > 0;) clearTimeout(this.watchIds.pop())
    },
    jdugeToStopResize: function () {
      var t = this.$window.width(),
        i = this.$window.height(),
        e = t == this.tmpWidth && i == this.tmpHeight;
      this.tmpWidth = t, this.tmpHeight = i, e && this.setup()
    },
    bindEvent: function () {
      this.$window.on("resize", this.watchWindowSize), this.$container.on("mouseenter", this.startEpicenter), this.$container.on("mousemove", this.moveEpicenter)
    },
    getAxis: function (t) {
      var i = this.$container.offset();
      return {
        x: t.clientX - i.left + this.$window.scrollLeft(),
        y: t.clientY - i.top + this.$window.scrollTop()
      }
    },
    startEpicenter: function (t) {
      this.axis = this.getAxis(t)
    },
    moveEpicenter: function (t) {
      var i = this.getAxis(t);
      this.axis || (this.axis = i), this.generateEpicenter(i.x, i.y, i.y - this.axis.y), this.axis = i
    },
    generateEpicenter: function (t, i, e) {
      if (!(i < this.height / 2 - this.THRESHOLD || i > this.height / 2 + this.THRESHOLD)) {
        var h = Math.round(t / this.pointInterval);
        h < 0 || h >= this.points.length || this.points[h].interfere(i, e)
      }
    },
    reverseVertical: function () {
      this.reverse = !this.reverse;
      for (var t = 0, i = this.fishes.length; t < i; t++) this.fishes[t].reverseVertical()
    },
    controlStatus: function () {
      for (var t = 0, i = this.points.length; t < i; t++) this.points[t].updateSelf();
      for (t = 0, i = this.points.length; t < i; t++) this.points[t].updateNeighbors();
      this.fishes.length < this.fishCount && 0 == --this.intervalCount && (this.intervalCount = this.MAX_INTERVAL_COUNT, this.fishes.push(new FISH(this)))
    },
    render: function () {
      requestAnimationFrame(this.render), this.controlStatus(), this.context.clearRect(0, 0, this.width, this.height), this.context.fillStyle = this.FISH_COLOR;
      for (var t = 0, i = this.fishes.length; t < i; t++) this.fishes[t].render(this.context);
      this.context.save(), this.context.globalCompositeOperation = "xor", this.context.beginPath(), this.context.moveTo(0, this.reverse ? 0 : this.height);
      for (t = 0, i = this.points.length; t < i; t++) this.points[t].render(this.context);
      this.context.lineTo(this.width, this.reverse ? 0 : this.height), this.context.closePath(), this.context.fill(), this.context.restore()
    }
  },
  SURFACE_POINT = function (t, i) {
    this.renderer = t, this.x = i, this.init()
  };
  SURFACE_POINT.prototype = {
    SPRING_CONSTANT: .03,
    SPRING_FRICTION: .9,
    WAVE_SPREAD: .3,
    ACCELARATION_RATE: .01,
    init: function () {
      this.initHeight = this.renderer.height * this.renderer.INIT_HEIGHT_RATE, this.height = this.initHeight, this.fy = 0, this.force = {
        previous: 0,
        next: 0
      }
    },
    setPreviousPoint: function (t) {
      this.previous = t
    },
    setNextPoint: function (t) {
      this.next = t
    },
    interfere: function (t, i) {
      this.fy = this.renderer.height * this.ACCELARATION_RATE * (this.renderer.height - this.height - t >= 0 ? -1 : 1) * Math.abs(i)
    },
    updateSelf: function () {
      this.fy += this.SPRING_CONSTANT * (this.initHeight - this.height), this.fy *= this.SPRING_FRICTION, this.height += this.fy
    },
    updateNeighbors: function () {
      this.previous && (this.force.previous = this.WAVE_SPREAD * (this.height - this.previous.height)), this.next && (this.force.next = this.WAVE_SPREAD * (this.height - this.next.height))
    },
    render: function (t) {
      this.previous && (this.previous.height += this.force.previous, this.previous.fy += this.force.previous), this.next && (this.next.height += this.force.next, this.next.fy += this.force.next), t.lineTo(this.x, this.renderer.height - this.height)
    }
  };
  var FISH = function (t) {
    this.renderer = t, this.init()
  };
  FISH.prototype = {
    GRAVITY: .4,
    init: function () {
      this.direction = Math.random() < .5, this.x = this.direction ? this.renderer.width + this.renderer.THRESHOLD : -this.renderer.THRESHOLD, this.previousY = this.y, this.vx = this.getRandomValue(4, 10) * (this.direction ? -1 : 1), this.renderer.reverse ? (this.y = this.getRandomValue(1 * this.renderer.height / 10, 4 * this.renderer.height / 10), this.vy = this.getRandomValue(2, 5), this.ay = this.getRandomValue(.05, .2)) : (this.y = this.getRandomValue(6 * this.renderer.height / 10, 9 * this.renderer.height / 10), this.vy = this.getRandomValue(-5, -2), this.ay = this.getRandomValue(-.2, -.05)), this.isOut = !1, this.theta = 0, this.phi = 0
    },
    getRandomValue: function (t, i) {
      return t + (i - t) * Math.random()
    },
    reverseVertical: function () {
      this.isOut = !this.isOut, this.ay *= -1
    },
    controlStatus: function (t) {
      this.previousY = this.y, this.x += this.vx, this.y += this.vy, this.vy += this.ay, this.renderer.reverse ? this.y > this.renderer.height * this.renderer.INIT_HEIGHT_RATE ? (this.vy -= this.GRAVITY, this.isOut = !0) : (this.isOut && (this.ay = this.getRandomValue(.05, .2)), this.isOut = !1) : this.y < this.renderer.height * this.renderer.INIT_HEIGHT_RATE ? (this.vy += this.GRAVITY, this.isOut = !0) : (this.isOut && (this.ay = this.getRandomValue(-.2, -.05)), this.isOut = !1), this.isOut || (this.theta += Math.PI / 20, this.theta %= 2 * Math.PI, this.phi += Math.PI / 30, this.phi %= 2 * Math.PI), this.renderer.generateEpicenter(this.x + (this.direction ? -1 : 1) * this.renderer.THRESHOLD, this.y, this.y - this.previousY), (this.vx > 0 && this.x > this.renderer.width + this.renderer.THRESHOLD || this.vx < 0 && this.x < -this.renderer.THRESHOLD) && this.init()
    },
    render: function (t) {
      t.save(), t.translate(this.x, this.y), t.rotate(Math.PI + Math.atan2(this.vy, this.vx)), t.scale(1, this.direction ? 1 : -1), t.beginPath(), t.moveTo(-30, 0), t.bezierCurveTo(-20, 15, 15, 10, 40, 0), t.bezierCurveTo(15, -10, -20, -15, -30, 0), t.fill(), t.save(), t.translate(40, 0), t.scale(.9 + .2 * Math.sin(this.theta), 1), t.beginPath(), t.moveTo(0, 0), t.quadraticCurveTo(5, 10, 20, 8), t.quadraticCurveTo(12, 5, 10, 0), t.quadraticCurveTo(12, -5, 20, -8), t.quadraticCurveTo(5, -10, 0, 0), t.fill(), t.restore(), t.save(), t.translate(-3, 0), t.rotate((Math.PI / 3 + Math.PI / 10 * Math.sin(this.phi)) * (this.renderer.reverse ? -1 : 1)), t.beginPath(), this.renderer.reverse ? (t.moveTo(5, 0), t.bezierCurveTo(10, 10, 10, 30, 0, 40), t.bezierCurveTo(-12, 25, -8, 10, 0, 0)) : (t.moveTo(-5, 0), t.bezierCurveTo(-10, -10, -10, -30, 0, -40), t.bezierCurveTo(12, -25, 8, -10, 0, 0)), t.closePath(), t.fill(), t.restore(), t.restore(), this.controlStatus(t)
    }
  }, $(function () {
    RENDERER.init()
    $('.dark').click(function () {
      setTimeout(() => {
        RENDERER.setFishColor();
        RENDERER.context.fill();
      });
    })
  });
</script>
  
  <div class="footer bg-color">
    <div class="footer-main">
      
        
          <div class="link">
            
              
                <a href="/atom.xml" class="social">
                  
                    <i class="fas fa-rss" aria-hidden="true"></i>
                  
                </a>
              
            
              
                <a href="mailto:liaochenlanruo@webmail.hzau.edu.cn" class="social">
                  
                    <i class="fas fa-envelope" aria-hidden="true"></i>
                  
                </a>
              
            
              
                <a target="_blank" rel="noopener" href="https://github.com/liaochenlanruo" class="social">
                  
                    <i class="fab fa-github" aria-hidden="true"></i>
                  
                </a>
              
            
              
                <a href="tencent://AddContact/?fromId=50&amp;fromSubId=1&amp;subcmd=all&amp;uin=549830082" class="social">
                  
                    <i class="fab fa-qq" aria-hidden="true"></i>
                  
                </a>
              
            
          </div>
        
      
        
          <div class="footer-copyright">
            <p>Copyright © 2018 - 2024 <a target="_blank" rel="noopener" href="https://github.com/liaochenlanruo">liaochenlanruo</a> | Powered by <a target="_blank" rel="noopener" href="https://hexo.io/zh-cn/docs/">Hexo</a> | Theme <a target="_blank" rel="noopener" href="https://github.com/yuang01/theme">Bamboo</a> </p>

          </div>
        
      
        
          
            <!-- 不蒜子统计 -->
            <!-- 不蒜子统计 -->
<span id="busuanzi_container_site_pv">
      <i class="fas fa-eye" aria-hidden="true"></i>本站总访问量：<span id="busuanzi_value_site_pv"></span> 次
</span>
<span class="post-meta-divider">|</span>
<span id="busuanzi_container_site_uv" style='display:none'>
      <i class="fas fa-users" aria-hidden="true"></i>本站访客数：<span id="busuanzi_value_site_uv"></span> 人
</span>

          
        
      
        
          <div class="footer-custom">
            
              <div><p><a target="_blank" rel="noopener" href="http://beian.miit.gov.cn/">鄂ICP备2021021941号</a>    <img src="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/logo/%E5%A4%87%E6%A1%88%E5%9B%BE%E6%A0%87.png" class="lazyload placeholder" data-srcset="https://cdn.jsdelivr.net/gh/liaochenlanruo/cdn@master/img/logo/%E5%A4%87%E6%A1%88%E5%9B%BE%E6%A0%87.png" srcset="https://pic1.zhimg.com/v2-cd38920285d125be80b3eb504052c550_b.webp" alt="备案图标"> <a target="_blank" rel="noopener" href="http://www.beian.gov.cn/portal/registerSystemInfo?recordcode=42011102004725">鄂公网安备 42011102004725号</a></p>
</div>
            
          </div>
        
      
    </div>
  </div>



    <!-- 渲染暗黑按钮 -->
    
      <div class="dark" onclick="toggleDarkMode()">
  <div class="dark-content">
    <i class="fas" id="darkIcon" aria-hidden="true"></i>
  </div>
</div>

<script defer>
  $(function() {
    // 初始化暗黑模式状态
    let isDark = JSON.parse(localStorage.getItem('dark')) || JSON.parse('false');
    updateDarkModeIcon(isDark);
  });

  function toggleDarkMode() {
    const isDark = $(document.body).hasClass('darkModel');
    $(document.body).toggleClass('darkModel');
    localStorage.setItem('dark', !isDark);
    updateDarkModeIcon(!isDark);
  }

  function updateDarkModeIcon(isDark) {
    const iconElement = document.getElementById('darkIcon');
    if (isDark) {
      iconElement.classList.remove('fa-moon');
      iconElement.classList.add('fa-lightbulb');
    } else {
      iconElement.classList.remove('fa-lightbulb');
      iconElement.classList.add('fa-moon');
    }
  }
</script>

    
    <!-- 渲染回到顶部按钮 -->
    
      <div class="goTop top-btn-color" pointer>
  <i class="fas fa-arrow-up" aria-hidden="true"></i>
</div>
<script defer src="/js/goTop.js"></script>

    
    <!-- 渲染左下角音乐播放器 -->
    
      <link rel="stylesheet" href="/js/aplayer/APlayer@1.10.1.min.css">
<style>
.aplayer .aplayer-lrc p {
  
  font-size: 12px;
  font-weight: 700;
  line-height: 16px !important;
}

.aplayer .aplayer-lrc p.aplayer-lrc-current {
  
  font-size: 15px;
  color: #42b983;
}


.aplayer.aplayer-fixed.aplayer-narrow .aplayer-body {
  left: -66px !important;
}

.aplayer.aplayer-fixed.aplayer-narrow .aplayer-body:hover {
  left: 0px !important;
}


</style>
<meting-js  
  class=""
  server="tencent"
  type="playlist"
  id="8062553743"
  fixed='true'
  autoplay='false'
  theme='#42b983'
  loop='all'
  order='random'
  preload='auto'
  volume='0.7'
  list-folded='true'
>
</meting-js>

<!-- <style>
  #aplayer {
    position: fixed;
    left: 0;
    bottom: 300px;
  }
</style> -->
<script defer src="https://unpkg.com/aplayer@1.10.1/dist/APlayer.min.js"></script>
<script defer src="https://unpkg.com/meting@2/dist/Meting.min.js"></script>
    

    <!-- 图片放大 -->
    
      <script src="https://unpkg.com/@fancyapps/ui@5.0/dist/fancybox/fancybox.umd.js"></script>
    

    <!-- 百度解析 -->
    <!-- Baidu Analytics -->

<script async>
    var _hmt = _hmt || [];
    (function () {
        var hm = document.createElement("script");
        hm.src = "https://hm.baidu.com/hm.js?aea6c1471d6299bdd1c87a0836f18675";
        var s = document.getElementsByTagName("script")[0];
        s.parentNode.insertBefore(hm, s);
    })();
</script>

    <!-- Baidu Push -->

<script async>
    (function () {
        var bp = document.createElement('script');
        var curProtocol = window.location.protocol.split(':')[0];
        if (curProtocol === 'https') {
            bp.src = 'https://zz.bdstatic.com/linksubmit/push.js';
        } else {
            bp.src = 'http://push.zhanzhang.baidu.com/push.js';
        }
        var s = document.getElementsByTagName("script")[0];
        s.parentNode.insertBefore(bp, s);
    })();
</script>

    
    <!-- 背景彩带 -->
    
      <script async type="text/javascript" size="100" alpha='0.4' zIndex="-1" src="/js/ribbon.min.js"></script>
    

    <script src="/js/utils/index.js"></script>
    <script src="/js/app.js"></script>
    
    <!-- 文章目录所需js -->
<!-- <link href="/js/tocbot/tocbot.css" rel="stylesheet">
<script src="/js/tocbot/tocbot.min.js"></script> -->

<script src="https://cdnjs.cloudflare.com/ajax/libs/tocbot/4.18.2/tocbot.min.js"></script>


<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/tocbot/4.18.2/tocbot.css">

<script>
  var headerEl = 'h1, h2, h3, h4',  //headers 
    content = '.post-detail',//文章容器
    idArr = {};  //标题数组以确定是否增加索引id
  //add #id
  var option = {
    // Where to render the table of contents.
    tocSelector: '.toc',
    // Where to grab the headings to build the table of contents.
    contentSelector: content,
    // Which headings to grab inside of the contentSelector element.
    headingSelector: headerEl,
    scrollSmooth: true,
    scrollSmoothOffset: -70,
    // headingsOffset: -($(window).height() * 0.4 - 45),
    headingsOffset: -($(window).height() * 0.4 - 70),
    // positionFixedSelector: '.toc-main',
    // positionFixedClass: 'is-position-fixed',
    fixedSidebarOffset: 'auto',
    activeLinkClass: 'is-active-link',
    orderedList: true,
    collapseDepth: 20,
    // onClick: function (e) {},
  }
  if ($('.toc').length > 0) {

    $(content).children(headerEl).each(function () {
      //去除空格以及多余标点
      var headerId = $(this).text().replace(/[\s|\~|`|\!|\@|\#|\$|\%|\^|\&|\*|\(|\)|\_|\+|\=|\||\|\[|\]|\{|\}|\;|\:|\"|\'|\,|\<|\.|\>|\/|\?|\：|\，|\。]/g, '');

      headerId = headerId.toLowerCase();
      if (idArr[headerId]) {
        //id已经存在
        $(this).attr('id', headerId + '-' + idArr[headerId]);
        idArr[headerId]++;
      }
      else {
        //id未存在
        idArr[headerId] = 1;
        $(this).attr('id', headerId);
      }
    });

    document.addEventListener("DOMContentLoaded", function () {
      tocbot.init(option);
      mobileTocClick();
    });

  }

  window.tocScrollFn = function () {
    return bamboo.throttle(function () {
      findHeadPosition();
    }, 100)()
  }
  window.addEventListener('scroll', tocScrollFn);

  const findHeadPosition = function (top) {
    if ($('.toc-list').length <= 0) {
      return false;
    }
    setTimeout(() => {  // or DOMContentLoaded 
      autoScrollToc();
    }, 0);
  }

  const autoScrollToc = function () {
    const $activeItem = document.querySelector('.is-active-link');
    const $cardToc = document.querySelector('.toc-content');
    const activePosition = $activeItem.getBoundingClientRect().top
    const sidebarScrollTop = $cardToc.scrollTop
    if (activePosition > (document.documentElement.clientHeight - 100)) {
      $cardToc.scrollTop = sidebarScrollTop + 150
    }
    if (activePosition < 150) {
      $cardToc.scrollTop = sidebarScrollTop - 150
    }
  }

  document.addEventListener('pjax:send', function () {
    if ($('.toc').length) {
      tocbot.destroy();
    }
  });

  document.addEventListener('pjax:complete', function () {
    if ($('.toc').length) {
      tocbot.init(option);
      mobileTocClick();
    }
  });
  
  // 手机端toc按钮点击出现目录
  const mobileTocClick = function () {
    const $cardTocLayout = document.getElementsByClassName('side_toc')[0];
    const $cardToc = $cardTocLayout.getElementsByClassName('toc-content')[0];
    let right = '45px';
    if (window.innerWidth >= 551 && window.innerWidth <= 992) {
      right = '100px'
    }
    const mobileToc = {
      open: () => {
        $cardTocLayout.style.cssText = 'animation: toc-open .3s; opacity: 1; right: ' + right
      },

      close: () => {
        $cardTocLayout.style.animation = 'toc-close .2s'
        setTimeout(() => {
          $cardTocLayout.style.cssText = "opacity:''; animation: ''; right: ''"
        }, 100)
      }
    }
    document.getElementById('toc-mobile-btn').addEventListener('click', () => {
      if (window.getComputedStyle($cardTocLayout).getPropertyValue('opacity') === '0') mobileToc.open()
      else mobileToc.close()
    })

    $cardToc.addEventListener('click', (e) => {
      if (window.innerWidth < 992) { // 小于992px的时候
        mobileToc.close()
      }
    })
  }
</script>

<style>
  /* .is-position-fixed {
    position: sticky !important;
    top: 74px;
  }

  .toc-main ul {
    counter-reset: show-list;
  }

  .toc-main ul li::before {
    content: counter(item)".";
    display: block;
    position: absolute;
    left: 12px;
    top: 0;
  } */
</style>
 

<!-- 设置导航背景 -->
<script>
  let setHeaderClass = () => {
    const nav = $('#navHeader');
    const navTop = nav.outerHeight();
    const winTop = $(window).scrollTop();
    if(winTop > navTop) {
      nav.addClass('header-bg-color');
    }
    else {
      nav.removeClass('header-bg-color');
    }
  };

  let scrollCollect = () => {
    return bamboo.throttle(function (e) {
      setHeaderClass();
    }, 200)()
  }

  let initHeaderBg = () => {
    setHeaderClass();
  }

  setHeaderClass();
  window.addEventListener('scroll', scrollCollect);

  document.addEventListener('pjax:send', function () {
    window.removeEventListener('scroll', scrollCollect)
  })
  document.addEventListener('pjax:complete', function () {
    window.addEventListener('scroll', scrollCollect);
    setHeaderClass();
  })
</script> 

<!-- 渲染issues标签里的内容 -->
<script>
  function loadIssuesJS() {
    if ($(".post-detail").find(".issues-api").length == 0) {
      return;
    } 
    loadScript('/js/issues/index.js');
  };
  $(function () {
    loadIssuesJS();
  });
  document.addEventListener('pjax:complete', function () {
    if (typeof IssuesAPI == "undefined") {
      loadIssuesJS();
    }
  })
</script>

<!-- 渲染远程json加载的图片标签(getPhotoOnline)里的内容 -->
<script>
  function loadPhotoOnlineJS() {
    if ($(".post-detail").find(".getJsonPhoto-api").length == 0) {
      return;
    } 
    loadScript('/js/getPhotoOnline/index.js');
  };
  $(function () {
    loadPhotoOnlineJS();
  });
  document.addEventListener('pjax:complete', function () {
    if (typeof getPhotoJson == "undefined") {
      loadPhotoOnlineJS();
    }
  })
</script>

<!-- 渲染远程json加载的talk标签(getTalkOnline)里的内容 -->
<script>
  function loadTalkOnlineJS() {
    if ($(".post-detail").find(".getJsonTalk-api").length == 0) {
      return;
    } 
    loadScript('https://cdnjs.cloudflare.com/ajax/libs/waterfall.js/1.0.2/waterfall.min.js'); // 瀑布流插件，https://raphamorim.io/waterfall.js/
    loadScript('/js/getTalkOnline/index.js');
  };
  $(function () {
    loadTalkOnlineJS();
  });
  document.addEventListener('pjax:complete', function () {
    if (typeof getTalkJson == "undefined") {
      loadTalkOnlineJS();
    }
  })
</script>

<!-- 渲染远程json加载的site-card标签(getSiteOnline)里的内容 -->
<script>
  function loadSiteOnlineJS() {
    if ($(".post-detail").find(".getJsonSite-api").length == 0) {
      return;
    } 
    loadScript('/js/getSiteOnline/index.js');
  };
  $(function () {
    loadSiteOnlineJS();
  });
  document.addEventListener('pjax:complete', function () {
    if (typeof getSiteJson == "undefined") {
      loadSiteOnlineJS();
    }
  })
</script>

<!-- 输入框打字特效 -->
<!-- 输入框打字特效 -->

  <script src="/js/activate-power-mode.js"></script>
  <script>
    POWERMODE.colorful = true;  // 打开随机颜色特效
    POWERMODE.shake = false;    // 关闭输入框抖动
    document.body.addEventListener('input', POWERMODE);//监听打字事件
  </script>


<!-- markdown代码一键复制功能 -->

  <link rel="stylesheet" href="https://unpkg.com/v-plugs-ayu/lib/ayu.css">
  <script src="https://unpkg.com/v-plugs-ayu/lib/ayu.umd.min.js"></script>
  <script src="/js/clipboard/clipboard.min.js"></script>
  <div id="appCopy">
  </div>
  <script data-pjax>
    var vm = new Vue({
      el: '#appCopy',
      data: {
      },
      computed: {
      },
      mounted() {
        const that = this;
        var copy = '复制';
        /* code */
        var initCopyCode = function () {
          var copyHtml = '';
          copyHtml += '<button class="btn-copy" data-clipboard-snippet="" style="position:absolute;top:0;right:0;z-index:1;">';
          copyHtml += '<i class="fas fa-copy"></i><span>' + copy + '</span>';
          copyHtml += '</button>';
          $(".post-detail pre").not('.gutter pre').wrap("<div class='codeBox' style='position:relative;width:100%;'></div>")
          $(".post-detail pre").not('.gutter pre').before(copyHtml);
          new ClipboardJS('.btn-copy', {
            target: function (trigger) {
              return trigger.nextElementSibling;
            }
          });
        }
        initCopyCode();
        $('.btn-copy').unbind('click').bind('click', function () {
          doSomething();
        })
        $(document).unbind('keypress').bind('keypress', function (e) {
          if (e.ctrlKey && e.keyCode == 67) {
            doSomething();
          }
        })

        function doSomething() {
          that.$notify({
            title: "成功",
            content: "代码已复制，请遵守相关授权协议。",
            type: 'success'
          })
        }
      },
      methods: {
      },
      created() { }
    })
  </script>
  

<!-- 图片懒加载 -->
<script defer src="https://unpkg.com/vanilla-lazyload@17.1.0/dist/lazyload.min.js"></script>
<script>
  // https://www.npmjs.com/package/vanilla-lazyload
  // Set the options globally
  // to make LazyLoad self-initialize
  window.lazyLoadOptions = {
    elements_selector: ".lazyload",
    threshold: 0
  };
  // Listen to the initialization event
  // and get the instance of LazyLoad
  window.addEventListener(
    "LazyLoad::Initialized",
    function (event) {
      window.lazyLoadInstance = event.detail.instance;
    },
    false
  );
  document.addEventListener('DOMContentLoaded', function () {
    lazyLoadInstance.update();
  });
  document.addEventListener('pjax:complete', function () {
    lazyLoadInstance.update();
  });
</script>


<!-- 卡片滚动动画 -->
   

<!-- 评论所需js -->

  
        <script type="text/javascript">
  var utteranceComment = {};

  function check_utterance() {
    let isDark = JSON.parse(localStorage.getItem('dark')) || JSON.parse('false');
    if (isDark) {
      utteranceComment.Theme = 'github-dark';
    } else {
      utteranceComment.Theme = 'github-light';
    }

    return document.getElementById("gitment-container");
  }
  comment_el = '#gitment-container';
  load_utterance = function () {
    if ($(comment_el).length) {
      // 匿名函数，防止污染全局变量
      const HEAD = check_utterance();

      var utterances = document.createElement('script');
      utterances.type = 'text/javascript';
      utterances.async = true;
      utterances.setAttribute('issue-term', 'pathname')
      utterances.setAttribute('theme', utteranceComment.Theme)
      utterances.setAttribute('repo', 'liaochenlanruo/liaochenlanruo.github.io')
      utterances.crossorigin = 'anonymous';
      utterances.src = 'https://utteranc.es/client.js';
      // content 是要插入评论的地方
      document.getElementById('gitment-container').appendChild(utterances);

    }
  }

  function dark_utterance() {
    const HEAD = check_utterance();
    if (!HEAD) return;
    const message = {
      type: 'set-theme',
      theme: utteranceComment.Theme
    };
    const utteranceIframe = document.querySelector('iframe');
    utteranceIframe.contentWindow.postMessage(message, 'https://utteranc.es');
  }

  $(document).ready(load_utterance);
  document.addEventListener('pjax:complete', function () {
    load_utterance();
  });

  $('.dark').click(function () {
    setTimeout(() => {
      dark_utterance();
    });
  })

</script>

<style>
  .utterances {
    max-width: inherit !important;
  }
</style>
      


<!-- 鼠标点击特效 -->
<!-- 爱心点击 -->

  
    <script async src="/js/cursor/fireworks.js"></script>
  




  <script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js" data-pjax></script>


<!-- 轮播图标签 -->
<script>
  var bambooSwiperTag = {};
  function load_swiper() {
    if (!document.querySelectorAll(".post-swiper-container")[0]) return;
    loadCSS("https://unpkg.com/swiper@6/swiper-bundle.min.css")
    loadScript("https://unpkg.com/swiper@6/swiper-bundle.min.js").then(() => {
      pjax_swiper();
    });
  }

  load_swiper();

  function pjax_swiper() {
    bambooSwiperTag.swiper = new Swiper('.post-swiper-container', {
      slidesPerView: 'auto',
      spaceBetween: 8,
      centeredSlides: true,
      loop: true,
      autoplay: true ? {
        delay: 3000,
        stopOnLastSlide: false,
        disableOnInteraction: false,
      } : false,
      pagination: {
        el: '.swiper-pagination',
        clickable: true,
      },
      navigation: {
        nextEl: '.swiper-button-next',
        prevEl: '.swiper-button-prev',
      },
      on:{
        init: function(){
          swiperAnimateCache(this); //隐藏动画元素 
          swiperAnimate(this); //初始化完成开始动画
        }, 
        slideChangeTransitionEnd: function(){ 
          swiperAnimate(this); //每个slide切换结束时也运行当前slide动画
          //this.slides.eq(this.activeIndex).find('.ani').removeClass('ani'); 动画只展现一次，去除ani类名
        } 
      }
    });
  }

  document.addEventListener('pjax:complete', function () {
    if (!document.querySelectorAll(".post-swiper-container")[0]) return;
    if (typeof bambooSwiperTag.swiper === "undefined") {
      load_swiper();
    } else {
      pjax_swiper();
    }
  });
</script>
    <!-- pjax -->
    

<!-- pjax -->


  <script src="/js/pjax@0.2.8/index.js"></script>
  
    <div class="pjax-animate">
  
    <div class="loading-circle"><div id="loader-circle"></div></div>
    <script>
      window.ShowLoading = function() {
        $(".loading-circle").css("display", "block");
      };
      window.HideLoading = function() {
        $(".loading-circle").css("display", "none");
      }
    </script>
  
	<script>
    document.addEventListener('pjax:complete', function () {
      window.HideLoading();
    })
    document.addEventListener('pjax:send', function () {
      window.ShowLoading();
    })
    document.addEventListener('pjax:error', function () {
      window.HideLoading();
    })
	</script>
</div>

  

  <script>
    var pjax = new Pjax({
      elements: 'a[href]:not([href^="#"]):not([href="javascript:void(0)"]):not([no-pjax])',   // 拦截正常带链接的 a 标签
      selectors: ["#pjax-container","title"],                                   // 根据实际需要确认重载区域
      cacheBust: false,   // url 地址追加时间戳，用以避免浏览器缓存
      timeout: 5000
    });

    document.addEventListener('pjax:send', function (e) {

      try {
        var currentUrl = window.location.pathname;
        var targetUrl = e.triggerElement.href;
        var banUrl = [""];
        if (banUrl[0] != "") {
          banUrl.forEach(item => {
            if(currentUrl.indexOf(item) != -1 || targetUrl.indexOf(item) != -1) {
              window.location.href = targetUrl;
            }
          });
        }
      } catch (error) {}

      $(window).unbind('resize');
      $(window).unbind('scroll');
      $(document).unbind('scroll');
      $(document).unbind('click');
      $('body').unbind('click');

    })
    
    document.addEventListener('pjax:complete', function () {
      $('script[data-pjax], .pjax-reload script').each(function () {
        $(this).parent().append($(this).remove());
      });
    });

    document.addEventListener('pjax:error', function (e) {
      window.location.href = e.triggerElement.href;
    })
    
    // 刷新不从顶部开始
    document.addEventListener("DOMContentLoaded", function () {
      history.scrollRestoration = 'auto';
    })
  </script>



  </body>
</html>